// @unicitylabs/state-transition-sdk/lib/api/json-rpc/JsonRpcHttpTransport.js
import { v4 as uuid } from 'uuid';
import { JsonRpcDataError } from './JsonRpcDataError.js';
import { JsonRpcNetworkError } from './JsonRpcNetworkError.js';
/**
 * JSON-RPC HTTP service.
 */
export class JsonRpcHttpTransport {
    url;
    /**
     * JSON-RPC HTTP service constructor.
     */
    constructor(url) {
        this.url = url;
    }
    /**
     * Send a JSON-RPC request.
     */
    async request(method, params, headers = new Headers()) {
        headers.set('Content-Type', 'application/json');
        const response = await fetch(this.url, {
            body: JSON.stringify({
                id: uuid(),
                jsonrpc: '2.0',
                method,
                params,
            }),
            headers,
            method: 'POST',
        });
        if (!response.ok) {
            throw new JsonRpcNetworkError(response.status, await response.text());
        }
        const data = (await response.json());
        if (data.error) {
            throw new JsonRpcDataError(data.error);
        }
        return data.result;
    }
}
