// @unicitylabs/nostr-js-sdk/dist/cjs/nametag/NametagUtils.js
"use strict";
/**
 * NametagUtils - Privacy-preserving nametag normalization and hashing.
 * Supports phone number normalization to E.164 format.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.NAMETAG_MAX_LENGTH = exports.NAMETAG_MIN_LENGTH = void 0;
exports.sha256Hex = sha256Hex;
exports.normalizeNametag = normalizeNametag;
exports.hashNametag = hashNametag;
exports.areSameNametag = areSameNametag;
exports.formatForDisplay = formatForDisplay;
exports.isPhoneNumber = isPhoneNumber;
exports.isValidNametag = isValidNametag;
exports.hashAddressForTag = hashAddressForTag;
exports.encryptNametag = encryptNametag;
exports.decryptNametag = decryptNametag;
const sha256_1 = require("@noble/hashes/sha256");
const hkdf_1 = require("@noble/hashes/hkdf");
const utils_1 = require("@noble/hashes/utils");
const libphonenumber_js_1 = require("libphonenumber-js");
/**
 * Get the Web Crypto API (works in both Node.js and browser).
 */
async function getWebCrypto() {
    if (typeof globalThis.crypto?.subtle !== 'undefined') {
        return globalThis.crypto;
    }
    // Node.js environment - import webcrypto
    const nodeCrypto = await Promise.resolve().then(() => __importStar(require('crypto')));
    return nodeCrypto.webcrypto;
}
/**
 * Base64 encode (works in both Node.js and browser).
 */
function toBase64(bytes) {
    if (typeof Buffer !== 'undefined') {
        return Buffer.from(bytes).toString('base64');
    }
    return btoa(String.fromCharCode(...Array.from(bytes)));
}
/**
 * Base64 decode (works in both Node.js and browser).
 */
function fromBase64(base64) {
    if (typeof Buffer !== 'undefined') {
        return new Uint8Array(Buffer.from(base64, 'base64'));
    }
    return Uint8Array.from(atob(base64), c => c.charCodeAt(0));
}
/** Salt prefix for nametag hashing */
const NAMETAG_SALT = 'unicity:nametag:';
/** Minimum nametag length (after normalization) */
exports.NAMETAG_MIN_LENGTH = 3;
/** Maximum nametag length (after normalization) */
exports.NAMETAG_MAX_LENGTH = 20;
/** Default country code for phone number normalization */
const DEFAULT_COUNTRY = 'US';
/** Salt prefix for address hashing */
const ADDRESS_SALT = 'unicity:address:';
/**
 * Compute SHA-256 hash of a string.
 * @param input String to hash
 * @returns Hex-encoded hash
 */
function sha256Hex(input) {
    const bytes = new TextEncoder().encode(input);
    return (0, utils_1.bytesToHex)((0, sha256_1.sha256)(bytes));
}
/**
 * Check if a string looks like a phone number.
 * Heuristic: starts with + OR has >50% digits AND >= 7 digits total.
 * @param str String to check
 * @returns true if the string looks like a phone number
 */
function isLikelyPhoneNumber(str) {
    if (str.startsWith('+')) {
        return true;
    }
    const digitsOnly = str.replace(/\D/g, '');
    const digitCount = digitsOnly.length;
    if (digitCount < 7) {
        return false;
    }
    // Count non-digit characters (excluding common phone number chars)
    const cleanedLength = str.replace(/[\s\-().]/g, '').length;
    const digitRatio = digitCount / cleanedLength;
    return digitRatio > 0.5;
}
/**
 * Normalize a phone number to E.164 format.
 * @param phoneNumber Phone number string
 * @param defaultCountry Default country code
 * @returns E.164 formatted phone number, or null if invalid
 */
function normalizePhoneNumber(phoneNumber, defaultCountry) {
    try {
        // Try to parse with default country
        if ((0, libphonenumber_js_1.isValidPhoneNumber)(phoneNumber, defaultCountry)) {
            const parsed = (0, libphonenumber_js_1.parsePhoneNumber)(phoneNumber, defaultCountry);
            return parsed.format('E.164');
        }
        // Try without default country (for numbers with country code)
        if ((0, libphonenumber_js_1.isValidPhoneNumber)(phoneNumber)) {
            const parsed = (0, libphonenumber_js_1.parsePhoneNumber)(phoneNumber);
            return parsed.format('E.164');
        }
        return null;
    }
    catch {
        return null;
    }
}
/**
 * Normalize a nametag for hashing.
 * - If it looks like a phone number, normalize to E.164
 * - Otherwise, lowercase and remove @unicity suffix
 * @param nametag Nametag to normalize
 * @param defaultCountry Default country code for phone normalization
 * @returns Normalized nametag
 */
function normalizeNametag(nametag, defaultCountry = DEFAULT_COUNTRY) {
    const trimmed = nametag.trim();
    if (isLikelyPhoneNumber(trimmed)) {
        const normalized = normalizePhoneNumber(trimmed, defaultCountry);
        if (normalized) {
            return normalized;
        }
        // If phone normalization fails, fall through to standard normalization
    }
    // Standard normalization: lowercase, remove @unicity suffix
    let normalized = trimmed.toLowerCase();
    if (normalized.endsWith('@unicity')) {
        normalized = normalized.slice(0, -8);
    }
    return normalized;
}
/**
 * Hash a nametag with the standard salt.
 * @param nametag Nametag to hash
 * @param defaultCountry Default country code for phone normalization
 * @returns Hex-encoded SHA-256 hash
 */
function hashNametag(nametag, defaultCountry = DEFAULT_COUNTRY) {
    const normalized = normalizeNametag(nametag, defaultCountry);
    return sha256Hex(NAMETAG_SALT + normalized);
}
/**
 * Compare two nametags for equality (handling format variations).
 * @param tag1 First nametag
 * @param tag2 Second nametag
 * @param defaultCountry Default country code for phone normalization
 * @returns true if the nametags represent the same identity
 */
function areSameNametag(tag1, tag2, defaultCountry = DEFAULT_COUNTRY) {
    const normalized1 = normalizeNametag(tag1, defaultCountry);
    const normalized2 = normalizeNametag(tag2, defaultCountry);
    return normalized1 === normalized2;
}
/**
 * Format a nametag for display (privacy-preserving).
 * For phone numbers, hides middle digits.
 * @param nametag Nametag to format
 * @param defaultCountry Default country code for phone normalization
 * @returns Display-safe formatted nametag
 */
function formatForDisplay(nametag, defaultCountry = DEFAULT_COUNTRY) {
    const trimmed = nametag.trim();
    if (isLikelyPhoneNumber(trimmed)) {
        const normalized = normalizePhoneNumber(trimmed, defaultCountry);
        if (normalized) {
            // Hide middle digits: +1415***2671
            const digits = normalized.slice(1); // Remove +
            if (digits.length > 6) {
                const start = digits.slice(0, 4);
                const end = digits.slice(-4);
                return '+' + start + '***' + end;
            }
        }
    }
    return normalizeNametag(nametag, defaultCountry);
}
/**
 * Check if a string is a valid phone number.
 * @param str String to check
 * @param defaultCountry Default country code
 * @returns true if the string is a valid phone number
 */
function isPhoneNumber(str, defaultCountry = DEFAULT_COUNTRY) {
    try {
        return (0, libphonenumber_js_1.isValidPhoneNumber)(str, defaultCountry) ||
            (0, libphonenumber_js_1.isValidPhoneNumber)(str);
    }
    catch {
        return false;
    }
}
/**
 * Validate a nametag string. Strips leading @, normalizes, then checks format.
 * Regular nametags: lowercase alphanumeric, underscore, hyphen, 3-20 chars.
 * Phone numbers: validated via libphonenumber-js.
 * @param nametag Nametag to validate
 * @param defaultCountry Default country code for phone normalization
 * @returns true if the nametag is valid
 */
function isValidNametag(nametag, defaultCountry = DEFAULT_COUNTRY) {
    const stripped = nametag.startsWith('@') ? nametag.slice(1) : nametag;
    const normalized = normalizeNametag(stripped, defaultCountry);
    if (isPhoneNumber(normalized)) {
        return true;
    }
    const pattern = new RegExp(`^[a-z0-9_-]{${exports.NAMETAG_MIN_LENGTH},${exports.NAMETAG_MAX_LENGTH}}$`);
    return pattern.test(normalized);
}
/**
 * Hash an address for use as an indexed relay tag.
 * Enables reverse lookup: address → binding event.
 * @param address Address string (e.g., DIRECT://..., alpha1..., PROXY://...)
 * @returns Hex-encoded SHA-256 hash
 */
function hashAddressForTag(address) {
    return sha256Hex(ADDRESS_SALT + address);
}
/**
 * Derive an AES-256 encryption key from a private key using HKDF-SHA256.
 * @param privateKeyHex Hex-encoded private key
 * @returns 32-byte derived key
 */
function deriveNametagEncryptionKey(privateKeyHex) {
    const privateKeyBytes = hexToBytes(privateKeyHex);
    const saltInput = new TextEncoder().encode('sphere-nametag-salt');
    const salt = (0, sha256_1.sha256)(saltInput);
    const info = new TextEncoder().encode('nametag-encryption');
    return (0, hkdf_1.hkdf)(sha256_1.sha256, privateKeyBytes, salt, info, 32);
}
/**
 * Convert hex string to Uint8Array.
 */
function hexToBytes(hex) {
    const bytes = new Uint8Array(hex.length / 2);
    for (let i = 0; i < hex.length; i += 2) {
        bytes[i / 2] = parseInt(hex.substring(i, i + 2), 16);
    }
    return bytes;
}
/**
 * Encrypt a nametag with AES-GCM using a key derived from the private key.
 * Enables nametag recovery on wallet import.
 * @param nametag Plain text nametag
 * @param privateKeyHex Hex-encoded private key for key derivation
 * @returns Base64-encoded encrypted data (IV + ciphertext + auth tag)
 */
async function encryptNametag(nametag, privateKeyHex) {
    const webCrypto = await getWebCrypto();
    const key = deriveNametagEncryptionKey(privateKeyHex);
    const iv = webCrypto.getRandomValues(new Uint8Array(12));
    const data = new TextEncoder().encode(nametag);
    const cryptoKey = await webCrypto.subtle.importKey('raw', new Uint8Array(key).buffer, { name: 'AES-GCM' }, false, ['encrypt']);
    const encrypted = await webCrypto.subtle.encrypt({ name: 'AES-GCM', iv: new Uint8Array(iv).buffer }, cryptoKey, new Uint8Array(data).buffer);
    const combined = new Uint8Array(iv.length + encrypted.byteLength);
    combined.set(iv, 0);
    combined.set(new Uint8Array(encrypted), iv.length);
    return toBase64(combined);
}
/**
 * Decrypt a nametag encrypted with encryptNametag().
 * @param encryptedBase64 Base64-encoded encrypted data (IV + ciphertext + auth tag)
 * @param privateKeyHex Hex-encoded private key for key derivation
 * @returns Decrypted nametag, or null if decryption fails (wrong key)
 */
async function decryptNametag(encryptedBase64, privateKeyHex) {
    try {
        const webCrypto = await getWebCrypto();
        const key = deriveNametagEncryptionKey(privateKeyHex);
        const combined = fromBase64(encryptedBase64);
        const iv = combined.slice(0, 12);
        const ciphertext = combined.slice(12);
        const cryptoKey = await webCrypto.subtle.importKey('raw', new Uint8Array(key).buffer, { name: 'AES-GCM' }, false, ['decrypt']);
        const decrypted = await webCrypto.subtle.decrypt({ name: 'AES-GCM', iv: new Uint8Array(iv).buffer }, cryptoKey, new Uint8Array(ciphertext).buffer);
        return new TextDecoder().decode(decrypted);
    }
    catch {
        return null;
    }
}
