// @unicitylabs/nostr-js-sdk/dist/cjs/index.js
"use strict";
/**
 * Unicity Nostr SDK
 *
 * A TypeScript SDK for Nostr protocol with Unicity extensions.
 * Works in both Node.js and browser environments.
 *
 * Features:
 * - BIP-340 Schnorr signatures
 * - NIP-04 encrypted direct messages with GZIP compression
 * - Multi-relay WebSocket connections with auto-reconnection
 * - Token transfers over Nostr
 * - Privacy-preserving nametag bindings
 *
 * @packageDocumentation
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.NIP17 = exports.PaymentRequestProtocol = exports.TokenTransferProtocol = exports.NametagBinding = exports.NametagUtils = exports.EventKinds = exports.NIP44 = exports.NIP04 = exports.SchnorrSigner = exports.Bech32 = exports.NostrKeyManager = void 0;
// Core key management
var NostrKeyManager_js_1 = require("./NostrKeyManager.js");
Object.defineProperty(exports, "NostrKeyManager", { enumerable: true, get: function () { return NostrKeyManager_js_1.NostrKeyManager; } });
// Crypto utilities
__exportStar(require("./crypto/index.js"), exports);
var index_js_1 = require("./crypto/index.js");
Object.defineProperty(exports, "Bech32", { enumerable: true, get: function () { return index_js_1.Bech32; } });
Object.defineProperty(exports, "SchnorrSigner", { enumerable: true, get: function () { return index_js_1.SchnorrSigner; } });
Object.defineProperty(exports, "NIP04", { enumerable: true, get: function () { return index_js_1.NIP04; } });
Object.defineProperty(exports, "NIP44", { enumerable: true, get: function () { return index_js_1.NIP44; } });
// Protocol types and classes
__exportStar(require("./protocol/index.js"), exports);
var index_js_2 = require("./protocol/index.js");
Object.defineProperty(exports, "EventKinds", { enumerable: true, get: function () { return index_js_2.EventKinds; } });
// Client
__exportStar(require("./client/index.js"), exports);
// Nametag utilities
__exportStar(require("./nametag/index.js"), exports);
var index_js_3 = require("./nametag/index.js");
Object.defineProperty(exports, "NametagUtils", { enumerable: true, get: function () { return index_js_3.NametagUtils; } });
Object.defineProperty(exports, "NametagBinding", { enumerable: true, get: function () { return index_js_3.NametagBinding; } });
// Token transfer
var index_js_4 = require("./token/index.js");
Object.defineProperty(exports, "TokenTransferProtocol", { enumerable: true, get: function () { return index_js_4.TokenTransferProtocol; } });
// Payment requests
var index_js_5 = require("./payment/index.js");
Object.defineProperty(exports, "PaymentRequestProtocol", { enumerable: true, get: function () { return index_js_5.PaymentRequestProtocol; } });
// NIP-17 Private messaging
__exportStar(require("./messaging/index.js"), exports);
exports.NIP17 = __importStar(require("./messaging/nip17.js"));
