// @unicitylabs/state-transition-sdk/lib/smt/plain/SparseMerkleTreeRootNode.js
import { FinalizedLeafBranch } from './FinalizedLeafBranch.js';
import { SparseMerkleTreePath } from './SparseMerkleTreePath.js';
import { SparseMerkleTreePathStep } from './SparseMerkleTreePathStep.js';
import { CborSerializer } from '../../serialization/cbor/CborSerializer.js';
import { BigintConverter } from '../../util/BigintConverter.js';
import { dedent } from '../../util/StringUtils.js';
import { calculateCommonPath } from '../SparseMerkleTreePathUtils.js';
/**
 * Sparse Merkle Tree root node implementation.
 */
export class SparseMerkleTreeRootNode {
    left;
    right;
    hash;
    path = 1n;
    constructor(left, right, hash) {
        this.left = left;
        this.right = right;
        this.hash = hash;
    }
    /**
     * Creates a new instance of MerkleTreeRootNode.
     * @param left Root node left branch.
     * @param right Root node right branch.
     * @param factory Factory to create data hashers.
     * @return A promise that resolves to a new MerkleTreeRootNode instance.
     */
    static async create(left, right, factory) {
        const hash = await factory
            .create()
            .update(CborSerializer.encodeArray(CborSerializer.encodeByteString(BigintConverter.encode(1n)), CborSerializer.encodeNullable(left?.hash.data, CborSerializer.encodeByteString), CborSerializer.encodeNullable(right?.hash.data, CborSerializer.encodeByteString)))
            .digest();
        return new SparseMerkleTreeRootNode(left ?? null, right ?? null, hash);
    }
    static generatePath(remainingPath, parent) {
        if (parent instanceof FinalizedLeafBranch) {
            return [new SparseMerkleTreePathStep(parent.path, parent.data)];
        }
        const commonPath = calculateCommonPath(remainingPath, parent.path);
        remainingPath = remainingPath >> commonPath.length;
        if (commonPath.path !== parent.path || remainingPath === 1n) {
            return [
                new SparseMerkleTreePathStep(0n, parent.left?.hash.data ?? null),
                new SparseMerkleTreePathStep(parent.path, parent.right?.hash.data ?? null),
            ];
        }
        const isRight = remainingPath & 1n;
        const branch = isRight ? parent.right : parent.left;
        const siblingBranch = isRight ? parent.left : parent.right;
        const step = new SparseMerkleTreePathStep(parent.path, siblingBranch?.hash.data ?? null);
        if (branch === null) {
            return [new SparseMerkleTreePathStep(isRight, null), step];
        }
        return [...this.generatePath(remainingPath, branch), step];
    }
    /**
     * Generates a merkle tree traversal path.
     * @param path The path for which to generate the Merkle tree path.
     * @returns A MerkleTreePath instance representing the path in the tree.
     */
    getPath(path) {
        return new SparseMerkleTreePath(this.hash, SparseMerkleTreeRootNode.generatePath(path, this));
    }
    /**
     * Returns a string representation of the MerkleTreeRootNode.
     */
    toString() {
        return dedent `
      Root: 
        Hash: ${this.hash.toString()}
      Left: 
        ${this.left ? this.left.toString() : 'null'}
      Right: 
        ${this.right ? this.right.toString() : 'null'}`;
    }
}
