// @unicitylabs/state-transition-sdk/lib/smt/plain/SparseMerkleTreePathStep.js
import { CborDeserializer } from '../../serialization/cbor/CborDeserializer.js';
import { CborSerializer } from '../../serialization/cbor/CborSerializer.js';
import { BigintConverter } from '../../util/BigintConverter.js';
import { HexConverter } from '../../util/HexConverter.js';
import { dedent } from '../../util/StringUtils.js';
/**
 * Single step along a plain sparse Merkle tree path.
 */
export class SparseMerkleTreePathStep {
    path;
    _data;
    constructor(path, _data) {
        this.path = path;
        this._data = _data;
        if (path < 0n) {
            throw new Error('Path should be non negative.');
        }
    }
    /**
     * @returns {Uint8Array|null} Copy of the step data bytes, or `null`.
     */
    get data() {
        return this._data ? new Uint8Array(this._data) : null;
    }
    /**
     * Create SparseMerkleTreePathStep from CBOR bytes.
     *
     * @param {Uint8Array} bytes CBOR bytes.
     * @returns {SparseMerkleTreePathStep} Decoded step.
     */
    static fromCBOR(bytes) {
        const data = CborDeserializer.decodeArray(bytes, 2);
        return new SparseMerkleTreePathStep(BigintConverter.decode(CborDeserializer.decodeByteString(data[0])), CborDeserializer.decodeNullable(data[1], CborDeserializer.decodeByteString));
    }
    /**
     * Convert SparseMerkleTreePathStep to CBOR bytes.
     *
     * @returns {Uint8Array} CBOR bytes.
     */
    toCBOR() {
        return CborSerializer.encodeArray(CborSerializer.encodeByteString(BigintConverter.encode(this.path)), CborSerializer.encodeNullable(this._data, CborSerializer.encodeByteString));
    }
    /**
     * @returns {string} String representation of the step.
     */
    toString() {
        return dedent `
      Merkle Tree Path Step
        Path: ${this.path.toString(2)}
        Data: ${this._data ? HexConverter.encode(this._data) : 'null'}`;
    }
}
