// @unicitylabs/nostr-js-sdk/dist/cjs/payment/PaymentRequestProtocol.js
"use strict";
/**
 * PaymentRequestProtocol - Request payments from other users over Nostr.
 * Uses NIP-04 encryption for secure payment request transmission.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResponseStatus = exports.DEFAULT_DEADLINE_MS = void 0;
exports.createPaymentRequestEvent = createPaymentRequestEvent;
exports.parsePaymentRequest = parsePaymentRequest;
exports.isPaymentRequest = isPaymentRequest;
exports.getAmount = getAmount;
exports.getRecipientNametag = getRecipientNametag;
exports.getTarget = getTarget;
exports.getSender = getSender;
exports.formatAmount = formatAmount;
exports.parseAmount = parseAmount;
exports.isExpired = isExpired;
exports.getRemainingTimeMs = getRemainingTimeMs;
exports.createPaymentRequestResponseEvent = createPaymentRequestResponseEvent;
exports.parsePaymentRequestResponse = parsePaymentRequestResponse;
exports.isPaymentRequestResponse = isPaymentRequestResponse;
exports.getResponseStatus = getResponseStatus;
exports.getOriginalEventId = getOriginalEventId;
const Event_js_1 = require("../protocol/Event.js");
const EventKinds = __importStar(require("../protocol/EventKinds.js"));
/** Prefix for payment request messages */
const MESSAGE_PREFIX = 'payment_request:';
/** Prefix for payment request response messages */
const RESPONSE_PREFIX = 'payment_request_response:';
/** Default deadline duration: 5 minutes in milliseconds */
exports.DEFAULT_DEADLINE_MS = 5 * 60 * 1000;
/**
 * Payment request response status.
 */
var ResponseStatus;
(function (ResponseStatus) {
    /** Payment request was declined by the recipient */
    ResponseStatus["DECLINED"] = "DECLINED";
    /** Payment request expired (deadline passed) */
    ResponseStatus["EXPIRED"] = "EXPIRED";
})(ResponseStatus || (exports.ResponseStatus = ResponseStatus = {}));
/**
 * Generate a short unique request ID.
 */
function generateRequestId() {
    const bytes = new Uint8Array(4);
    // eslint-disable-next-line no-undef
    if (typeof crypto !== 'undefined' && crypto.getRandomValues) {
        // eslint-disable-next-line no-undef
        crypto.getRandomValues(bytes);
    }
    else {
        // Fallback for environments without crypto.getRandomValues
        for (let i = 0; i < bytes.length; i++) {
            bytes[i] = Math.floor(Math.random() * 256);
        }
    }
    return Array.from(bytes)
        .map((b) => b.toString(16).padStart(2, '0'))
        .join('');
}
/**
 * Create a payment request event.
 *
 * Event structure:
 * - Kind: 31115 (PAYMENT_REQUEST - Unicity custom)
 * - Tags:
 *   - ["p", "<target_pubkey_hex>"] - Target (who should pay)
 *   - ["type", "payment_request"] - Event type
 *   - ["amount", "<amount>"] - Amount for relay filtering
 *   - ["recipient", "<nametag>"] - Recipient nametag
 * - Content: NIP-04 encrypted "payment_request:{json}"
 *
 * @param keyManager Key manager with signing keys
 * @param targetPubkeyHex Target's public key (who should pay)
 * @param request Payment request details
 * @returns Signed event
 */
async function createPaymentRequestEvent(keyManager, targetPubkeyHex, request) {
    // Generate request ID if not provided
    const requestId = request.requestId || generateRequestId();
    // Calculate deadline: use provided value, or default to 5 minutes from now
    // If explicitly set to null, no deadline
    const deadline = request.deadline === null
        ? null
        : request.deadline !== undefined
            ? request.deadline
            : Date.now() + exports.DEFAULT_DEADLINE_MS;
    // Serialize request to JSON
    const requestJson = JSON.stringify({
        amount: String(request.amount), // Convert to string for JSON compatibility with bigint
        coinId: request.coinId,
        message: request.message,
        recipientNametag: request.recipientNametag,
        requestId: requestId,
        deadline: deadline,
    });
    // Add prefix and encrypt
    const message = MESSAGE_PREFIX + requestJson;
    const encryptedContent = await keyManager.encryptHex(message, targetPubkeyHex);
    // Build tags
    const tags = [
        ['p', targetPubkeyHex],
        ['type', 'payment_request'],
        ['amount', String(request.amount)],
    ];
    if (request.recipientNametag) {
        tags.push(['recipient', request.recipientNametag]);
    }
    const event = Event_js_1.Event.create(keyManager, {
        kind: EventKinds.PAYMENT_REQUEST,
        tags,
        content: encryptedContent,
    });
    return event;
}
/**
 * Parse a payment request event.
 * Decrypts and parses the payment request data.
 *
 * @param event Payment request event
 * @param keyManager Key manager for decryption
 * @returns Parsed payment request
 * @throws Error if the event is not a valid payment request
 */
async function parsePaymentRequest(event, keyManager) {
    // Verify event kind
    if (event.kind !== EventKinds.PAYMENT_REQUEST) {
        throw new Error('Event is not a payment request');
    }
    // Verify event type tag
    const eventType = event.getTagValue('type');
    if (eventType !== 'payment_request') {
        throw new Error('Event type is not payment_request');
    }
    // Determine the peer's public key for decryption
    let peerPubkeyHex;
    if (keyManager.isMyPublicKey(event.pubkey)) {
        // We sent this event, decrypt with target's key
        const targetPubkey = event.getTagValue('p');
        if (!targetPubkey) {
            throw new Error('No target found in event');
        }
        peerPubkeyHex = targetPubkey;
    }
    else {
        // We received this event, decrypt with sender's key
        peerPubkeyHex = event.pubkey;
    }
    // Decrypt the content
    const decrypted = await keyManager.decryptHex(event.content, peerPubkeyHex);
    // Validate prefix
    if (!decrypted.startsWith(MESSAGE_PREFIX)) {
        throw new Error('Invalid payment request format: missing prefix');
    }
    // Parse JSON
    const requestJson = decrypted.slice(MESSAGE_PREFIX.length);
    const parsed = JSON.parse(requestJson);
    return {
        amount: BigInt(parsed.amount),
        coinId: parsed.coinId,
        message: parsed.message,
        recipientNametag: parsed.recipientNametag,
        requestId: parsed.requestId,
        senderPubkey: event.pubkey,
        timestamp: event.created_at * 1000, // Convert to milliseconds
        eventId: event.id,
        deadline: parsed.deadline !== undefined ? parsed.deadline : null,
    };
}
/**
 * Check if an event is a payment request.
 * @param event Event to check
 * @returns true if the event is a payment request
 */
function isPaymentRequest(event) {
    return (event.kind === EventKinds.PAYMENT_REQUEST &&
        event.getTagValue('type') === 'payment_request');
}
/**
 * Get the amount from a payment request event (from unencrypted tag).
 * @param event Payment request event
 * @returns Amount, or undefined if not specified
 */
function getAmount(event) {
    const amount = event.getTagValue('amount');
    if (amount === undefined) {
        return undefined;
    }
    try {
        return BigInt(amount);
    }
    catch {
        return undefined;
    }
}
/**
 * Get the recipient nametag from a payment request event (from unencrypted tag).
 * @param event Payment request event
 * @returns Recipient nametag, or undefined if not specified
 */
function getRecipientNametag(event) {
    return event.getTagValue('recipient');
}
/**
 * Get the target public key from a payment request event.
 * @param event Payment request event
 * @returns Target public key (hex), or undefined if not found
 */
function getTarget(event) {
    return event.getTagValue('p');
}
/**
 * Get the sender public key from a payment request event.
 * @param event Payment request event
 * @returns Sender public key (hex)
 */
function getSender(event) {
    return event.pubkey;
}
/**
 * Format an amount for display with proper decimals.
 * @param amount Amount in smallest units
 * @param decimals Number of decimal places (default: 8)
 * @returns Formatted amount string
 */
function formatAmount(amount, decimals = 8) {
    const amountBigInt = typeof amount === 'bigint' ? amount : BigInt(amount);
    const divisor = BigInt(10) ** BigInt(decimals);
    const wholePart = amountBigInt / divisor;
    const fractionalPart = amountBigInt % divisor;
    if (fractionalPart === BigInt(0)) {
        return wholePart.toString();
    }
    // Format fractional part with leading zeros
    const fractionalStr = fractionalPart.toString().padStart(decimals, '0');
    // Remove trailing zeros
    const trimmedFractional = fractionalStr.replace(/0+$/, '');
    return `${wholePart}.${trimmedFractional}`;
}
/**
 * Parse an amount string to smallest units.
 * @param amountStr Amount string (e.g., "1.5")
 * @param decimals Number of decimal places (default: 8)
 * @returns Amount in smallest units
 */
function parseAmount(amountStr, decimals = 8) {
    const multiplier = BigInt(10) ** BigInt(decimals);
    const parts = amountStr.split('.');
    const wholePart = BigInt(parts[0] || '0');
    if (parts.length === 1) {
        return wholePart * multiplier;
    }
    // Handle fractional part
    let fractionalStr = parts[1] || '0';
    // Pad or truncate to correct number of decimals
    if (fractionalStr.length < decimals) {
        fractionalStr = fractionalStr.padEnd(decimals, '0');
    }
    else if (fractionalStr.length > decimals) {
        fractionalStr = fractionalStr.slice(0, decimals);
    }
    const fractionalPart = BigInt(fractionalStr);
    return wholePart * multiplier + fractionalPart;
}
// ============================================================================
// Payment Request Response Functions
// ============================================================================
/**
 * Check if a parsed payment request has expired.
 * @param request Parsed payment request
 * @returns true if the request has a deadline and it has passed
 */
function isExpired(request) {
    return request.deadline !== null && Date.now() > request.deadline;
}
/**
 * Get remaining time until deadline in milliseconds.
 * @param request Parsed payment request
 * @returns Remaining time in ms, 0 if expired, null if no deadline
 */
function getRemainingTimeMs(request) {
    if (request.deadline === null)
        return null;
    const remaining = request.deadline - Date.now();
    return remaining > 0 ? remaining : 0;
}
/**
 * Create a payment request response event (for decline/expiration).
 *
 * Event structure:
 * - Kind: 31116 (PAYMENT_REQUEST_RESPONSE)
 * - Tags:
 *   - ["p", "<target_pubkey_hex>"] - Original requester
 *   - ["type", "payment_request_response"]
 *   - ["status", "DECLINED" | "EXPIRED"]
 *   - ["e", "<original_event_id>", "", "reply"] - Reference to original request
 * - Content: NIP-04 encrypted response JSON
 *
 * @param keyManager Key manager with signing keys
 * @param targetPubkeyHex Original requester's public key
 * @param response Response details
 * @returns Signed event
 */
async function createPaymentRequestResponseEvent(keyManager, targetPubkeyHex, response) {
    // Serialize response to JSON
    const responseJson = JSON.stringify({
        requestId: response.requestId,
        originalEventId: response.originalEventId,
        status: response.status,
        reason: response.reason,
    });
    // Add prefix and encrypt
    const message = RESPONSE_PREFIX + responseJson;
    const encryptedContent = await keyManager.encryptHex(message, targetPubkeyHex);
    // Build tags
    const tags = [
        ['p', targetPubkeyHex],
        ['type', 'payment_request_response'],
        ['status', response.status],
    ];
    // Add reference to original event
    if (response.originalEventId) {
        tags.push(['e', response.originalEventId, '', 'reply']);
    }
    const event = Event_js_1.Event.create(keyManager, {
        kind: EventKinds.PAYMENT_REQUEST_RESPONSE,
        tags,
        content: encryptedContent,
    });
    return event;
}
/**
 * Parse a payment request response event.
 * Decrypts and parses the response data.
 *
 * @param event Payment request response event
 * @param keyManager Key manager for decryption
 * @returns Parsed payment request response
 * @throws Error if the event is not a valid payment request response
 */
async function parsePaymentRequestResponse(event, keyManager) {
    // Verify event kind
    if (event.kind !== EventKinds.PAYMENT_REQUEST_RESPONSE) {
        throw new Error('Event is not a payment request response');
    }
    // Determine the peer's public key for decryption
    let peerPubkeyHex;
    if (keyManager.isMyPublicKey(event.pubkey)) {
        // We sent this response, decrypt with target's key
        const targetPubkey = event.getTagValue('p');
        if (!targetPubkey) {
            throw new Error('No target found in event');
        }
        peerPubkeyHex = targetPubkey;
    }
    else {
        // We received this response, decrypt with sender's key
        peerPubkeyHex = event.pubkey;
    }
    // Decrypt the content
    const decrypted = await keyManager.decryptHex(event.content, peerPubkeyHex);
    // Validate prefix
    if (!decrypted.startsWith(RESPONSE_PREFIX)) {
        throw new Error('Invalid payment request response format: missing prefix');
    }
    // Parse JSON
    const responseJson = decrypted.slice(RESPONSE_PREFIX.length);
    const parsed = JSON.parse(responseJson);
    return {
        requestId: parsed.requestId,
        originalEventId: parsed.originalEventId,
        status: parsed.status,
        reason: parsed.reason,
        senderPubkey: event.pubkey,
        eventId: event.id,
        timestamp: event.created_at * 1000,
    };
}
/**
 * Check if an event is a payment request response.
 * @param event Event to check
 * @returns true if the event is a payment request response
 */
function isPaymentRequestResponse(event) {
    return (event.kind === EventKinds.PAYMENT_REQUEST_RESPONSE &&
        event.getTagValue('type') === 'payment_request_response');
}
/**
 * Get the response status from a payment request response event (from unencrypted tag).
 * @param event Payment request response event
 * @returns Status string, or undefined if not found
 */
function getResponseStatus(event) {
    return event.getTagValue('status');
}
/**
 * Get the referenced original event ID from the response event.
 * @param event Payment request response event
 * @returns Original event ID, or undefined if not found
 */
function getOriginalEventId(event) {
    return event.getTagValue('e');
}
