// @unicitylabs/state-transition-sdk/lib/payment/SplitAssetProof.js
import { AssetId } from './asset/AssetId.js';
import { CborDeserializer } from '../serialization/cbor/CborDeserializer.js';
import { CborSerializer } from '../serialization/cbor/CborSerializer.js';
import { SparseMerkleTreePath } from '../smt/plain/SparseMerkleTreePath.js';
import { SparseMerkleSumTreePath } from '../smt/sum/SparseMerkleSumTreePath.js';
/**
 * Inclusion proof for a single asset within a split payment.
 */
export class SplitAssetProof {
    assetId;
    aggregationPath;
    assetTreePath;
    constructor(assetId, aggregationPath, assetTreePath) {
        this.assetId = assetId;
        this.aggregationPath = aggregationPath;
        this.assetTreePath = assetTreePath;
    }
    /**
     * Create a SplitAssetProof.
     *
     * @param {AssetId} assetId Asset id.
     * @param {SparseMerkleTreePath} aggregationPath Aggregation tree path.
     * @param {SparseMerkleSumTreePath} assetTreePath Asset tree path.
     * @returns {SplitAssetProof} New proof.
     */
    static create(assetId, aggregationPath, assetTreePath) {
        return new SplitAssetProof(assetId, aggregationPath, assetTreePath);
    }
    /**
     * Create split mint reason from CBOR bytes.
     *
     * @param bytes CBOR bytes
     * @return split mint reason proof
     */
    static fromCBOR(bytes) {
        const data = CborDeserializer.decodeArray(bytes, 3);
        return new SplitAssetProof(AssetId.fromCBOR(data[0]), SparseMerkleTreePath.fromCBOR(data[1]), SparseMerkleSumTreePath.fromCBOR(data[2]));
    }
    /**
     * Convert SplitAssetProof to CBOR bytes.
     *
     * @returns {Uint8Array} CBOR bytes.
     */
    toCBOR() {
        return CborSerializer.encodeArray(this.assetId.toCBOR(), this.aggregationPath.toCBOR(), this.assetTreePath.toCBOR());
    }
}
