// @unicitylabs/state-transition-sdk/lib/payment/SplitTokenRequest.js
import { TokenSalt } from '../transaction/TokenSalt.js';
import { TokenType } from '../transaction/TokenType.js';
/**
 * Request to mint one new token as part of a token split.
 */
export class SplitTokenRequest {
    recipient;
    tokenType;
    assets;
    salt;
    constructor(recipient, tokenType, assets, salt) {
        this.recipient = recipient;
        this.tokenType = tokenType;
        this.assets = assets;
        this.salt = salt;
    }
    /**
     * Create a SplitTokenRequest.
     *
     * @param {IPredicate} recipient Predicate that will lock the new token.
     * @param {PaymentAssetCollection} assets Assets the new token will receive.
     * @param {TokenType} tokenType Token type for the new token; defaults to a random token type.
     * @param {TokenSalt} salt Salt for the new token; defaults to a random 32-byte salt.
     * @returns {SplitTokenRequest} New request.
     */
    static create(recipient, assets, tokenType = TokenType.generate(), salt = TokenSalt.generate()) {
        return new SplitTokenRequest(recipient, tokenType, assets, salt);
    }
}
