// @unicitylabs/state-transition-sdk/lib/smt/sum/FinalizedLeafBranch.js
import { HexConverter } from '../../util/HexConverter.js';
import { dedent } from '../../util/StringUtils.js';
/**
 * Finalized leaf in a sparse Merkle sum tree.
 */
export class FinalizedLeafBranch {
    path;
    _data;
    value;
    hash;
    constructor(path, _data, value, hash) {
        this.path = path;
        this._data = _data;
        this.value = value;
        this.hash = hash;
        this._data = new Uint8Array(_data);
    }
    /**
     * @returns {Uint8Array} Copy of the leaf data bytes.
     */
    get data() {
        return new Uint8Array(this._data);
    }
    /**
     * @returns {Promise<FinalizedLeafBranch>} This branch (already finalized).
     */
    finalize() {
        return Promise.resolve(this);
    }
    /**
     * @returns {string} String representation of the leaf.
     */
    toString() {
        return dedent `
      Leaf[${this.path.toString(2)}]
        Hash: ${this.hash.toString()}
        Data: ${HexConverter.encode(this._data)}
        Sum: ${this.value}`;
    }
}
