// @unicitylabs/state-transition-sdk/lib/smt/sum/SparseMerkleSumTreeRootNode.js
import { FinalizedLeafBranch } from './FinalizedLeafBranch.js';
import { SparseMerkleSumTreePath } from './SparseMerkleSumTreePath.js';
import { SparseMerkleSumTreePathStep } from './SparseMerkleSumTreePathStep.js';
import { CborSerializer } from '../../serialization/cbor/CborSerializer.js';
import { BigintConverter } from '../../util/BigintConverter.js';
import { dedent } from '../../util/StringUtils.js';
import { calculateCommonPath } from '../SparseMerkleTreePathUtils.js';
/**
 * Sparse Merkle Sum Tree root node implementation.
 */
export class SparseMerkleSumTreeRootNode {
    left;
    right;
    value;
    hash;
    path = 1n;
    constructor(left, right, value, hash) {
        this.left = left;
        this.right = right;
        this.value = value;
        this.hash = hash;
    }
    /**
     * Creates a new instance of MerkleSumTreeRootNode.
     * @param left Root node left branch.
     * @param right Root node right branch.
     * @param factory Factory to create data hashers.
     * @return A promise that resolves to a new MerkleSumTreeRootNode instance.
     */
    static async create(left, right, factory) {
        const hash = await factory
            .create()
            .update(CborSerializer.encodeArray(CborSerializer.encodeByteString(BigintConverter.encode(1n)), CborSerializer.encodeNullable(left?.hash.data, CborSerializer.encodeByteString), CborSerializer.encodeByteString(BigintConverter.encode(left?.value ?? 0n)), CborSerializer.encodeNullable(right?.hash.data, CborSerializer.encodeByteString), CborSerializer.encodeByteString(BigintConverter.encode(right?.value ?? 0n))))
            .digest();
        return new SparseMerkleSumTreeRootNode(left ?? null, right ?? null, (left?.value ?? 0n) + (right?.value ?? 0n), hash);
    }
    static generatePath(remainingPath, parent) {
        if (parent instanceof FinalizedLeafBranch) {
            return [new SparseMerkleSumTreePathStep(parent.path, parent.data, parent.value)];
        }
        const commonPath = calculateCommonPath(remainingPath, parent.path);
        remainingPath = remainingPath >> commonPath.length;
        if (commonPath.path !== parent.path || remainingPath === 1n) {
            return [
                new SparseMerkleSumTreePathStep(0n, parent.left?.hash.data ?? null, parent.left?.value ?? 0n),
                new SparseMerkleSumTreePathStep(parent.path, parent.right?.hash.data ?? null, parent.right?.value ?? 0n),
            ];
        }
        const isRight = remainingPath & 1n;
        const branch = isRight ? parent.right : parent.left;
        const siblingBranch = isRight ? parent.left : parent.right;
        const step = new SparseMerkleSumTreePathStep(parent.path, siblingBranch?.hash.data ?? null, siblingBranch?.value ?? 0n);
        if (branch === null) {
            return [new SparseMerkleSumTreePathStep(isRight, null, 0n), step];
        }
        return [...this.generatePath(remainingPath, branch), step];
    }
    /**
     * Generates a merkle tree traversal path.
     * @param path The path to create the MerkleSumTreePath for.
     * @returns A MerkleSumTreePath for the given path.
     */
    getPath(path) {
        return new SparseMerkleSumTreePath(this.hash, SparseMerkleSumTreeRootNode.generatePath(path, this));
    }
    /**
     * Returns a string representation of the MerkleSumTreeRootNode.
     */
    toString() {
        return dedent `
      Left: 
        ${this.left ? this.left.toString() : 'null'}
      Right: 
        ${this.right ? this.right.toString() : 'null'}`;
    }
}
