// @unicitylabs/state-transition-sdk/lib/smt/sum/SparseMerkleSumTreePath.js
import { bitLen } from '@noble/curves/utils.js';
import { SparseMerkleSumTreePathStep } from './SparseMerkleSumTreePathStep.js';
import { DataHash } from '../../crypto/hash/DataHash.js';
import { DataHasher } from '../../crypto/hash/DataHasher.js';
import { InvalidJsonStructureError } from '../../InvalidJsonStructureError.js';
import { CborDeserializer } from '../../serialization/cbor/CborDeserializer.js';
import { CborSerializer } from '../../serialization/cbor/CborSerializer.js';
import { BigintConverter } from '../../util/BigintConverter.js';
import { HexConverter } from '../../util/HexConverter.js';
import { dedent } from '../../util/StringUtils.js';
import { PathVerificationResult } from '../PathVerificationResult.js';
/**
 * Path through a sparse Merkle sum tree from a leaf to the root.
 */
export class SparseMerkleSumTreePath {
    root;
    steps;
    constructor(root, steps) {
        this.root = root;
        this.steps = steps;
    }
    /**
     * Create SparseMerkleSumTreePath from CBOR bytes.
     *
     * @param {Uint8Array} bytes CBOR bytes.
     * @returns {SparseMerkleSumTreePath} Decoded path.
     */
    static fromCBOR(bytes) {
        const data = CborDeserializer.decodeArray(bytes, 2);
        const steps = CborDeserializer.decodeArray(data[1]);
        return new SparseMerkleSumTreePath(DataHash.fromImprint(CborDeserializer.decodeByteString(data[0])), steps.map((step) => SparseMerkleSumTreePathStep.fromCBOR(step)));
    }
    /**
     * Create SparseMerkleSumTreePath from JSON.
     *
     * @param {unknown} data JSON object.
     * @returns {SparseMerkleSumTreePath} Decoded path.
     * @throws {InvalidJsonStructureError} If the JSON does not match the expected shape.
     */
    static fromJSON(data) {
        if (!SparseMerkleSumTreePath.isJSON(data)) {
            throw new InvalidJsonStructureError();
        }
        return new SparseMerkleSumTreePath(DataHash.fromImprint(HexConverter.decode(data.root)), data.steps.map((step) => SparseMerkleSumTreePathStep.fromJSON(step)));
    }
    /**
     * Type guard for the JSON shape of a path.
     *
     * @param {unknown} data Value to test.
     * @returns {boolean} True if `data` matches {@link ISparseMerkleSumTreePathJson}.
     */
    static isJSON(data) {
        return (typeof data === 'object' &&
            data !== null &&
            'root' in data &&
            typeof data.root === 'string' &&
            'steps' in data &&
            Array.isArray(data.steps) &&
            data.steps.length > 0);
    }
    /**
     * Convert SparseMerkleSumTreePath to CBOR bytes.
     *
     * @returns {Uint8Array} CBOR bytes.
     */
    toCBOR() {
        return CborSerializer.encodeArray(CborSerializer.encodeByteString(this.root.imprint), CborSerializer.encodeArray(...this.steps.map((step) => step.toCBOR())));
    }
    /**
     * @returns {string} String representation of the path.
     */
    toString() {
        return dedent `
      Merkle Tree Path
        Root: ${this.root.toString()} 
        Steps: [
          ${this.steps.map((step) => step?.toString() ?? 'null').join('\n')}
        ]`;
    }
    /**
     * Verifies the tree path against the root hash and state ID.
     * @param stateId The state ID as bit string in bigint format to verify against the path.
     * @returns A Promise resolving to a PathVerificationResult indicating success or failure.
     */
    async verify(stateId) {
        let step = this.steps[0];
        let currentData;
        let currentPath = step.path;
        let currentSum = step.value;
        if (step.path > 0) {
            const hash = await new DataHasher(this.root.algorithm)
                .update(CborSerializer.encodeArray(CborSerializer.encodeByteString(BigintConverter.encode(step.path)), CborSerializer.encodeNullable(step.data, CborSerializer.encodeByteString), CborSerializer.encodeByteString(BigintConverter.encode(step.value))))
                .digest();
            currentData = hash.data;
        }
        else {
            currentPath = 1n;
            currentData = step.data;
        }
        for (let i = 1; i < this.steps.length; i++) {
            step = this.steps[i];
            const isRight = this.steps[i - 1].path & 1n;
            const left = {
                data: isRight ? step.data : currentData,
                value: isRight ? step.value : currentSum,
            };
            const right = {
                data: isRight ? currentData : step.data,
                value: isRight ? currentSum : step.value,
            };
            const hash = await new DataHasher(this.root.algorithm)
                .update(CborSerializer.encodeArray(CborSerializer.encodeByteString(BigintConverter.encode(step.path)), CborSerializer.encodeNullable(left.data, CborSerializer.encodeByteString), CborSerializer.encodeByteString(BigintConverter.encode(left.value)), CborSerializer.encodeNullable(right.data, CborSerializer.encodeByteString), CborSerializer.encodeByteString(BigintConverter.encode(right.value))))
                .digest();
            currentData = hash.data;
            const length = BigInt(bitLen(step.path) - 1);
            if (length < 0n) {
                return new PathVerificationResult(false, false);
            }
            currentPath = (currentPath << length) | (step.path & ((1n << length) - 1n));
            currentSum += step.value;
        }
        const pathValid = currentData != null && this.root.equals(new DataHash(this.root.algorithm, currentData));
        const pathIncluded = stateId === currentPath;
        return new PathVerificationResult(pathValid, pathIncluded);
    }
}
