// @unicitylabs/state-transition-sdk/lib/crypto/secp256k1/SigningService.js
import { secp256k1 } from '@noble/curves/secp256k1.js';
import { Signature } from './Signature.js';
/**
 * Default secp256k1 signing service. Wraps a 32-byte private key and exposes
 * the matching compressed public key, plus helpers for sign/verify and for
 * recovering a public key from a signature.
 *
 * @implements {ISigningService}
 */
export class SigningService {
    privateKey;
    _publicKey;
    constructor(privateKey) {
        this.privateKey = privateKey;
        this.privateKey = new Uint8Array(privateKey);
        this._publicKey = secp256k1.getPublicKey(this.privateKey, true);
    }
    /**
     * @returns {string} Algorithm name (`secp256k1`).
     */
    get algorithm() {
        return 'secp256k1';
    }
    /**
     * @returns {Uint8Array} Copy of the compressed public key.
     */
    get publicKey() {
        return new Uint8Array(this._publicKey);
    }
    /**
     * Generate a signing service with a fresh random private key.
     *
     * @returns {SigningService} New signing service.
     */
    static generate() {
        return new SigningService(SigningService.generatePrivateKey());
    }
    /**
     * Generate a fresh random secp256k1 private key.
     *
     * @returns {Uint8Array} 32-byte private key.
     */
    static generatePrivateKey() {
        return secp256k1.utils.randomSecretKey();
    }
    /**
     * Check whether the given bytes form a valid compressed secp256k1 public key.
     *
     * @param {Uint8Array} publicKey Compressed public key bytes.
     * @returns {boolean} True if valid.
     */
    static isPublicKeyValid(publicKey) {
        return secp256k1.utils.isValidPublicKey(publicKey, true);
    }
    /**
     * Recover the public key from the signature's recovery byte and verify the
     * signature against `hash`.
     *
     * @param {DataHash} hash Hash that was signed.
     * @param {Signature} signature Recoverable signature.
     * @returns {Promise<boolean>} True if the signature verifies.
     */
    static verifySignatureWithRecoveredPublicKey(hash, signature) {
        const publicKey = secp256k1.Signature.fromBytes(new Uint8Array([signature.recovery, ...signature.bytes]), 'recovered')
            .recoverPublicKey(hash.data)
            .toBytes();
        return SigningService.verifyWithPublicKey(hash, signature.bytes, publicKey);
    }
    /**
     * Verify secp256k1 signature against the given public key.
     *
     * @param {DataHash} hash Signed hash.
     * @param {Uint8Array} signature Compact signature bytes.
     * @param {Uint8Array} publicKey Compressed public key.
     * @returns {Promise<boolean>} True if the signature verifies.
     */
    static verifyWithPublicKey(hash, signature, publicKey) {
        return Promise.resolve(secp256k1.verify(signature, hash.data, publicKey, { format: 'compact', prehash: false }));
    }
    /**
     * Sign a hash with this service's private key.
     *
     * @param {DataHash} hash Hash to sign.
     * @returns {Promise<Signature>} Recoverable signature.
     */
    sign(hash) {
        const signature = secp256k1.sign(hash.data, this.privateKey, { format: 'recovered', prehash: false });
        return Promise.resolve(new Signature(signature.slice(1), signature[0]));
    }
    /**
     * Verify a signature against this service's public key.
     *
     * @param {DataHash} hash Signed hash.
     * @param {Signature} signature Recoverable signature.
     * @returns {Promise<boolean>} True if the signature verifies.
     */
    verify(hash, signature) {
        return SigningService.verifyWithPublicKey(hash, signature.bytes, this._publicKey);
    }
}
