// @unicitylabs/state-transition-sdk/lib/crypto/secp256k1/Signature.js
import { CborDeserializer } from '../../serialization/cbor/CborDeserializer.js';
import { CborSerializer } from '../../serialization/cbor/CborSerializer.js';
import { HexConverter } from '../../util/HexConverter.js';
/**
 * secp256k1 recoverable signature: 64 bytes of compact `(r, s)` plus a
 * single recovery byte used to recover the signer's public key.
 */
export class Signature {
    _bytes;
    recovery;
    algorithm = 'secp256k1';
    constructor(_bytes, recovery) {
        this._bytes = _bytes;
        this.recovery = recovery;
        this._bytes = new Uint8Array(_bytes);
    }
    /**
     * @returns {Uint8Array} Copy of the 64-byte signature.
     */
    get bytes() {
        return new Uint8Array(this._bytes);
    }
    /**
     * Create Signature from its 65-byte encoding.
     *
     * @param {Uint8Array} bytes 64 signature bytes followed by recovery byte.
     * @returns {Signature} Decoded signature.
     * @throws {Error} If the input is not 65 bytes long.
     */
    static decode(bytes) {
        if (bytes.length !== 65) {
            throw new Error('Signature must contain signature and recovery byte.');
        }
        return new Signature(bytes.slice(0, -1), bytes[bytes.length - 1]);
    }
    /**
     * Create Signature from CBOR bytes.
     *
     * @param {Uint8Array} bytes CBOR bytes.
     * @returns {Signature} Decoded signature.
     */
    static fromCBOR(bytes) {
        return Signature.decode(CborDeserializer.decodeByteString(bytes));
    }
    /**
     * Create Signature from JSON.
     *
     * @param {string} data Hex string of the encoded signature.
     * @returns {Signature} Decoded signature.
     */
    static fromJSON(data) {
        return Signature.decode(HexConverter.decode(data));
    }
    /**
     * @returns {Uint8Array} 65-byte concatenation of signature bytes and recovery byte.
     */
    encode() {
        return new Uint8Array([...this._bytes, this.recovery]);
    }
    /**
     * Convert Signature to CBOR bytes.
     *
     * @returns {Uint8Array} CBOR bytes.
     */
    toCBOR() {
        return CborSerializer.encodeByteString(this.encode());
    }
    /**
     * Convert Signature to JSON.
     *
     * @returns {string} Hex string of the encoded signature.
     */
    toJSON() {
        return HexConverter.encode(this.encode());
    }
    /**
     * @returns {string} Hex string of the encoded signature.
     */
    toString() {
        return `${HexConverter.encode(this.encode())}`;
    }
}
