// @unicitylabs/sphere-sdk/dist/wallet-api/index.js
// node_modules/@noble/hashes/utils.js
function isBytes(a) {
  return a instanceof Uint8Array || ArrayBuffer.isView(a) && a.constructor.name === "Uint8Array" && "BYTES_PER_ELEMENT" in a && a.BYTES_PER_ELEMENT === 1;
}
function abytes(value, length, title = "") {
  const bytes = isBytes(value);
  const len = value?.length;
  const needsLen = length !== void 0;
  if (!bytes || needsLen && len !== length) {
    const prefix = title && `"${title}" `;
    const ofLen = needsLen ? ` of length ${length}` : "";
    const got = bytes ? `length=${len}` : `type=${typeof value}`;
    const message = prefix + "expected Uint8Array" + ofLen + ", got " + got;
    if (!bytes)
      throw new TypeError(message);
    throw new RangeError(message);
  }
  return value;
}
function aexists(instance, checkFinished = true) {
  if (instance.destroyed)
    throw new Error("Hash instance has been destroyed");
  if (checkFinished && instance.finished)
    throw new Error("Hash#digest() has already been called");
}
function aoutput(out, instance) {
  abytes(out, void 0, "digestInto() output");
  const min = instance.outputLen;
  if (out.length < min) {
    throw new RangeError('"digestInto() output" expected to be of length >=' + min);
  }
}
function clean(...arrays) {
  for (let i = 0; i < arrays.length; i++) {
    arrays[i].fill(0);
  }
}
function createView(arr) {
  return new DataView(arr.buffer, arr.byteOffset, arr.byteLength);
}
function rotr(word, shift) {
  return word << 32 - shift | word >>> shift;
}
function createHasher(hashCons, info = {}) {
  const hashC = (msg, opts) => hashCons(opts).update(msg).digest();
  const tmp = hashCons(void 0);
  hashC.outputLen = tmp.outputLen;
  hashC.blockLen = tmp.blockLen;
  hashC.canXOF = tmp.canXOF;
  hashC.create = (opts) => hashCons(opts);
  Object.assign(hashC, info);
  return Object.freeze(hashC);
}
var oidNist = (suffix) => ({
  // Current NIST hashAlgs suffixes used here fit in one DER subidentifier octet.
  // Larger suffix values would need base-128 OID encoding and a different length byte.
  oid: Uint8Array.from([6, 9, 96, 134, 72, 1, 101, 3, 4, 2, suffix])
});

// node_modules/@noble/hashes/_md.js
function Chi(a, b, c) {
  return a & b ^ ~a & c;
}
function Maj(a, b, c) {
  return a & b ^ a & c ^ b & c;
}
var HashMD = class {
  blockLen;
  outputLen;
  canXOF = false;
  padOffset;
  isLE;
  // For partial updates less than block size
  buffer;
  view;
  finished = false;
  length = 0;
  pos = 0;
  destroyed = false;
  constructor(blockLen, outputLen, padOffset, isLE) {
    this.blockLen = blockLen;
    this.outputLen = outputLen;
    this.padOffset = padOffset;
    this.isLE = isLE;
    this.buffer = new Uint8Array(blockLen);
    this.view = createView(this.buffer);
  }
  update(data) {
    aexists(this);
    abytes(data);
    const { view, buffer, blockLen } = this;
    const len = data.length;
    for (let pos = 0; pos < len; ) {
      const take = Math.min(blockLen - this.pos, len - pos);
      if (take === blockLen) {
        const dataView = createView(data);
        for (; blockLen <= len - pos; pos += blockLen)
          this.process(dataView, pos);
        continue;
      }
      buffer.set(data.subarray(pos, pos + take), this.pos);
      this.pos += take;
      pos += take;
      if (this.pos === blockLen) {
        this.process(view, 0);
        this.pos = 0;
      }
    }
    this.length += data.length;
    this.roundClean();
    return this;
  }
  digestInto(out) {
    aexists(this);
    aoutput(out, this);
    this.finished = true;
    const { buffer, view, blockLen, isLE } = this;
    let { pos } = this;
    buffer[pos++] = 128;
    clean(this.buffer.subarray(pos));
    if (this.padOffset > blockLen - pos) {
      this.process(view, 0);
      pos = 0;
    }
    for (let i = pos; i < blockLen; i++)
      buffer[i] = 0;
    view.setBigUint64(blockLen - 8, BigInt(this.length * 8), isLE);
    this.process(view, 0);
    const oview = createView(out);
    const len = this.outputLen;
    if (len % 4)
      throw new Error("_sha2: outputLen must be aligned to 32bit");
    const outLen = len / 4;
    const state = this.get();
    if (outLen > state.length)
      throw new Error("_sha2: outputLen bigger than state");
    for (let i = 0; i < outLen; i++)
      oview.setUint32(4 * i, state[i], isLE);
  }
  digest() {
    const { buffer, outputLen } = this;
    this.digestInto(buffer);
    const res = buffer.slice(0, outputLen);
    this.destroy();
    return res;
  }
  _cloneInto(to) {
    to ||= new this.constructor();
    to.set(...this.get());
    const { blockLen, buffer, length, finished, destroyed, pos } = this;
    to.destroyed = destroyed;
    to.finished = finished;
    to.length = length;
    to.pos = pos;
    if (length % blockLen)
      to.buffer.set(buffer);
    return to;
  }
  clone() {
    return this._cloneInto();
  }
};
var SHA256_IV = /* @__PURE__ */ Uint32Array.from([
  1779033703,
  3144134277,
  1013904242,
  2773480762,
  1359893119,
  2600822924,
  528734635,
  1541459225
]);

// node_modules/@noble/hashes/sha2.js
var SHA256_K = /* @__PURE__ */ Uint32Array.from([
  1116352408,
  1899447441,
  3049323471,
  3921009573,
  961987163,
  1508970993,
  2453635748,
  2870763221,
  3624381080,
  310598401,
  607225278,
  1426881987,
  1925078388,
  2162078206,
  2614888103,
  3248222580,
  3835390401,
  4022224774,
  264347078,
  604807628,
  770255983,
  1249150122,
  1555081692,
  1996064986,
  2554220882,
  2821834349,
  2952996808,
  3210313671,
  3336571891,
  3584528711,
  113926993,
  338241895,
  666307205,
  773529912,
  1294757372,
  1396182291,
  1695183700,
  1986661051,
  2177026350,
  2456956037,
  2730485921,
  2820302411,
  3259730800,
  3345764771,
  3516065817,
  3600352804,
  4094571909,
  275423344,
  430227734,
  506948616,
  659060556,
  883997877,
  958139571,
  1322822218,
  1537002063,
  1747873779,
  1955562222,
  2024104815,
  2227730452,
  2361852424,
  2428436474,
  2756734187,
  3204031479,
  3329325298
]);
var SHA256_W = /* @__PURE__ */ new Uint32Array(64);
var SHA2_32B = class extends HashMD {
  constructor(outputLen) {
    super(64, outputLen, 8, false);
  }
  get() {
    const { A, B, C, D, E, F, G, H } = this;
    return [A, B, C, D, E, F, G, H];
  }
  // prettier-ignore
  set(A, B, C, D, E, F, G, H) {
    this.A = A | 0;
    this.B = B | 0;
    this.C = C | 0;
    this.D = D | 0;
    this.E = E | 0;
    this.F = F | 0;
    this.G = G | 0;
    this.H = H | 0;
  }
  process(view, offset) {
    for (let i = 0; i < 16; i++, offset += 4)
      SHA256_W[i] = view.getUint32(offset, false);
    for (let i = 16; i < 64; i++) {
      const W15 = SHA256_W[i - 15];
      const W2 = SHA256_W[i - 2];
      const s0 = rotr(W15, 7) ^ rotr(W15, 18) ^ W15 >>> 3;
      const s1 = rotr(W2, 17) ^ rotr(W2, 19) ^ W2 >>> 10;
      SHA256_W[i] = s1 + SHA256_W[i - 7] + s0 + SHA256_W[i - 16] | 0;
    }
    let { A, B, C, D, E, F, G, H } = this;
    for (let i = 0; i < 64; i++) {
      const sigma1 = rotr(E, 6) ^ rotr(E, 11) ^ rotr(E, 25);
      const T1 = H + sigma1 + Chi(E, F, G) + SHA256_K[i] + SHA256_W[i] | 0;
      const sigma0 = rotr(A, 2) ^ rotr(A, 13) ^ rotr(A, 22);
      const T2 = sigma0 + Maj(A, B, C) | 0;
      H = G;
      G = F;
      F = E;
      E = D + T1 | 0;
      D = C;
      C = B;
      B = A;
      A = T1 + T2 | 0;
    }
    A = A + this.A | 0;
    B = B + this.B | 0;
    C = C + this.C | 0;
    D = D + this.D | 0;
    E = E + this.E | 0;
    F = F + this.F | 0;
    G = G + this.G | 0;
    H = H + this.H | 0;
    this.set(A, B, C, D, E, F, G, H);
  }
  roundClean() {
    clean(SHA256_W);
  }
  destroy() {
    this.destroyed = true;
    this.set(0, 0, 0, 0, 0, 0, 0, 0);
    clean(this.buffer);
  }
};
var _SHA256 = class extends SHA2_32B {
  // We cannot use array here since array allows indexing by variable
  // which means optimizer/compiler cannot use registers.
  A = SHA256_IV[0] | 0;
  B = SHA256_IV[1] | 0;
  C = SHA256_IV[2] | 0;
  D = SHA256_IV[3] | 0;
  E = SHA256_IV[4] | 0;
  F = SHA256_IV[5] | 0;
  G = SHA256_IV[6] | 0;
  H = SHA256_IV[7] | 0;
  constructor() {
    super(32);
  }
};
var sha256 = /* @__PURE__ */ createHasher(
  () => new _SHA256(),
  /* @__PURE__ */ oidNist(1)
);

// core/crypto.ts
import * as bip39 from "bip39";
import CryptoJS from "crypto-js";
import elliptic from "elliptic";

// core/errors.ts
var SphereError = class extends Error {
  code;
  cause;
  constructor(message, code, cause) {
    super(message);
    this.name = "SphereError";
    this.code = code;
    this.cause = cause;
  }
};

// core/crypto.ts
var ec = new elliptic.ec("secp256k1");
var CURVE_ORDER = BigInt(
  "0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141"
);
var SIGN_MESSAGE_PREFIX = "Sphere Signed Message:\n";
function varint(n) {
  if (n < 253) return new Uint8Array([n]);
  const buf = new Uint8Array(3);
  buf[0] = 253;
  buf[1] = n & 255;
  buf[2] = n >> 8 & 255;
  return buf;
}
function hashSignMessage(message) {
  const prefix = new TextEncoder().encode(SIGN_MESSAGE_PREFIX);
  const msg = new TextEncoder().encode(message);
  const prefixLen = varint(prefix.length);
  const msgLen = varint(msg.length);
  const full = new Uint8Array(prefixLen.length + prefix.length + msgLen.length + msg.length);
  let off = 0;
  full.set(prefixLen, off);
  off += prefixLen.length;
  full.set(prefix, off);
  off += prefix.length;
  full.set(msgLen, off);
  off += msgLen.length;
  full.set(msg, off);
  const hex = Array.from(full).map((b) => b.toString(16).padStart(2, "0")).join("");
  const h1 = CryptoJS.SHA256(CryptoJS.enc.Hex.parse(hex)).toString();
  return CryptoJS.SHA256(CryptoJS.enc.Hex.parse(h1)).toString();
}
function signMessage(privateKeyHex, message) {
  const keyPair = ec.keyFromPrivate(privateKeyHex, "hex");
  const hashHex = hashSignMessage(message);
  const hashBytes = Buffer.from(hashHex, "hex");
  const sig = keyPair.sign(hashBytes, { canonical: true });
  const pub = keyPair.getPublic();
  let recoveryParam = -1;
  for (let i = 0; i < 4; i++) {
    try {
      if (ec.recoverPubKey(hashBytes, sig, i).eq(pub)) {
        recoveryParam = i;
        break;
      }
    } catch {
    }
  }
  if (recoveryParam === -1) {
    throw new SphereError("Could not find recovery parameter", "SIGNING_ERROR");
  }
  const v = (31 + recoveryParam).toString(16).padStart(2, "0");
  const r = sig.r.toString("hex").padStart(64, "0");
  const s = sig.s.toString("hex").padStart(64, "0");
  return v + r + s;
}

// wallet-api/errors.ts
var STATUS_TO_CODE = {
  401: "UNAUTHORIZED",
  403: "FORBIDDEN",
  404: "NOT_FOUND",
  409: "CONFLICT",
  413: "TOO_LARGE",
  422: "VALIDATION",
  429: "RATE_LIMITED"
};
var WalletApiError = class _WalletApiError extends Error {
  code;
  /** HTTP status when the error came from a response. */
  status;
  constructor(message, code, status, cause) {
    super(message);
    this.name = "WalletApiError";
    this.code = code;
    this.status = status;
    if (cause !== void 0) this.cause = cause;
  }
  static fromStatus(status, message) {
    return new _WalletApiError(message, STATUS_TO_CODE[status] ?? "SERVER", status);
  }
};
var ChallengeTemplateError = class extends WalletApiError {
  constructor(message) {
    super(message, "CHALLENGE_TEMPLATE");
    this.name = "ChallengeTemplateError";
  }
};

// wallet-api/challenge.ts
var AUTH_CHALLENGE_PREFIX = "unicity:wallet-api:auth:v1\n";
var MAX_CLOCK_SKEW_MS = 5 * 60 * 1e3;
var MAX_VALIDITY_WINDOW_MS = 60 * 60 * 1e3;
function parseFields(body) {
  let parsed;
  try {
    parsed = JSON.parse(body);
  } catch {
    throw new ChallengeTemplateError("Challenge body is not the single-line JSON object the backend emits");
  }
  if (typeof parsed !== "object" || parsed === null || Array.isArray(parsed)) {
    throw new ChallengeTemplateError("Challenge body must be a JSON object");
  }
  const fields = /* @__PURE__ */ new Map();
  for (const [key, value] of Object.entries(parsed)) {
    if (typeof value !== "string") {
      throw new ChallengeTemplateError(`Challenge field "${key}" must be a string`);
    }
    fields.set(key, value);
  }
  return fields;
}
function requireField(fields, key) {
  const value = fields.get(key);
  if (value === void 0 || value === "") {
    throw new ChallengeTemplateError(`Challenge is missing the "${key}" field`);
  }
  return value;
}
function parseTimestamp(value, key) {
  const ms = Date.parse(value);
  if (Number.isNaN(ms)) {
    throw new ChallengeTemplateError(`Challenge "${key}" is not a parseable timestamp: ${value}`);
  }
  return ms;
}
function verifyChallengeTemplate(challenge, expect) {
  if (typeof challenge !== "string" || !challenge.startsWith(AUTH_CHALLENGE_PREFIX)) {
    throw new ChallengeTemplateError("Challenge does not start with the unicity:wallet-api:auth:v1 domain prefix");
  }
  const fields = parseFields(challenge.slice(AUTH_CHALLENGE_PREFIX.length));
  const pubkey = requireField(fields, "pubkey");
  if (pubkey.toLowerCase() !== expect.pubkey.toLowerCase()) {
    throw new ChallengeTemplateError("Challenge embeds a different pubkey than this wallet");
  }
  const nonce = requireField(fields, "nonce");
  if (nonce !== expect.nonce) {
    throw new ChallengeTemplateError("Challenge nonce does not match the nonce issued with it");
  }
  const network = requireField(fields, "network");
  if (network !== expect.network) {
    throw new ChallengeTemplateError(`Challenge is for network "${network}", expected "${expect.network}"`);
  }
  const issuedAt = parseTimestamp(requireField(fields, "issuedAt"), "issuedAt");
  const expiresAt = parseTimestamp(requireField(fields, "expiresAt"), "expiresAt");
  if (issuedAt > expect.nowMs + MAX_CLOCK_SKEW_MS) {
    throw new ChallengeTemplateError("Challenge issuedAt is in the future");
  }
  if (expiresAt <= expect.nowMs - MAX_CLOCK_SKEW_MS) {
    throw new ChallengeTemplateError("Challenge is already expired");
  }
  if (expiresAt <= issuedAt || expiresAt - issuedAt > MAX_VALIDITY_WINDOW_MS) {
    throw new ChallengeTemplateError("Challenge validity window is implausible");
  }
}

// wallet-api/codec.ts
function protocolError(message) {
  return new WalletApiError(message, "PROTOCOL");
}
function asRecord(value, what) {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw protocolError(`${what} is not an object`);
  }
  return value;
}
function asArray(value, what) {
  if (!Array.isArray(value)) throw protocolError(`${what} is not an array`);
  return value;
}
function asString(value, what) {
  if (typeof value !== "string") throw protocolError(`${what} is not a string`);
  return value;
}
function parseCounter(value, what) {
  if (typeof value === "string" && /^[0-9]+$/.test(value)) return BigInt(value);
  if (typeof value === "number" && Number.isSafeInteger(value) && value >= 0) return BigInt(value);
  throw protocolError(`${what} is not a non-negative integer counter`);
}
function parseAmount(value, what) {
  if (typeof value === "string" && /^[0-9]+$/.test(value)) return BigInt(value);
  throw protocolError(`${what} is not a decimal-string amount`);
}
function parseAssets(value, what) {
  if (value === void 0 || value === null) return void 0;
  return asArray(value, what).map((raw, i) => {
    const rec = asRecord(raw, `${what}[${i}]`);
    return {
      coinId: asString(rec.coinId, `${what}[${i}].coinId`),
      amount: parseAmount(rec.amount, `${what}[${i}].amount`)
    };
  });
}
function parseInventoryItem(raw, what) {
  const rec = asRecord(raw, what);
  const status = asString(rec.status, `${what}.status`);
  if (status !== "active" && status !== "removed") {
    throw protocolError(`${what}.status is neither 'active' nor 'removed'`);
  }
  return {
    tokenId: asString(rec.tokenId, `${what}.tokenId`),
    status,
    assets: parseAssets(rec.assets, `${what}.assets`),
    seq: parseCounter(rec.seq, `${what}.seq`)
  };
}
function parseInventoryPage(body) {
  const rec = asRecord(body, "inventory response");
  if (typeof rec.more !== "boolean") throw protocolError("inventory response .more is not a boolean");
  return {
    cursor: parseCounter(rec.cursor, "inventory response .cursor"),
    syncEpoch: parseCounter(rec.syncEpoch, "inventory response .syncEpoch"),
    more: rec.more,
    items: asArray(rec.items, "inventory response .items").map(
      (raw, i) => parseInventoryItem(raw, `items[${i}]`)
    )
  };
}
function parseBalances(body) {
  const rec = asRecord(body, "balances response");
  return asArray(rec.balances, "balances response .balances").map((raw, i) => {
    const b = asRecord(raw, `balances[${i}]`);
    const tokenCount = b.tokenCount;
    if (typeof tokenCount !== "number" || !Number.isSafeInteger(tokenCount) || tokenCount < 0) {
      throw protocolError(`balances[${i}].tokenCount is not a non-negative integer`);
    }
    return {
      coinId: asString(b.coinId, `balances[${i}].coinId`),
      total: parseAmount(b.total, `balances[${i}].total`),
      tokenCount
    };
  });
}
function parseBlobUrls(body) {
  const rec = asRecord(body, "blob-urls response");
  return asArray(rec.urls, "blob-urls response .urls").map((raw, i) => {
    const u = asRecord(raw, `urls[${i}]`);
    return {
      tokenId: asString(u.tokenId, `urls[${i}].tokenId`),
      getUrl: asString(u.getUrl, `urls[${i}].getUrl`)
    };
  });
}
function parseUploadUrls(body) {
  const rec = asRecord(body, "upload-urls response");
  return asArray(rec.urls, "upload-urls response .urls").map((raw, i) => {
    const u = asRecord(raw, `urls[${i}]`);
    return {
      sha256: asString(u.sha256, `urls[${i}].sha256`),
      key: asString(u.key, `urls[${i}].key`),
      putUrl: asString(u.putUrl, `urls[${i}].putUrl`)
    };
  });
}
function parseApplyResult(body) {
  const rec = asRecord(body, "apply response");
  return parseCounter(rec.cursor, "apply response .cursor");
}
function parseIntents(body) {
  const rec = asRecord(body, "intents response");
  return asArray(rec.intents, "intents response .intents").map((raw, i) => {
    const it = asRecord(raw, `intents[${i}]`);
    const status = asString(it.status, `intents[${i}].status`);
    if (status !== "open" && status !== "completed" && status !== "aborted") {
      throw protocolError(`intents[${i}].status is not a known intent status`);
    }
    const createdAt = Date.parse(asString(it.createdAt, `intents[${i}].createdAt`));
    if (Number.isNaN(createdAt)) throw protocolError(`intents[${i}].createdAt is not a timestamp`);
    return {
      transferId: asString(it.transferId, `intents[${i}].transferId`),
      payload: asString(it.payload, `intents[${i}].payload`),
      status,
      createdAt
    };
  });
}
function parseDepositResult(body) {
  const rec = asRecord(body, "mailbox deposit response");
  return asString(rec.entryId, "mailbox deposit response .entryId");
}
function parseMailboxEntry(raw, what) {
  const rec = asRecord(raw, what);
  const status = asString(rec.status, `${what}.status`);
  if (status !== "unclaimed" && status !== "claimed" && status !== "rejected") {
    throw protocolError(`${what}.status is not a known mailbox entry status`);
  }
  const createdAt = Date.parse(asString(rec.createdAt, `${what}.createdAt`));
  if (Number.isNaN(createdAt)) throw protocolError(`${what}.createdAt is not a timestamp`);
  return {
    entryId: asString(rec.entryId, `${what}.entryId`),
    seq: parseCounter(rec.seq, `${what}.seq`),
    status,
    transferId: asString(rec.transferId, `${what}.transferId`),
    tokenId: asString(rec.tokenId, `${what}.tokenId`),
    assets: (parseAssets(rec.assets, `${what}.assets`) ?? []).map((a) => ({ ...a })),
    senderPubkey: asString(rec.senderPubkey, `${what}.senderPubkey`),
    ...rec.memo !== void 0 && rec.memo !== null ? { memo: asString(rec.memo, `${what}.memo`) } : {},
    createdAt,
    ...rec.getUrl !== void 0 && rec.getUrl !== null ? { getUrl: asString(rec.getUrl, `${what}.getUrl`) } : {},
    ...rec.blobCollected === true ? { blobCollected: true } : {}
  };
}
function parseMailboxPage(body) {
  const rec = asRecord(body, "mailbox response");
  if (typeof rec.more !== "boolean") throw protocolError("mailbox response .more is not a boolean");
  return {
    readPointer: parseCounter(rec.readPointer, "mailbox response .readPointer"),
    syncEpoch: parseCounter(rec.syncEpoch, "mailbox response .syncEpoch"),
    more: rec.more,
    entries: asArray(rec.entries, "mailbox response .entries").map(
      (raw, i) => parseMailboxEntry(raw, `entries[${i}]`)
    )
  };
}
function parseClaimResult(body) {
  const rec = asRecord(body, "mailbox claim response");
  return {
    claimed: asArray(rec.claimed, "claim response .claimed").map(
      (raw, i) => asString(raw, `claimed[${i}]`)
    ),
    alreadyClaimed: asArray(rec.alreadyClaimed, "claim response .alreadyClaimed").map((raw, i) => {
      const a = asRecord(raw, `alreadyClaimed[${i}]`);
      if (typeof a.intoInventory !== "boolean") {
        throw protocolError(`alreadyClaimed[${i}].intoInventory is not a boolean`);
      }
      return { entryId: asString(a.entryId, `alreadyClaimed[${i}].entryId`), intoInventory: a.intoInventory };
    }),
    failed: asArray(rec.failed, "claim response .failed").map((raw, i) => {
      const f = asRecord(raw, `failed[${i}]`);
      return { entryId: asString(f.entryId, `failed[${i}].entryId`), code: asString(f.code, `failed[${i}].code`) };
    })
  };
}
function parseRejectResult(body) {
  const rec = asRecord(body, "mailbox reject response");
  return asArray(rec.rejected, "reject response .rejected").map((raw, i) => asString(raw, `rejected[${i}]`));
}
function parsePaymentRequestRecord(raw, what) {
  const rec = asRecord(raw, what);
  const status = asString(rec.status, `${what}.status`);
  if (status !== "open" && status !== "paid" && status !== "declined" && status !== "expired") {
    throw protocolError(`${what}.status is not a known payment-request status`);
  }
  const createdAt = Date.parse(asString(rec.createdAt, `${what}.createdAt`));
  if (Number.isNaN(createdAt)) throw protocolError(`${what}.createdAt is not a timestamp`);
  let expiresAt;
  if (rec.expiresAt !== void 0 && rec.expiresAt !== null) {
    expiresAt = Date.parse(asString(rec.expiresAt, `${what}.expiresAt`));
    if (Number.isNaN(expiresAt)) throw protocolError(`${what}.expiresAt is not a timestamp`);
  }
  if (rec.transferId !== null && typeof rec.transferId !== "string") {
    throw protocolError(`${what}.transferId is neither a string nor null`);
  }
  return {
    id: asString(rec.id, `${what}.id`),
    seq: parseCounter(rec.seq, `${what}.seq`),
    fromPubkey: asString(rec.fromPubkey, `${what}.fromPubkey`),
    toPubkey: asString(rec.toPubkey, `${what}.toPubkey`),
    assets: (parseAssets(rec.assets, `${what}.assets`) ?? []).map((a) => ({ ...a })),
    ...rec.memo !== void 0 && rec.memo !== null ? { memo: asString(rec.memo, `${what}.memo`) } : {},
    status,
    transferId: rec.transferId,
    createdAt,
    ...expiresAt !== void 0 ? { expiresAt } : {}
  };
}
function parsePaymentRequestsPage(body, role) {
  const rec = asRecord(body, "payment-requests response");
  if (typeof rec.more !== "boolean") throw protocolError("payment-requests response .more is not a boolean");
  const requests = asArray(rec.requests, "payment-requests response .requests").map(
    (raw, i) => parsePaymentRequestRecord(raw, `requests[${i}]`)
  );
  const syncEpoch = parseCounter(rec.syncEpoch, "payment-requests response .syncEpoch");
  if (role === "incoming") {
    return {
      role,
      requests,
      more: rec.more,
      cursor: parseCounter(rec.cursor, "payment-requests response .cursor"),
      syncEpoch
    };
  }
  if (rec.cursor !== null && typeof rec.cursor !== "string") {
    throw protocolError("payment-requests response .cursor is neither a keyset string nor null");
  }
  return { role, requests, more: rec.more, cursor: rec.cursor, syncEpoch };
}
function parseHistoryRecord(raw, what) {
  const rec = asRecord(raw, what);
  return {
    dedupKey: asString(rec.dedupKey, `${what}.dedupKey`),
    id: asString(rec.id, `${what}.id`),
    type: asString(rec.type, `${what}.type`),
    ts: asString(rec.ts, `${what}.ts`),
    assets: asArray(rec.assets, `${what}.assets`).map((a, i) => {
      const asset = asRecord(a, `${what}.assets[${i}]`);
      parseAmount(asset.amount, `${what}.assets[${i}].amount`);
      return {
        coinId: asString(asset.coinId, `${what}.assets[${i}].coinId`),
        amount: asString(asset.amount, `${what}.assets[${i}].amount`)
      };
    }),
    ...typeof rec.transferId === "string" ? { transferId: rec.transferId } : {},
    ...typeof rec.tokenId === "string" ? { tokenId: rec.tokenId } : {},
    ...typeof rec.counterpartyPubkey === "string" ? { counterpartyPubkey: rec.counterpartyPubkey } : {},
    ...typeof rec.memo === "string" ? { memo: rec.memo } : {},
    ...typeof rec.counterpartyNametag === "string" ? { counterpartyNametag: rec.counterpartyNametag } : {}
  };
}
function parseHistoryPage(body) {
  const rec = asRecord(body, "history response");
  if (typeof rec.more !== "boolean") throw protocolError("history response .more is not a boolean");
  if (rec.cursor !== null && typeof rec.cursor !== "string") {
    throw protocolError("history response .cursor is not a string or null");
  }
  return {
    records: asArray(rec.records, "history response .records").map(
      (raw, i) => parseHistoryRecord(raw, `records[${i}]`)
    ),
    more: rec.more,
    cursor: rec.cursor,
    syncEpoch: parseCounter(rec.syncEpoch, "history response .syncEpoch")
  };
}
function parseAuthChallenge(body) {
  const rec = asRecord(body, "auth/challenge response");
  return {
    nonce: asString(rec.nonce, "auth/challenge response .nonce"),
    challenge: asString(rec.challenge, "auth/challenge response .challenge")
  };
}
function parseAuthTokens(body) {
  const rec = asRecord(body, "auth token response");
  return {
    jwt: asString(rec.jwt, "auth token response .jwt"),
    refreshToken: asString(rec.refreshToken, "auth token response .refreshToken")
  };
}
function parseWsTicket(body) {
  const rec = asRecord(body, "ws-ticket response");
  return asString(rec.ticket, "ws-ticket response .ticket");
}
function parseWakeFrame(raw) {
  let body;
  try {
    body = JSON.parse(raw);
  } catch {
    return null;
  }
  if (typeof body !== "object" || body === null) return null;
  const rec = body;
  if (rec.type !== "wake") return null;
  const stream = rec.stream;
  if (stream !== "inventory" && stream !== "mailbox" && stream !== "payment_requests") return null;
  try {
    return { stream, syncEpoch: parseCounter(rec.syncEpoch, "wake .syncEpoch") };
  } catch {
    return null;
  }
}

// wallet-api/wake-supervisor.ts
var CLOSE_AUTH_GONE = 4401;
var DEFAULT_WAKE_TIMING = {
  backoffBaseMs: 500,
  backoffCapMs: 3e4,
  heartbeatMs: 3e4
};
var WakeSocketSupervisor = class {
  constructor(opts, env, timing = DEFAULT_WAKE_TIMING) {
    this.opts = opts;
    this.env = env;
    this.timing = timing;
  }
  status = "connecting";
  socket = null;
  reconnectTimer = null;
  watchdogTimer = null;
  attempt = 0;
  stopped = false;
  /** A reconnect cycle is already scheduled/in-flight — coalesce concurrent drop signals. */
  reconnecting = false;
  get currentStatus() {
    return this.status;
  }
  /** Open the first socket; subsequent drops self-heal until {@link close}. */
  start() {
    void this.connect();
  }
  /** Idempotent teardown — no leaked timers or sockets, and no further reconnects. */
  close() {
    if (this.stopped) return;
    this.stopped = true;
    this.clearReconnect();
    this.clearWatchdog();
    this.detachAndClose(this.socket);
    this.socket = null;
    this.setStatus("closed");
  }
  // ── connect / drop ──────────────────────────────────────────────────────────
  async connect() {
    if (this.stopped) return;
    this.clearReconnect();
    try {
      const ws = await this.env.openSocket();
      if (this.stopped) {
        this.detachAndClose(ws);
        return;
      }
      this.adopt(ws);
    } catch (err) {
      if (this.stopped) return;
      this.onDrop(err instanceof WalletApiError && err.status === 401);
    }
  }
  /** Wire a freshly opened socket: keep onclose/onerror LIVE, arm the watchdog, run catch-up. */
  adopt(ws) {
    this.socket = ws;
    this.attempt = 0;
    this.reconnecting = false;
    ws.onmessage = (ev) => this.onFrame(ev);
    ws.onerror = () => this.onDrop(false);
    ws.onclose = (ev) => this.onClose(ev);
    if ("onPing" in ws) ws.onPing = () => this.touch();
    if ("onPong" in ws) ws.onPong = () => this.touch();
    this.armWatchdog();
    this.setStatus("connected");
    this.opts.onReconnect?.();
  }
  onFrame(ev) {
    this.touch();
    const frame = parseWakeFrame(typeof ev.data === "string" ? ev.data : String(ev.data));
    if (frame) this.opts.onWake(frame);
  }
  onClose(ev) {
    this.onDrop(ev?.code === CLOSE_AUTH_GONE);
  }
  /** A drop from any cause: schedule a backed-off reconnect (unless intentionally stopped). */
  onDrop(authGone) {
    if (this.stopped) return;
    if (authGone) this.env.onAuthGone();
    if (this.reconnecting) return;
    this.reconnecting = true;
    this.clearWatchdog();
    this.detachAndClose(this.socket);
    this.socket = null;
    this.setStatus("reconnecting");
    this.scheduleReconnect();
  }
  scheduleReconnect() {
    const delay = this.backoffDelay(this.attempt);
    this.attempt += 1;
    this.reconnectTimer = this.env.setTimeout(() => {
      this.reconnectTimer = null;
      void this.connect();
    }, delay);
  }
  /** Exponential `base·2^n` capped at `cap`, then full jitter in `[0, capped]` (thundering-herd guard). */
  backoffDelay(attempt) {
    const capped = Math.min(this.timing.backoffCapMs, this.timing.backoffBaseMs * 2 ** attempt);
    return Math.floor(this.env.random() * capped);
  }
  // ── liveness watchdog ────────────────────────────────────────────────────────
  /** Mark the socket alive — re-arm the silence watchdog. */
  touch() {
    if (this.stopped || !this.socket) return;
    this.armWatchdog();
  }
  armWatchdog() {
    this.clearWatchdog();
    this.watchdogTimer = this.env.setTimeout(() => {
      this.watchdogTimer = null;
      this.onDrop(false);
    }, this.timing.heartbeatMs * 2);
  }
  clearWatchdog() {
    if (this.watchdogTimer !== null) {
      this.env.clearTimeout(this.watchdogTimer);
      this.watchdogTimer = null;
    }
  }
  clearReconnect() {
    if (this.reconnectTimer !== null) {
      this.env.clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }
  }
  /** Detach our handlers (so a late close from THIS socket can't re-trigger) then close it. */
  detachAndClose(ws) {
    if (!ws) return;
    ws.onmessage = null;
    ws.onerror = null;
    ws.onclose = null;
    if ("onPing" in ws) ws.onPing = null;
    if ("onPong" in ws) ws.onPong = null;
    try {
      ws.close();
    } catch {
    }
  }
  setStatus(status) {
    if (this.status === status) return;
    this.status = status;
    this.opts.onStatus?.(status);
  }
};

// wallet-api/client.ts
function isLoopbackHost(hostname) {
  return hostname === "localhost" || hostname.endsWith(".localhost") || /^127\.\d{1,3}\.\d{1,3}\.\d{1,3}$/.test(hostname) || hostname === "::1" || hostname === "[::1]";
}
function assertSecureBaseUrl(baseUrl) {
  let url;
  try {
    url = new URL(baseUrl);
  } catch {
    throw new WalletApiError(`Invalid wallet-api base URL: ${baseUrl}`, "CONFIG");
  }
  if (url.protocol !== "https:" && !(url.protocol === "http:" && isLoopbackHost(url.hostname))) {
    throw new WalletApiError(
      "wallet-api base URL must be https (plain http is allowed only on loopback \u2014 ARCHITECTURE \xA74)",
      "CONFIG"
    );
  }
  return url;
}
function bytesToBase64(bytes) {
  let bin = "";
  for (const b of bytes) bin += String.fromCharCode(b);
  return btoa(bin);
}
function defaultWebSocketFactory(url) {
  const Ctor = globalThis.WebSocket;
  if (!Ctor) {
    throw new WalletApiError(
      'No global WebSocket available \u2014 inject webSocketFactory (e.g. the "ws" package on Node < 22)',
      "CONFIG"
    );
  }
  return new Ctor(url);
}
var DEFAULT_RETRY = {
  maxAttempts: 3,
  baseMs: 200,
  capMs: 1e4,
  jitter: "full",
  honorRetryAfter: true,
  maxRetryAfterMs: 6e4
};
function finiteAtLeast(value, fallback, min) {
  return typeof value === "number" && Number.isFinite(value) && value >= min ? value : fallback;
}
function resolveRetry(cfg) {
  if (cfg === false) return { ...DEFAULT_RETRY, maxAttempts: 1 };
  if (!cfg) return { ...DEFAULT_RETRY };
  return {
    maxAttempts: Math.max(1, Math.floor(finiteAtLeast(cfg.maxAttempts, DEFAULT_RETRY.maxAttempts, 1))),
    baseMs: finiteAtLeast(cfg.baseMs, DEFAULT_RETRY.baseMs, 0),
    capMs: finiteAtLeast(cfg.capMs, DEFAULT_RETRY.capMs, 0),
    jitter: cfg.jitter === "none" ? "none" : "full",
    honorRetryAfter: cfg.honorRetryAfter ?? DEFAULT_RETRY.honorRetryAfter,
    maxRetryAfterMs: finiteAtLeast(cfg.maxRetryAfterMs, DEFAULT_RETRY.maxRetryAfterMs, 0)
  };
}
var WalletApiClient = class _WalletApiClient {
  /** Network name (also the blob-key prefix `<network>/t/<sha256>` — §5.2). */
  network;
  baseUrl;
  deviceId;
  storage;
  fetchFn;
  wsFactory;
  now;
  retry;
  identity = null;
  jwt = null;
  /** Serializes concurrent re-auth attempts. */
  authInFlight = null;
  /**
   * #583: monotonic auth-session counter. Bumped by every {@link setIdentity}
   * and {@link logout}. A sign-in/refresh attempt captures it at its START and
   * bails before any shared-state mutation (`this.jwt`, the refresh token) if it
   * no longer matches — i.e. the identity moved on while the attempt was async,
   * so its result is stale and would contaminate the new owner. See
   * {@link isStale}.
   */
  authGeneration = 0;
  /** The construction config, retained so {@link clone} can mint a sibling. */
  config;
  constructor(config) {
    const url = assertSecureBaseUrl(config.baseUrl);
    this.baseUrl = url.toString().replace(/\/$/, "");
    if (!config.network) throw new WalletApiError("network is required", "CONFIG");
    this.network = config.network;
    this.deviceId = config.deviceId;
    this.storage = config.storage;
    if (!config.fetchFn && !globalThis.fetch) {
      throw new WalletApiError("No fetch available \u2014 inject fetchFn", "CONFIG");
    }
    this.fetchFn = config.fetchFn ?? ((u, init) => globalThis.fetch(u, init));
    this.wsFactory = config.webSocketFactory ?? defaultWebSocketFactory;
    this.now = config.now ?? (() => Date.now());
    this.retry = resolveRetry(config.retry);
    this.config = config;
  }
  /**
   * #583: mint a FRESH, identity-less sibling client from the SAME construction
   * config (baseUrl, network, deviceId, storage, fetch/ws factories, clock).
   * Per-address client isolation builds one client per HD address from this, so
   * each address's providers authenticate as their OWN owner with their OWN JWT
   * — an orphaned previous-address poll pump can never drive a client that has
   * been re-authed to a DIFFERENT owner. The shared `storage` is intentional:
   * its keys are namespaced per (network, chainPubkey, deviceId), so per-owner
   * refresh tokens / cursors stay separate (no cross-owner collision). The clone
   * starts with no identity and no JWT — the caller binds it via setIdentity.
   */
  clone() {
    return new _WalletApiClient(this.config);
  }
  /** Bind the wallet identity this client authenticates as. Resets the session. */
  setIdentity(identity) {
    this.identity = identity;
    this.jwt = null;
    this.authGeneration += 1;
    this.authInFlight = null;
  }
  // ── storage keys (scoped per network + pubkey + device) ────────────────────
  requireIdentity() {
    if (!this.identity) throw new WalletApiError("No identity set \u2014 call setIdentity() first", "CONFIG");
    return this.identity;
  }
  scopedKey(kind) {
    const { chainPubkey } = this.requireIdentity();
    return `wallet-api:${kind}:${this.network}:${chainPubkey}`;
  }
  refreshTokenKey() {
    return `${this.scopedKey("refresh")}:${this.deviceId}`;
  }
  /**
   * #583: true once the identity has moved on since `gen` was captured (a
   * {@link setIdentity}/{@link logout} bumped {@link authGeneration}). A sign-in
   * attempt checks this before every mutation of `this.jwt` or the refresh token
   * and bails out when stale — its tokens belong to a now-discarded owner.
   */
  isStale(gen) {
    return this.authGeneration !== gen;
  }
  // ── auth (ARCHITECTURE §4) ──────────────────────────────────────────────────
  /**
   * Establish a session: try the stored refresh token first (rotating), fall
   * back to a fresh challenge→sign→verify cycle. Safe to call repeatedly.
   */
  async signIn() {
    if (this.authInFlight) return this.authInFlight;
    const gen = this.authGeneration;
    const inFlight = this.signInInner(gen).finally(() => {
      if (this.authInFlight === inFlight) this.authInFlight = null;
    });
    this.authInFlight = inFlight;
    return inFlight;
  }
  async signInInner(gen) {
    if (await this.tryRefresh(gen)) return;
    if (this.isStale(gen)) return;
    await this.challengeSignIn(gen);
  }
  /**
   * `POST /v1/auth/refresh` with the stored token; rotates on success. Any
   * 4xx (expired, revoked, rotation-reuse revocation) clears the stored token
   * and reports `false` — the caller falls back to a challenge cycle.
   */
  async tryRefresh(gen) {
    const key = this.refreshTokenKey();
    const stored = await this.storage.get(key);
    if (!stored) return false;
    const res = await this.rawFetch("POST", "/v1/auth/refresh", { refreshToken: stored });
    if (this.isStale(gen)) return false;
    if (res.status === 200) {
      const tokens = parseAuthTokens(await this.readJson(res, "auth/refresh"));
      this.jwt = tokens.jwt;
      await this.storage.set(key, tokens.refreshToken);
      return true;
    }
    await this.storage.remove(key);
    return false;
  }
  /** The challenge→verify cycle (§4 steps 1–3) with template verification (S1). */
  async challengeSignIn(gen) {
    const identity = this.requireIdentity();
    const key = this.refreshTokenKey();
    const challengeRes = await this.rawFetch("POST", "/v1/auth/challenge", {
      pubkey: identity.chainPubkey
    });
    if (challengeRes.status !== 200) throw await this.toError(challengeRes, "auth/challenge");
    const { nonce, challenge } = parseAuthChallenge(await this.readJson(challengeRes, "auth/challenge"));
    verifyChallengeTemplate(challenge, {
      pubkey: identity.chainPubkey,
      nonce,
      network: this.network,
      nowMs: this.now()
    });
    const signature = signMessage(identity.privateKey, challenge);
    const verifyRes = await this.rawFetch("POST", "/v1/auth/verify", {
      nonce,
      signature,
      deviceId: this.deviceId
    });
    if (verifyRes.status !== 200) throw await this.toError(verifyRes, "auth/verify");
    const tokens = parseAuthTokens(await this.readJson(verifyRes, "auth/verify"));
    if (this.isStale(gen)) return;
    this.jwt = tokens.jwt;
    await this.storage.set(key, tokens.refreshToken);
  }
  /** Revoke the session server-side and drop all local credentials. */
  async logout() {
    this.authGeneration += 1;
    this.authInFlight = null;
    if (this.jwt) {
      try {
        await this.rawFetch("POST", "/v1/auth/logout", {}, this.jwt);
      } finally {
        this.jwt = null;
      }
    }
    await this.storage.remove(this.refreshTokenKey());
  }
  // ── request core ────────────────────────────────────────────────────────────
  async rawFetch(method, path, body, jwt) {
    const headers = {};
    if (body !== void 0) headers["content-type"] = "application/json";
    if (jwt) headers["authorization"] = `Bearer ${jwt}`;
    try {
      return await this.fetchFn(`${this.baseUrl}${path}`, {
        method,
        headers,
        body: body !== void 0 ? JSON.stringify(body) : void 0
      });
    } catch (err) {
      throw new WalletApiError(`wallet-api request failed: ${method} ${path}`, "NETWORK", void 0, err);
    }
  }
  async readJson(res, what) {
    const text = await res.text();
    if (text === "") return null;
    try {
      return JSON.parse(text);
    } catch (err) {
      throw new WalletApiError(`${what}: response is not JSON`, "PROTOCOL", res.status, err);
    }
  }
  async toError(res, what) {
    let message = `${what} failed with status ${res.status}`;
    try {
      const body = await this.readJson(res, what);
      if (body?.error?.message) message = `${what}: ${body.error.code ?? ""} ${body.error.message}`.trim();
    } catch {
    }
    return WalletApiError.fromStatus(res.status, message);
  }
  /**
   * Retry transient REST failures with bounded exponential backoff + jitter:
   * - a thrown transient `NETWORK` failure (dropped/reset connection, DNS blip)
   *   retries on **idempotent GETs only** — a lost-response retry of a write
   *   could double-apply;
   * - a `429` response retries on **any** method (a 429 is rejected before the
   *   handler runs — the write never executed), honoring the server `Retry-After`;
   * - a `503` response retries on **GETs only** (a write may have started).
   * Every other outcome (2xx, or a non-retryable status) is returned to the
   * caller, which maps a non-2xx to a typed {@link WalletApiError}.
   */
  async rawFetchWithRetry(method, path, body, jwt) {
    const idempotent = method === "GET";
    for (let attempt = 1; ; attempt += 1) {
      let res;
      try {
        res = await this.rawFetch(method, path, body, jwt);
      } catch (err) {
        const transient = err instanceof WalletApiError && err.code === "NETWORK";
        if (!transient || !idempotent || attempt >= this.retry.maxAttempts) throw err;
        await this.sleep(this.backoffMs(attempt));
        continue;
      }
      const waitMs = this.retryDelayForStatus(res, idempotent, attempt);
      if (waitMs === null || attempt >= this.retry.maxAttempts) return res;
      await this.sleep(waitMs);
    }
  }
  /** Bounded full-jitter exponential backoff (mirrors the wake supervisor). */
  backoffMs(attempt) {
    const capped = Math.min(this.retry.capMs, this.retry.baseMs * 2 ** (attempt - 1));
    return this.retry.jitter === "none" ? capped : Math.floor(Math.random() * capped);
  }
  /**
   * Delay before retrying a non-2xx response, or `null` to not retry. `429`
   * retries on any method (rejected before execution); `503` on idempotent GETs.
   * Honors a capped `Retry-After` when present, else falls back to backoff.
   */
  retryDelayForStatus(res, idempotent, attempt) {
    const retryable = res.status === 429 || res.status === 503 && idempotent;
    if (!retryable) return null;
    const hinted = this.retry.honorRetryAfter ? this.parseRetryAfter(res) : null;
    return hinted === null ? this.backoffMs(attempt) : Math.min(hinted, this.retry.maxRetryAfterMs);
  }
  /** Parse `Retry-After` (delta-seconds or an HTTP-date) to ms, or `null` if absent/unparseable. */
  parseRetryAfter(res) {
    const raw = res.headers?.get("retry-after");
    if (!raw) return null;
    const seconds = Number(raw);
    if (Number.isFinite(seconds)) return Math.max(0, seconds * 1e3);
    const when = Date.parse(raw);
    return Number.isNaN(when) ? null : Math.max(0, when - this.now());
  }
  sleep(ms) {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
  /**
   * Authenticated JSON request. A 401 triggers one silent re-auth
   * (refresh → challenge fallback) and one retry; {@link rawFetchWithRetry}
   * additionally rides out transient failures — a dropped connection on an
   * idempotent GET, or a `429`/`503` with `Retry-After` — so a single blip or a
   * brief rate-limit window doesn't fail the call.
   */
  async requestJson(method, path, body) {
    if (!this.jwt) await this.signIn();
    let res = await this.rawFetchWithRetry(method, path, body, this.jwt);
    if (res.status === 401) {
      this.jwt = null;
      await this.signIn();
      res = await this.rawFetchWithRetry(method, path, body, this.jwt);
    }
    if (res.status === 204) return null;
    if (!res.ok) throw await this.toError(res, `${method} ${path}`);
    return this.readJson(res, `${method} ${path}`);
  }
  // ── inventory / blobs (§16) ─────────────────────────────────────────────────
  /** `GET /v1/inventory?since=` — one page; the caller loops while `more`. */
  async listInventory(since) {
    const query = since !== void 0 ? `?since=${since.toString()}` : "";
    const page = parseInventoryPage(await this.requestJson("GET", `/v1/inventory${query}`));
    await this.noteSyncEpoch(page.syncEpoch);
    return page;
  }
  /** `GET /v1/balances` — active rows only. */
  async getBalances() {
    return parseBalances(await this.requestJson("GET", "/v1/balances"));
  }
  /** `POST /v1/tokens/blob-urls` — owner's active *or tombstoned* rows (§5.3 recovery). */
  async getBlobUrls(tokenIds) {
    return parseBlobUrls(await this.requestJson("POST", "/v1/tokens/blob-urls", { tokenIds }));
  }
  /** `POST /v1/tokens/upload-urls` — checksum/length-bound presigned PUTs (§5.2). */
  async getUploadUrls(blobs) {
    return parseUploadUrls(await this.requestJson("POST", "/v1/tokens/upload-urls", { blobs }));
  }
  /**
   * `POST /v1/inventory/apply` (§5.3). Also marks the local intent copy
   * completed — the server completes the intent in the same transaction (§16).
   */
  async applyInventoryDelta(req) {
    const cursor = parseApplyResult(
      await this.requestJson("POST", "/v1/inventory/apply", {
        transferId: req.transferId,
        spent: req.spent,
        added: req.added,
        ...req.externalDelivery !== void 0 ? { externalDelivery: req.externalDelivery } : {}
      })
    );
    await this.setLocalIntentStatus(req.transferId, "completed");
    return cursor;
  }
  /** Download blob bytes from a signed GET URL (the URL itself is the credential). */
  async fetchBlob(getUrl) {
    let res;
    try {
      res = await this.fetchFn(getUrl, { method: "GET" });
    } catch (err) {
      throw new WalletApiError("blob download failed", "NETWORK", void 0, err);
    }
    if (!res.ok) throw WalletApiError.fromStatus(res.status, `blob download failed with status ${res.status}`);
    return new Uint8Array(await res.arrayBuffer());
  }
  /**
   * Upload blob bytes to a signed PUT URL. A `412 Precondition Failed` means
   * the identical blob already exists (content addressing, `If-None-Match: *`
   * — §5.2) and is treated as success.
   *
   * The §5.2 presign binds `x-amz-checksum-sha256` and `if-none-match` (plus
   * `content-length`) as SIGNED HEADERS — a real S3 endpoint rejects the
   * SigV4 signature unless the uploader sends them verbatim. (Caught by the
   * phase-2 harness on first contact with real MinIO — the in-process fake
   * had only validated the body, not the signed headers.)
   */
  async uploadBlob(putUrl, bytes) {
    let res;
    try {
      res = await this.fetchFn(putUrl, {
        method: "PUT",
        headers: {
          "content-type": "application/octet-stream",
          "x-amz-checksum-sha256": bytesToBase64(sha256(bytes)),
          "if-none-match": "*"
        },
        body: bytes
      });
    } catch (err) {
      throw new WalletApiError("blob upload failed", "NETWORK", void 0, err);
    }
    if (res.ok || res.status === 412) return;
    throw WalletApiError.fromStatus(res.status, `blob upload failed with status ${res.status}`);
  }
  // ── mailbox (§6/§16) ────────────────────────────────────────────────────────
  /**
   * `POST /v1/mailbox` — deposit an already-uploaded finished blob to the
   * recipient's mailbox. Idempotent by content-derived `entry_id` (§6): an
   * existing entry in any status returns `200` with its id, provided the
   * request's recipient and key match the stored entry (a mismatch is `409`).
   */
  async depositMailbox(req) {
    return parseDepositResult(
      await this.requestJson("POST", "/v1/mailbox", {
        recipientPubkey: req.recipientPubkey,
        key: req.key,
        transferId: req.transferId,
        stateHash: req.stateHash,
        tokenId: req.tokenId,
        ...req.memo !== void 0 ? { memo: req.memo } : {}
      })
    );
  }
  /**
   * `GET /v1/mailbox?since=<seq>` — entries of EVERY status are listable for
   * any client-chosen `since`; a claimed entry carries a working `getUrl`
   * while its blob is within retention, `blobCollected: true` afterwards (§6).
   */
  async listMailbox(since) {
    const query = since !== void 0 ? `?since=${since.toString()}` : "";
    const page = parseMailboxPage(await this.requestJson("GET", `/v1/mailbox${query}`));
    await this.noteSyncEpoch(page.syncEpoch);
    return page;
  }
  /**
   * `POST /v1/mailbox/claim` — addressee-only, idempotent ownership handoff
   * (§6). `intoInventory:false` is the delivery-only claim: the entry resolves
   * and the pointer advances with ZERO inventory writes.
   */
  async claimMailbox(entryIds, intoInventory) {
    return parseClaimResult(
      await this.requestJson("POST", "/v1/mailbox/claim", { entryIds, intoInventory })
    );
  }
  /**
   * `POST /v1/mailbox/reject` — addressee-only; terminal for DISCOVERY only:
   * the entry counts toward read-pointer contiguity but remains claimable and
   * its blob is retained (§6 — reject is never a destruction path).
   */
  async rejectMailbox(entryIds) {
    return parseRejectResult(await this.requestJson("POST", "/v1/mailbox/reject", { entryIds }));
  }
  // ── history (§10/§16) ───────────────────────────────────────────────────────
  /**
   * `POST /v1/history` — client-asserted records, deduped by `dedupKey` (the
   * server never writes history rows — §10). `memo`/`counterpartyNametag`
   * MUST already be S6 envelopes.
   */
  async postHistoryRecords(records) {
    await this.requestJson("POST", "/v1/history", { records });
  }
  /** `GET /v1/history?before=&limit=` — newest-first keyset pages (§10). */
  async listHistory(options = {}) {
    const params = new URLSearchParams();
    if (options.before !== void 0) params.set("before", options.before);
    if (options.limit !== void 0) params.set("limit", String(options.limit));
    const qs = params.toString();
    return parseHistoryPage(await this.requestJson("GET", `/v1/history${qs ? `?${qs}` : ""}`));
  }
  // ── payment requests (§10/§16) ──────────────────────────────────────────────
  /**
   * `POST /v1/payment-requests` — create a request addressed to a payer (the
   * payer's account is auto-provisioned even if they never authenticated —
   * §4/§9; the per-payer open cap → 429, §5.5). `memo` MUST already be an S6
   * `enc1.` envelope (§8.3) — only the requester's wallet key decrypts it.
   */
  async createPaymentRequest(input) {
    return parsePaymentRequestRecord(
      await this.requestJson("POST", "/v1/payment-requests", {
        toPubkey: input.toPubkey,
        // Amounts are decimal strings in every JSON body (§11).
        assets: input.assets.map((a) => ({ coinId: a.coinId, amount: a.amount.toString() })),
        ...input.memo !== void 0 ? { memo: input.memo } : {},
        ...input.expiresAt !== void 0 ? { expiresAt: new Date(input.expiresAt).toISOString() } : {}
      }),
      "payment-request create response"
    );
  }
  /**
   * `GET /v1/payment-requests?role=…` (§16): `'incoming'` is the payer's
   * gap-free `?since=<seq>` stream (cursor = bigint, the standard §16 since
   * contract); `'outgoing'` is the requester's newest-first
   * `?before=<opaque keyset>` backfill (cursor = string | null). The two
   * cursor families never mix — the server 422s a mismatched parameter, and
   * the parameter types make it unrepresentable here.
   */
  async listPaymentRequests(params) {
    const p = params;
    const query = new URLSearchParams();
    query.set("role", p.role);
    if (p.status !== void 0) query.set("status", p.status);
    if (p.role === "incoming") {
      if (p.since !== void 0) query.set("since", p.since.toString());
    } else if (p.before !== void 0) {
      query.set("before", p.before);
    }
    const page = parsePaymentRequestsPage(
      await this.requestJson("GET", `/v1/payment-requests?${query.toString()}`),
      p.role
    );
    await this.noteSyncEpoch(page.syncEpoch);
    return page;
  }
  /**
   * `POST /v1/payment-requests/{id}/respond` — addressee-only (§10, non-
   * addressee → 403); only an `open` request may be responded to (else 409).
   * `paid` REQUIRES and links the fulfilling send's transferId, `declined`
   * carries none — the pairing is enforced by {@link RespondPaymentRequestInput}.
   */
  async respondPaymentRequest(id, response) {
    return parsePaymentRequestRecord(
      await this.requestJson("POST", `/v1/payment-requests/${id}/respond`, response),
      "payment-request respond response"
    );
  }
  // ── intents (E.3) ───────────────────────────────────────────────────────────
  intentsKey() {
    return this.scopedKey("intents");
  }
  async readLocalIntents() {
    const raw = await this.storage.get(this.intentsKey());
    if (!raw) return {};
    try {
      return JSON.parse(raw);
    } catch {
      return {};
    }
  }
  async writeLocalIntents(intents) {
    await this.storage.set(this.intentsKey(), JSON.stringify(intents));
  }
  async setLocalIntentStatus(transferId, status, opts = {}) {
    const intents = await this.readLocalIntents();
    const intent = intents[transferId];
    if (!intent) return;
    if (intent.status === "completed") return;
    const abortPending = opts.abortPending === true;
    if (intent.status === status && intent.abortPending === true === abortPending) return;
    intent.status = status;
    if (abortPending) intent.abortPending = true;
    else delete intent.abortPending;
    await this.writeLocalIntents(intents);
  }
  /**
   * Persist a (client-encrypted — S6) intent payload: LOCAL copy first, then
   * `PUT /v1/intents/{transferId}` and await the server ack (E.3 — the engine
   * MUST NOT be called before this resolves). A failed server PUT throws, but
   * the local copy stays as the restore backstop and is re-PUT by
   * {@link resyncOpenIntents}.
   */
  async putIntent(transferId, payloadEnvelope) {
    const intents = await this.readLocalIntents();
    if (!intents[transferId]) {
      intents[transferId] = { payload: payloadEnvelope, status: "open", createdAt: this.now() };
      await this.writeLocalIntents(intents);
    }
    await this.requestJson("PUT", `/v1/intents/${transferId}`, { payload: payloadEnvelope });
  }
  /** `GET /v1/intents?status=` — server-side intent list. */
  async listIntents(status) {
    return parseIntents(await this.requestJson("GET", `/v1/intents?status=${status}`));
  }
  /**
   * `POST /v1/intents/{id}/abort` — soft, recoverable (§16).
   *
   * #516: the LOCAL copy flips to 'aborted' even when the server abort cannot
   * land (dead backend) — leaving it 'open' would make {@link resyncOpenIntents}
   * re-PUT it and the next sign-in's `resumeOpenIntents` RE-EXECUTE a send the
   * user watched fail (and possibly retried manually): a double-pay hazard.
   * An unlanded abort is marked pending and replayed by
   * {@link resyncOpenIntents} (PUT + abort) so the server converges too.
   */
  async abortIntent(transferId) {
    try {
      await this.requestJson("POST", `/v1/intents/${transferId}/abort`);
    } catch (err) {
      await this.setLocalIntentStatus(transferId, "aborted", { abortPending: true });
      throw err;
    }
    await this.setLocalIntentStatus(transferId, "aborted");
  }
  /** `POST /v1/intents/{id}/complete` — the uniform client-side close (E.3). */
  async completeIntent(transferId) {
    await this.requestJson("POST", `/v1/intents/${transferId}/complete`);
    await this.setLocalIntentStatus(transferId, "completed");
  }
  /** The locally-known open intents (the E.3 restore backstop). */
  async listLocalOpenIntents() {
    const intents = await this.readLocalIntents();
    return Object.entries(intents).filter(([, i]) => i.status === "open").map(([transferId, i]) => ({ transferId, payload: i.payload, status: "open", createdAt: i.createdAt }));
  }
  /**
   * Re-PUT every locally-known open intent (idempotent — the server PUT is
   * write-once while open/completed). Called after a `syncEpoch` change
   * (server restore — §5.4): intents are the one server table not
   * re-derivable from blobs.
   *
   * #516: also replays locally-aborted intents whose abort never landed on
   * the server (`abortPending`) — PUT (no-op while open/completed; (re)creates
   * the row after a restore or a failed original PUT) then abort, so the
   * server row lands as 'aborted' instead of an 'open' intent that resume
   * would re-execute. Completed-wins is preserved server-side (§16).
   */
  async resyncOpenIntents() {
    const intents = await this.readLocalIntents();
    for (const [transferId, intent] of Object.entries(intents)) {
      if (intent.status === "open") {
        await this.requestJson("PUT", `/v1/intents/${transferId}`, { payload: intent.payload });
      } else if (intent.status === "aborted" && intent.abortPending === true) {
        await this.requestJson("PUT", `/v1/intents/${transferId}`, { payload: intent.payload });
        await this.requestJson("POST", `/v1/intents/${transferId}/abort`);
        await this.setLocalIntentStatus(transferId, "aborted");
      }
    }
  }
  // ── syncEpoch (§5.4/§9) ─────────────────────────────────────────────────────
  syncEpochKey() {
    return this.scopedKey("syncEpoch");
  }
  /**
   * Track the server `syncEpoch` carried by cursor-bearing responses and
   * wakes. On a change (server restore), re-PUT locally-known open intents
   * before anything resumes (E.3). Storage-provider cursor invalidation is
   * the provider's own duty (S2) — it observes the same epoch values.
   */
  async noteSyncEpoch(epoch) {
    const stored = await this.storage.get(this.syncEpochKey());
    if (stored === epoch.toString()) return;
    await this.storage.set(this.syncEpochKey(), epoch.toString());
    if (stored !== null) {
      await this.resyncOpenIntents();
    }
  }
  // ── realtime (§9) ───────────────────────────────────────────────────────────
  /**
   * Open the wake channel: `POST /v1/ws-ticket` (JWT-authed, single-use,
   * short TTL) then `GET /v1/ws?ticket=…` — the JWT never appears in a URL.
   * Wakes are nudges only; correctness comes from the `?since=` cursors.
   */
  async connectWakeSocket(onWake) {
    const ws = await this.openWakeSocket();
    ws.onerror = null;
    ws.onclose = null;
    ws.onmessage = (ev) => {
      const frame = parseWakeFrame(typeof ev.data === "string" ? ev.data : String(ev.data));
      if (!frame) return;
      void this.noteSyncEpoch(frame.syncEpoch);
      onWake(frame);
    };
    return { close: () => ws.close() };
  }
  /**
   * Mint a single-use ticket (a 401 silently re-auths via {@link requestJson}),
   * upgrade to a wake socket and resolve once it opens. ONE socket, no
   * auto-reconnect — the building block reused by {@link connectWakeSocket} and
   * the {@link superviseWakeSocket} supervisor.
   */
  async openWakeSocket() {
    const ticket = parseWsTicket(await this.requestJson("POST", "/v1/ws-ticket"));
    const wsBase = this.baseUrl.replace(/^http/, "ws");
    const ws = this.wsFactory(`${wsBase}/v1/ws?ticket=${encodeURIComponent(ticket)}`);
    await new Promise((resolve, reject) => {
      ws.onopen = () => resolve();
      ws.onerror = (err) => reject(new WalletApiError("wake socket failed to connect", "NETWORK", void 0, err));
      ws.onclose = () => reject(new WalletApiError("wake socket closed before connecting", "NETWORK"));
    });
    return ws;
  }
  /**
   * Open a SELF-HEALING wake channel (§9): the supervisor keeps the socket
   * alive across drops — bounded backoff + jitter reconnects (fresh ticket each
   * attempt; a 4401/auth-gone close also refreshes the token), a client-side
   * liveness watchdog that force-reconnects a half-open socket, and a full
   * catch-up pull of every stream on each (re)connect (`onReconnect`) so a dead
   * window converges WITHOUT waiting for the poll. `onStatus` surfaces true
   * socket liveness, decoupled from sign-in state. `close()` tears it down with
   * no further reconnects (a deliberate logout-close never loops).
   */
  superviseWakeSocket(options, timing = DEFAULT_WAKE_TIMING) {
    const supervisor = new WakeSocketSupervisor(
      {
        onWake: (wake) => {
          void this.noteSyncEpoch(wake.syncEpoch);
          options.onWake(wake);
        },
        onReconnect: options.onReconnect,
        onStatus: options.onStatus
      },
      {
        openSocket: () => this.openWakeSocket(),
        // A 4401/expiry close means the bound token is gone — drop the cached
        // JWT so the next ticket mint re-auths (refresh → challenge fallback).
        onAuthGone: () => {
          this.jwt = null;
        },
        setTimeout: (cb, ms) => setTimeout(cb, ms),
        clearTimeout: (h) => clearTimeout(h),
        random: () => Math.random()
      },
      timing
    );
    supervisor.start();
    return {
      close: () => supervisor.close(),
      get status() {
        return supervisor.currentStatus;
      }
    };
  }
};
export {
  AUTH_CHALLENGE_PREFIX,
  ChallengeTemplateError,
  WalletApiClient,
  WalletApiError,
  verifyChallengeTemplate
};
