// @unicitylabs/state-transition-sdk/lib/smt/radix/FinalizedNodeBranch.js
import { dedent } from '../../util/StringUtils.js';
/**
 * Finalized interior node in a radix sparse Merkle tree.
 */
export class FinalizedNodeBranch {
    path;
    depth;
    left;
    right;
    hash;
    constructor(path, depth, left, right, hash) {
        this.path = path;
        this.depth = depth;
        this.left = left;
        this.right = right;
        this.hash = hash;
    }
    /**
     * Hash a {@link PendingNodeBranch} into a finalized node.
     *
     * @param {IDataHasherFactory<IDataHasher>} factory Hasher factory.
     * @param {PendingNodeBranch} node Pending node to finalize.
     * @returns {Promise<FinalizedNodeBranch>} Finalized node branch.
     */
    static async create(factory, node) {
        const [left, right] = await Promise.all([node.left.finalize(factory), node.right.finalize(factory)]);
        const hash = await factory
            .create()
            .update(new Uint8Array([0x01, node.depth]))
            .update(left.hash.data)
            .update(right.hash.data)
            .digest();
        return new FinalizedNodeBranch(node.path, node.depth, left, right, hash);
    }
    /**
     * @returns {Promise<FinalizedNodeBranch>} This branch (already finalized).
     */
    finalize() {
        return Promise.resolve(this);
    }
    /**
     * @returns {string} String representation of the node.
     */
    toString() {
        return dedent `
      Node[${this.path.toString(2)}]
        Hash: ${this.hash.toString()}
        Depth: ${this.depth}
        Left: ${this.left.toString()}
        Right: ${this.right.toString()}`;
    }
}
