// @unicitylabs/state-transition-sdk/lib/smt/radix/FinalizedLeafBranch.js
import { HexConverter } from '../../util/HexConverter.js';
import { dedent } from '../../util/StringUtils.js';
/**
 * Finalized leaf in a radix sparse Merkle tree.
 */
export class FinalizedLeafBranch {
    path;
    _key;
    _data;
    hash;
    constructor(path, _key, _data, hash) {
        this.path = path;
        this._key = _key;
        this._data = _data;
        this.hash = hash;
    }
    /**
     * @returns {Uint8Array} Copy of the leaf data bytes.
     */
    get data() {
        return this._data.slice();
    }
    /**
     * @returns {Uint8Array} Copy of the leaf key bytes.
     */
    get key() {
        return this._key.slice();
    }
    /**
     * Hash a {@link PendingLeafBranch} into a finalized leaf.
     *
     * @param {IDataHasherFactory<IDataHasher>} factory Hasher factory.
     * @param {PendingLeafBranch} leaf Pending leaf to finalize.
     * @returns {Promise<FinalizedLeafBranch>} Finalized leaf branch.
     */
    static async fromPendingLeaf(factory, leaf) {
        const key = leaf.key;
        const data = leaf.data;
        const hash = await factory
            .create()
            .update(new Uint8Array([0x00]))
            .update(key)
            .update(data)
            .digest();
        return new FinalizedLeafBranch(leaf.path, key, data, hash);
    }
    /**
     * @returns {Promise<FinalizedLeafBranch>} This branch (already finalized).
     */
    finalize() {
        return Promise.resolve(this);
    }
    /**
     * @returns {string} String representation of the leaf.
     */
    toString() {
        return dedent `
      FinalizedLeaf[${this.path.toString(2)}]
        Key: ${HexConverter.encode(this._key)}
        Data: ${HexConverter.encode(this._data)}
        Hash: ${this.hash.toString()}`;
    }
}
