// @unicitylabs/state-transition-sdk/lib/smt/radix/PendingLeafBranch.js
import { FinalizedLeafBranch } from './FinalizedLeafBranch.js';
import { HexConverter } from '../../util/HexConverter.js';
import { dedent } from '../../util/StringUtils.js';
/**
 * Pending leaf in a radix sparse Merkle tree, awaiting hashing.
 */
export class PendingLeafBranch {
    path;
    _key;
    _data;
    constructor(path, _key, _data) {
        this.path = path;
        this._key = _key;
        this._data = _data;
        this._key = new Uint8Array(_key);
        this._data = new Uint8Array(_data);
    }
    /**
     * @returns {Uint8Array} Copy of the leaf data bytes.
     */
    get data() {
        return this._data.slice();
    }
    /**
     * @returns {Uint8Array} Copy of the leaf key bytes.
     */
    get key() {
        return this._key.slice();
    }
    /**
     * Hash this leaf to produce a finalized branch.
     *
     * @param {IDataHasherFactory<IDataHasher>} factory Hasher factory.
     * @returns {Promise<FinalizedLeafBranch>} Finalized leaf branch.
     */
    finalize(factory) {
        return FinalizedLeafBranch.fromPendingLeaf(factory, this);
    }
    /**
     * @returns {string} String representation of the leaf.
     */
    toString() {
        return dedent `
      PendingLeaf[${this.path.toString(2)}]
        Key: ${HexConverter.encode(this._key)}
        Data: ${HexConverter.encode(this._data)}`;
    }
}
