// @unicitylabs/state-transition-sdk/lib/smt/radix/SparseMerkleTreeRootNode.js
import { FinalizedLeafBranch } from './FinalizedLeafBranch.js';
import { DataHash } from '../../crypto/hash/DataHash.js';
import { HashAlgorithm } from '../../crypto/hash/HashAlgorithm.js';
import { BitString } from '../../util/BitString.js';
import { dedent } from '../../util/StringUtils.js';
import { areUint8ArraysEqual } from '../../util/TypedArrayUtils.js';
/**
 * Sparse Merkle Tree root node implementation.
 */
export class SparseMerkleTreeRootNode {
    left;
    right;
    hash;
    depth = 0;
    path = 1n;
    constructor(left, right, hash) {
        this.left = left;
        this.right = right;
        this.hash = hash;
    }
    /**
     * Creates a new instance of MerkleTreeRootNode.
     * @param left Root node left branch.
     * @param right Root node right branch.
     * @param factory Factory to create data hashers.
     * @return A promise that resolves to a new MerkleTreeRootNode instance.
     */
    static async create(left, right, factory) {
        if (left != null && right == null) {
            return new SparseMerkleTreeRootNode(left, null, left.hash);
        }
        if (left == null && right != null) {
            return new SparseMerkleTreeRootNode(null, right, right.hash);
        }
        if (left != null && right != null) {
            return new SparseMerkleTreeRootNode(left, right, await factory
                .create()
                .update(new Uint8Array([0x01, 0x00]))
                .update(left.hash.data)
                .update(right.hash.data)
                .digest());
        }
        return new SparseMerkleTreeRootNode(null, null, new DataHash(HashAlgorithm.SHA256, new Uint8Array(32)));
    }
    /**
     * Check whether a leaf with the given key is in the tree.
     *
     * @param {Uint8Array} key Leaf key.
     * @returns {boolean} True if the key is present.
     */
    has(key) {
        const keyPath = BitString.fromBytesReversedLSB(key).toBigInt();
        // eslint-disable-next-line @typescript-eslint/no-this-alias
        let node = this;
        while (node != null) {
            if (node instanceof FinalizedLeafBranch) {
                return areUint8ArraysEqual(node.key, key);
            }
            const isRight = (keyPath >> BigInt(node.depth)) & 1n;
            node = isRight ? node.right : node.left;
        }
        return false;
    }
    /**
     * Returns a string representation of the MerkleTreeRootNode.
     */
    toString() {
        return dedent `
      Root: 
        Hash: ${this.hash.toString()}
      Left: 
        ${this.left ? this.left.toString() : 'null'}
      Right: 
        ${this.right ? this.right.toString() : 'null'}`;
    }
}
