// @unicitylabs/state-transition-sdk/lib/api/CertificationResponse.js
import { InvalidJsonStructureError } from '../InvalidJsonStructureError.js';
/**
 * Possible results from the aggregator when submitting a certification request.
 */
export var CertificationStatus;
(function (CertificationStatus) {
    /** The certification request was accepted and stored. */
    CertificationStatus["SUCCESS"] = "SUCCESS";
    /** State identifier did not match the payload. */
    CertificationStatus["STATE_ID_MISMATCH"] = "STATE_ID_MISMATCH";
    /** Signature verification failed. */
    CertificationStatus["SIGNATURE_VERIFICATION_FAILED"] = "SIGNATURE_VERIFICATION_FAILED";
    /** The signature format is invalid. */
    CertificationStatus["INVALID_SIGNATURE_FORMAT"] = "INVALID_SIGNATURE_FORMAT";
    /** The public key format is invalid. */
    CertificationStatus["INVALID_PUBLIC_KEY_FORMAT"] = "INVALID_PUBLIC_KEY_FORMAT";
    /** The source state hash format is invalid. */
    CertificationStatus["INVALID_SOURCE_STATE_HASH_FORMAT"] = "INVALID_SOURCE_STATE_HASH_FORMAT";
    /** The transaction hash format is invalid. */
    CertificationStatus["INVALID_TRANSACTION_HASH_FORMAT"] = "INVALID_TRANSACTION_HASH_FORMAT";
    /** The specified algorithm is not supported. */
    CertificationStatus["UNSUPPORTED_ALGORITHM"] = "UNSUPPORTED_ALGORITHM";
    /** The certification request was submitted to an invalid shard. */
    CertificationStatus["INVALID_SHARD"] = "INVALID_SHARD";
})(CertificationStatus || (CertificationStatus = {}));
/**
 * Response object returned by the aggregator on certification request.
 *
 * The status is deliberately tolerant: the aggregator may emit status strings unknown to this SDK
 * version (older deployments, or statuses added later). An unknown status parses successfully and
 * must be treated by callers as "the request was not accepted" — the caller then probes
 * `getInclusionProof` to learn the state's actual certification, instead of failing on parse.
 */
export class CertificationResponse {
    status;
    /** The aggregator's status string. Known values are enumerated in {@link CertificationStatus};
     * compare against those members. Unknown strings are preserved as-is. */
    constructor(status) {
        this.status = status;
    }
    /**
     * Create a new certification response.
     * @param {string} status Certification response status (unknown strings allowed — see class doc)
     *
     * @returns {CertificationResponse} Created certification response
     */
    static create(status) {
        return new CertificationResponse(status);
    }
    /**
     * Parse a JSON response object.
     *
     * @param {unknown} data Raw response
     * @returns {Promise<>CertificationResponse>} Parsed response
     * @throws {InvalidJsonStructureError} Error if the data does not match the expected shape
     */
    static fromJSON(data) {
        if (!CertificationResponse.isJSON(data)) {
            throw new InvalidJsonStructureError();
        }
        return new CertificationResponse(data.status);
    }
    /**
     * Check if the given data is a valid JSON response object.
     *
     * @param {unknown} data Raw response
     * @returns {boolean} True if the data is a valid JSON response object
     */
    static isJSON(data) {
        return (typeof data === 'object' &&
            data !== null &&
            'status' in data &&
            typeof data.status === 'string' &&
            data.status.length > 0);
    }
    /**
     * Convert the response to a JSON object.
     *
     * @returns {ICertificationResponseJson} JSON representation of the response
     */
    toJSON() {
        return {
            status: this.status,
        };
    }
}
