// @unicitylabs/state-transition-sdk/lib/api/CertificationData.js
import { DataHash } from '../crypto/hash/DataHash.js';
import { HashAlgorithm } from '../crypto/hash/HashAlgorithm.js';
import { MintSigningService } from '../crypto/MintSigningService.js';
import { SignaturePredicateUnlockScript } from '../predicate/builtin/SignaturePredicateUnlockScript.js';
import { EncodedPredicate } from '../predicate/EncodedPredicate.js';
import { CborDeserializer } from '../serialization/cbor/CborDeserializer.js';
import { CborError } from '../serialization/cbor/CborError.js';
import { CborSerializer } from '../serialization/cbor/CborSerializer.js';
import { HexConverter } from '../util/HexConverter.js';
import { dedent } from '../util/StringUtils.js';
/**
 * Certification request data.
 */
export class CertificationData {
    lockScript;
    sourceStateHash;
    transactionHash;
    _unlockScript;
    static CBOR_TAG = 39031n;
    static VERSION = 1n;
    /**
     * Create a certification data object.
     * @param {IPredicate} lockScript
     * @param {DataHash} sourceStateHash
     * @param {DataHash} transactionHash
     * @param {Uint8Array} _unlockScript Unlock script bytes
     */
    constructor(lockScript, sourceStateHash, transactionHash, _unlockScript) {
        this.lockScript = lockScript;
        this.sourceStateHash = sourceStateHash;
        this.transactionHash = transactionHash;
        this._unlockScript = _unlockScript;
        this._unlockScript = new Uint8Array(_unlockScript);
    }
    /**
     * Get the unlock script.
     */
    get unlockScript() {
        return new Uint8Array(this._unlockScript);
    }
    /**
     * @returns {bigint} Wire-format version of this certification data.
     */
    get version() {
        return CertificationData.VERSION;
    }
    /**
     * Create CertificationData from CBOR bytes.
     * @param {Uint8Array} bytes CBOR bytes
     *
     * @returns {CertificationData} certification data
     */
    static fromCBOR(bytes) {
        const tag = CborDeserializer.decodeTag(bytes);
        if (tag.tag !== CertificationData.CBOR_TAG) {
            throw new CborError(`Invalid CBOR tag for CertificationData: ${tag.tag}`);
        }
        const data = CborDeserializer.decodeArray(tag.data, 5);
        const version = CborDeserializer.decodeUnsignedInteger(data[0]);
        if (version !== CertificationData.VERSION) {
            throw new CborError(`Unsupported CertificationData version: ${version}`);
        }
        return new CertificationData(EncodedPredicate.fromCBOR(data[1]), new DataHash(HashAlgorithm.SHA256, CborDeserializer.decodeByteString(data[2])), new DataHash(HashAlgorithm.SHA256, CborDeserializer.decodeByteString(data[3])), CborDeserializer.decodeByteString(data[4]));
    }
    /**
     * Create CertificationData from a mint transaction.
     *
     * @param {MintTransaction} transaction Mint transaction to certify.
     * @returns {Promise<CertificationData>} Certification data.
     */
    static async fromMintTransaction(transaction) {
        const signingService = await MintSigningService.create(transaction.tokenId);
        return CertificationData.fromTransaction(transaction, await SignaturePredicateUnlockScript.create(transaction, signingService));
    }
    /**
     * Create CertificationData from a transaction and its unlock script.
     *
     * @param {ITransaction} transaction Transaction to certify.
     * @param {IUnlockScript} unlockScript Unlock script for the transaction.
     * @returns {Promise<CertificationData>} Certification data.
     */
    static async fromTransaction(transaction, unlockScript) {
        const transactionHash = await transaction.calculateTransactionHash();
        return new CertificationData(transaction.lockScript, transaction.sourceStateHash, transactionHash, unlockScript.encode());
    }
    /**
     * Convert the certification data to CBOR bytes.
     *
     * @returns {Uint8Array} CBOR bytes
     */
    toCBOR() {
        return CborSerializer.encodeTag(CertificationData.CBOR_TAG, CborSerializer.encodeArray(CborSerializer.encodeUnsignedInteger(this.version), this.lockScript.toCBOR(), CborSerializer.encodeByteString(this.sourceStateHash.data), CborSerializer.encodeByteString(this.transactionHash.data), CborSerializer.encodeByteString(this._unlockScript)));
    }
    /**
     * Returns a string representation of the CertificationData.
     * @returns The string representation.
     */
    toString() {
        return dedent `
      Certification Data
        Version: ${CertificationData.VERSION}
        Owner Predicate: 
          ${this.lockScript.toString()}
        Source State Hash: ${this.sourceStateHash.toString()}
        Transaction Hash: ${this.transactionHash.toString()}
        Witness: ${HexConverter.encode(this._unlockScript)}`;
    }
}
