// @unicitylabs/state-transition-sdk/lib/api/CertificationRequest.js
import { StateId } from './StateId.js';
import { CborSerializer } from '../serialization/cbor/CborSerializer.js';
/**
 * Certification request object sent by the client to the aggregator.
 */
export class CertificationRequest {
    stateId;
    certificationData;
    static CBOR_TAG = 39030n;
    static VERSION = 1n;
    /**
     * Constructs a CertificationRequest instance.
     * @param {StateId} stateId Unique state identifier.
     * @param {CertificationData} certificationData Certification data.
     */
    constructor(stateId, certificationData) {
        this.stateId = stateId;
        this.certificationData = certificationData;
    }
    /**
     * @returns {bigint} Wire-format version of this request.
     */
    get version() {
        return CertificationRequest.VERSION;
    }
    /**
     * Create a new CertificationRequest instance.
     * @param {CertificationData} certificationData Certification data.
     *
     * @returns {Promise<CertificationRequest>} A promise that resolves to a CertificationRequest instance.
     */
    static async create(certificationData) {
        return new CertificationRequest(await StateId.fromCertificationData(certificationData), certificationData);
    }
    /**
     * Convert the request to a CBOR bytes.
     *
     * @returns CBOR bytes
     */
    toCBOR() {
        return CborSerializer.encodeTag(CertificationRequest.CBOR_TAG, CborSerializer.encodeArray(CborSerializer.encodeUnsignedInteger(this.version), this.stateId.toCBOR(), this.certificationData.toCBOR(), CborSerializer.encodeUnsignedInteger(0)));
    }
}
