// @unicitylabs/state-transition-sdk/lib/api/InclusionProof.js
import { UnicityCertificate } from './bft/UnicityCertificate.js';
import { CertificationData } from './CertificationData.js';
import { InclusionCertificate } from './InclusionCertificate.js';
import { CborDeserializer } from '../serialization/cbor/CborDeserializer.js';
import { CborError } from '../serialization/cbor/CborError.js';
import { CborSerializer } from '../serialization/cbor/CborSerializer.js';
import { dedent } from '../util/StringUtils.js';
/**
 * Represents a proof of inclusion or non inclusion in a sparse merkle tree.
 */
export class InclusionProof {
    certificationData;
    inclusionCertificate;
    unicityCertificate;
    static CBOR_TAG = 39033n;
    static VERSION = 1n;
    /**
     * Constructs an InclusionProof instance.
     * @param certificationData Certification data.
     * @param inclusionCertificate Inclusion certificate.
     * @param unicityCertificate Unicity certificate.
     */
    constructor(certificationData, inclusionCertificate, unicityCertificate) {
        this.certificationData = certificationData;
        this.inclusionCertificate = inclusionCertificate;
        this.unicityCertificate = unicityCertificate;
    }
    /**
     * @returns {bigint} Wire-format version of this inclusion proof.
     */
    get version() {
        return InclusionProof.VERSION;
    }
    /**
     * Decodes an InclusionProof from CBOR bytes.
     * @param bytes The CBOR-encoded bytes.
     * @returns An InclusionProof instance.
     */
    static fromCBOR(bytes) {
        const tag = CborDeserializer.decodeTag(bytes);
        if (tag.tag !== InclusionProof.CBOR_TAG) {
            throw new CborError(`Invalid CBOR tag for InclusionProof: ${tag.tag}`);
        }
        const data = CborDeserializer.decodeArray(tag.data, 4);
        const version = CborDeserializer.decodeUnsignedInteger(data[0]);
        if (version !== InclusionProof.VERSION) {
            throw new CborError(`Unsupported InclusionProof version: ${version}`);
        }
        return new InclusionProof(CborDeserializer.decodeNullable(data[1], CertificationData.fromCBOR), CborDeserializer.decodeNullable(data[2], (inclusionCertificate) => InclusionCertificate.decode(CborDeserializer.decodeByteString(inclusionCertificate))), UnicityCertificate.fromCBOR(data[3]));
    }
    /**
     * Encodes the InclusionProof to CBOR format.
     * @returns The CBOR-encoded bytes.
     */
    toCBOR() {
        return CborSerializer.encodeTag(InclusionProof.CBOR_TAG, CborSerializer.encodeArray(CborSerializer.encodeUnsignedInteger(this.version), CborSerializer.encodeNullable(this.certificationData, (certificationData) => certificationData.toCBOR()), CborSerializer.encodeNullable(this.inclusionCertificate, (inclusionCertificate) => CborSerializer.encodeByteString(inclusionCertificate.encode())), this.unicityCertificate.toCBOR()));
    }
    /**
     * Returns a string representation of the InclusionProof.
     * @returns The string representation.
     */
    toString() {
        return dedent `
      Inclusion Proof
        ${this.inclusionCertificate?.toString()}
        ${this.certificationData?.toString()}
        ${this.unicityCertificate.toString()}`;
    }
}
