// @unicitylabs/state-transition-sdk/lib/api/NetworkId.js
/**
 * Unicity network identifier (α). Used to scope token ids and other
 * network-bound values so they cannot be replayed across networks.
 */
export class NetworkId {
    id;
    name;
    static LOCAL = new NetworkId(3, 'LOCAL');
    static MAINNET = new NetworkId(1, 'MAINNET');
    static TESTNET = new NetworkId(2, 'TESTNET');
    constructor(id, name) {
        this.id = id;
        this.name = name;
    }
    /**
     * Resolve a NetworkId from its numeric identifier. Returns the registered
     * singleton for known ids; constructs a new (unnamed) instance for any
     * other value in the 16-bit unsigned range.
     *
     * @param {number|bigint} id Numeric network identifier.
     * @returns {NetworkId} NetworkId for the given identifier.
     * @throws {Error} If `id` is outside the 16-bit unsigned range.
     */
    static fromId(id) {
        const value = BigInt(id);
        if (value < 1n || value > 0xffffn) {
            throw new Error(`Network identifier out of allowed 16-bit unsigned range: ${id}.`);
        }
        const numeric = Number(value);
        switch (numeric) {
            case NetworkId.MAINNET.id:
                return NetworkId.MAINNET;
            case NetworkId.TESTNET.id:
                return NetworkId.TESTNET;
            case NetworkId.LOCAL.id:
                return NetworkId.LOCAL;
            default:
                return new NetworkId(numeric);
        }
    }
    /**
     * Equality check against another NetworkId.
     *
     * @param {NetworkId} other Other network identifier.
     * @returns {boolean} True if both share the same numeric id.
     */
    equals(other) {
        return this.id === other.id;
    }
    /**
     * @returns {string} `NetworkId[<name>]` for registered networks, `NetworkId[<id>]` otherwise.
     */
    toString() {
        return `NetworkId[${this.name ?? this.id}]`;
    }
}
