// @unicitylabs/nostr-js-sdk/dist/cjs/crypto/schnorr.js
"use strict";
/**
 * BIP-340 Schnorr signature implementation using secp256k1.
 * Uses @noble/curves for cryptographic operations.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPublicKey = getPublicKey;
exports.getPublicKeyHex = getPublicKeyHex;
exports.sign = sign;
exports.signHex = signHex;
exports.verify = verify;
exports.verifyHex = verifyHex;
const secp256k1_1 = require("@noble/curves/secp256k1");
const utils_1 = require("@noble/hashes/utils");
/**
 * Get the x-only public key from a private key (BIP-340)
 * @param privateKey 32-byte private key
 * @returns 32-byte x-only public key
 */
function getPublicKey(privateKey) {
    if (privateKey.length !== 32) {
        throw new Error('Private key must be 32 bytes');
    }
    return secp256k1_1.schnorr.getPublicKey(privateKey);
}
/**
 * Get the x-only public key from a hex-encoded private key
 * @param privateKeyHex Hex-encoded private key
 * @returns Hex-encoded x-only public key
 */
function getPublicKeyHex(privateKeyHex) {
    const privateKey = (0, utils_1.hexToBytes)(privateKeyHex);
    const publicKey = getPublicKey(privateKey);
    return (0, utils_1.bytesToHex)(publicKey);
}
/**
 * Sign a message using BIP-340 Schnorr signature
 * @param message 32-byte message hash to sign
 * @param privateKey 32-byte private key
 * @returns 64-byte Schnorr signature (R.x || s)
 */
function sign(message, privateKey) {
    if (message.length !== 32) {
        throw new Error('Message must be 32 bytes');
    }
    if (privateKey.length !== 32) {
        throw new Error('Private key must be 32 bytes');
    }
    return secp256k1_1.schnorr.sign(message, privateKey);
}
/**
 * Sign a message and return hex-encoded signature
 * @param message 32-byte message hash to sign
 * @param privateKey 32-byte private key
 * @returns Hex-encoded 64-byte Schnorr signature
 */
function signHex(message, privateKey) {
    return (0, utils_1.bytesToHex)(sign(message, privateKey));
}
/**
 * Verify a BIP-340 Schnorr signature
 * @param signature 64-byte Schnorr signature
 * @param message 32-byte message hash
 * @param publicKey 32-byte x-only public key
 * @returns true if the signature is valid
 */
function verify(signature, message, publicKey) {
    if (signature.length !== 64) {
        return false;
    }
    if (message.length !== 32) {
        return false;
    }
    if (publicKey.length !== 32) {
        return false;
    }
    try {
        return secp256k1_1.schnorr.verify(signature, message, publicKey);
    }
    catch {
        return false;
    }
}
/**
 * Verify a hex-encoded Schnorr signature
 * @param signatureHex Hex-encoded 64-byte signature
 * @param message 32-byte message hash
 * @param publicKeyHex Hex-encoded 32-byte public key
 * @returns true if the signature is valid
 */
function verifyHex(signatureHex, message, publicKeyHex) {
    try {
        const signature = (0, utils_1.hexToBytes)(signatureHex);
        const publicKey = (0, utils_1.hexToBytes)(publicKeyHex);
        return verify(signature, message, publicKey);
    }
    catch {
        return false;
    }
}
