// @unicitylabs/state-transition-sdk/lib/predicate/builtin/verification/UnicityIdPredicateVerifier.js
import { VerificationResult } from '../../../verification/VerificationResult.js';
import { VerificationStatus } from '../../../verification/VerificationStatus.js';
import { EncodedPredicate } from '../../EncodedPredicate.js';
import { BuiltInPredicateType } from '../BuiltInPredicateType.js';
import { UnicityIdPredicate } from '../UnicityIdPredicate.js';
import { UnicityIdPredicateUnlockScript } from '../UnicityIdPredicateUnlockScript.js';
/**
 * Verifier for {@link UnicityIdPredicate}: confirms the unlock script
 * carries a valid unicity-id token whose id matches the predicate, then
 * recursively verifies the token's target predicate against the spend.
 */
export class UnicityIdPredicateVerifier {
    verifier;
    trustBase;
    issuerPublicKey;
    constructor(verifier, trustBase, issuerPublicKey) {
        this.verifier = verifier;
        this.trustBase = trustBase;
        this.issuerPublicKey = issuerPublicKey;
    }
    /**
     * @returns {BuiltInPredicateType} UnicityId predicate type id.
     */
    get type() {
        return BuiltInPredicateType.UnicityId;
    }
    /**
     * @inheritDoc
     */
    async verify(encodedPredicate, sourceStateHash, transactionHash, unlockScript) {
        const predicate = UnicityIdPredicate.fromPredicate(encodedPredicate);
        const decodedUnlockScript = await UnicityIdPredicateUnlockScript.decode(unlockScript);
        const tokenId = await predicate.unicityId.toTokenId();
        if (!tokenId.equals(decodedUnlockScript.token.id)) {
            return new VerificationResult('UnicityIdPredicateVerifier', VerificationStatus.FAIL, 'Token ID mismatch.');
        }
        let result = await decodedUnlockScript.token.verify(this.trustBase, this.verifier, this.issuerPublicKey);
        if (result.status !== VerificationStatus.OK) {
            return new VerificationResult('UnicityIdPredicateVerifier', VerificationStatus.FAIL, 'Could not verify unicity id token.', [result]);
        }
        result = await this.verifier.verify(EncodedPredicate.fromPredicate(decodedUnlockScript.token.genesis.targetPredicate), sourceStateHash, transactionHash, decodedUnlockScript.unlockScript);
        if (result.status !== VerificationStatus.OK) {
            return new VerificationResult('UnicityIdPredicateVerifier', VerificationStatus.FAIL, 'Could not verify target predicate.', [result]);
        }
        return new VerificationResult('UnicityIdPredicateVerifier', VerificationStatus.OK);
    }
}
