// @unicitylabs/state-transition-sdk/lib/serialization/cbor/CborDeserializer.js
import { CborError } from './CborError.js';
import { CborMap } from './CborMap.js';
import { CborMapEntry } from './CborMapEntry.js';
import { CborReader } from './CborReader.js';
import { MajorType } from './MajorType.js';
import { areUint8ArraysEqual } from '../../util/TypedArrayUtils.js';
/**
 * Static helpers that decode canonical CBOR bytes into TypeScript values.
 * Every method consumes the entire input and asserts there are no trailing
 * bytes.
 */
export class CborDeserializer {
    static FALSE_CBOR = new Uint8Array([0xf4]);
    static NULL_CBOR = new Uint8Array([0xf6]);
    static TRUE_CBOR = new Uint8Array([0xf5]);
    /**
     * Decode a CBOR array.
     *
     * @param {Uint8Array} data CBOR bytes.
     * @param {number|null} expectedLength If set, throws when the array has a different length.
     * @returns {Uint8Array[]} CBOR bytes of each element.
     * @throws {CborError} On wrong length or trailing bytes.
     */
    static decodeArray(data, expectedLength = null) {
        const reader = new CborReader(data);
        const length = reader.readLength(MajorType.ARRAY);
        if (length > 0xffffffff) {
            throw new CborError('Array too long.');
        }
        const result = [];
        for (let i = 0; i < length; i++) {
            result.push(reader.readRawCbor());
        }
        if (expectedLength !== null && expectedLength !== result.length) {
            throw new CborError(`Expected array length ${expectedLength}, got ${result.length}.`);
        }
        reader.assertExhausted();
        return result;
    }
    /**
     * Decode a CBOR boolean.
     *
     * @param {Uint8Array} data CBOR bytes.
     * @returns {boolean} Boolean value.
     * @throws {CborError} If the input encodes anything other than `true` or `false`.
     */
    static decodeBoolean(data) {
        const reader = new CborReader(data);
        const cbor = reader.readRawCbor();
        reader.assertExhausted();
        if (areUint8ArraysEqual(cbor, CborDeserializer.TRUE_CBOR)) {
            return true;
        }
        if (areUint8ArraysEqual(cbor, CborDeserializer.FALSE_CBOR)) {
            return false;
        }
        throw new CborError('Type mismatch, expected boolean.');
    }
    /**
     * Decode a CBOR byte string.
     *
     * @param {Uint8Array} data CBOR bytes.
     * @returns {Uint8Array} Decoded payload bytes.
     */
    static decodeByteString(data) {
        const reader = new CborReader(data);
        const length = reader.readLength(MajorType.BYTE_STRING);
        if (length > 0xffffffff) {
            throw new CborError('Byte string too long.');
        }
        const result = reader.read(Number(length));
        reader.assertExhausted();
        return result;
    }
    /**
     * Decode a CBOR map, enforcing canonical key ordering and rejecting
     * duplicate keys.
     *
     * @param {Uint8Array} data CBOR bytes.
     * @returns {CborMapEntry[]} Decoded entries.
     * @throws {CborError} On duplicate or out-of-order keys.
     */
    static decodeMap(data) {
        const reader = new CborReader(data);
        const length = reader.readLength(MajorType.MAP);
        if (length > 0xffffffff) {
            throw new CborError('Map too long.');
        }
        const result = [];
        for (let i = 0; i < length; i++) {
            const entry = new CborMapEntry(reader.readRawCbor(), reader.readRawCbor());
            if (result.length > 0) {
                const comparison = CborMap.compareEntries(result[result.length - 1], entry);
                if (comparison === 0) {
                    throw new CborError('Duplicate map key found.');
                }
                if (comparison > 0) {
                    throw new CborError('Map keys are not in canonical order.');
                }
            }
            result.push(entry);
        }
        reader.assertExhausted();
        return result;
    }
    /**
     * Decode a CBOR negative integer.
     *
     * @returns {bigint} Decoded integer.
     * @throws {CborError} Always; not yet implemented.
     */
    static decodeNegativeInteger() {
        throw new CborError('Not implemented.');
    }
    /**
     * Decode `data` with `decode`, or return `null` if the input is CBOR `null`.
     *
     * @param {Uint8Array} data CBOR bytes.
     * @param {(data: Uint8Array) => T} decode Decoder for the inner value.
     * @returns {T|null} Decoded value or `null`.
     */
    static decodeNullable(data, decode) {
        const reader = new CborReader(data);
        const cbor = reader.readRawCbor();
        reader.assertExhausted();
        if (areUint8ArraysEqual(cbor, CborDeserializer.NULL_CBOR)) {
            return null;
        }
        return decode(cbor);
    }
    /**
     * Decode a CBOR tag.
     *
     * @param {Uint8Array} data CBOR bytes.
     * @returns {{data: Uint8Array, tag: bigint}} Tag number and CBOR bytes of the tagged value.
     */
    static decodeTag(data) {
        const reader = new CborReader(data);
        const tag = reader.readLength(MajorType.TAG);
        const cbor = reader.readRawCbor();
        reader.assertExhausted();
        return { data: cbor, tag };
    }
    /**
     * Decode a CBOR text string.
     *
     * @param {Uint8Array} data CBOR bytes.
     * @returns {string} UTF-8 decoded string.
     */
    static decodeTextString(data) {
        const reader = new CborReader(data);
        const length = reader.readLength(MajorType.TEXT_STRING);
        if (length > BigInt(Number.MAX_SAFE_INTEGER)) {
            throw new CborError('Text string too long.');
        }
        const result = reader.read(Number(length));
        reader.assertExhausted();
        return new TextDecoder().decode(result);
    }
    /**
     * Decode a CBOR unsigned integer.
     *
     * @param {Uint8Array} data CBOR bytes.
     * @returns {bigint} Decoded integer.
     */
    static decodeUnsignedInteger(data) {
        const reader = new CborReader(data);
        const result = reader.readLength(MajorType.UNSIGNED_INTEGER);
        reader.assertExhausted();
        return result;
    }
}
