// @unicitylabs/state-transition-sdk/lib/transaction/CertifiedTransferTransaction.js
import { TransferTransaction } from './TransferTransaction.js';
import { InclusionProof } from '../api/InclusionProof.js';
import { CborDeserializer } from '../serialization/cbor/CborDeserializer.js';
import { CborSerializer } from '../serialization/cbor/CborSerializer.js';
import { dedent } from '../util/StringUtils.js';
import { VerificationError } from '../verification/VerificationError.js';
import { InclusionProofVerificationRule, InclusionProofVerificationStatus, } from './verification/rule/InclusionProofVerificationRule.js';
/**
 * Transfer transaction bundled with a verified inclusion proof.
 */
export class CertifiedTransferTransaction {
    transaction;
    inclusionProof;
    constructor(transaction, inclusionProof) {
        this.transaction = transaction;
        this.inclusionProof = inclusionProof;
    }
    /**
     * @returns {Uint8Array|null} Data payload of the inner transaction.
     */
    get data() {
        return this.transaction.data;
    }
    /**
     * @returns {EncodedPredicate} Lock script of the inner transaction.
     */
    get lockScript() {
        return this.transaction.lockScript;
    }
    /**
     * @returns {EncodedPredicate} Recipient predicate of the inner transaction.
     */
    get recipient() {
        return this.transaction.recipient;
    }
    /**
     * @returns {DataHash} Source state hash of the inner transaction.
     */
    get sourceStateHash() {
        return this.transaction.sourceStateHash;
    }
    /**
     * @returns {Uint8Array} State mask of the inner transaction.
     */
    get stateMask() {
        return this.transaction.stateMask;
    }
    /**
     * Create CertifiedTransferTransaction from CBOR bytes.
     *
     * @param {Uint8Array} bytes CBOR bytes.
     * @param {Token} token Token providing context for the transfer transaction.
     * @returns {Promise<CertifiedTransferTransaction>} Decoded certified transaction.
     */
    static async fromCBOR(bytes, token) {
        const data = CborDeserializer.decodeArray(bytes, 2);
        return new CertifiedTransferTransaction(await TransferTransaction.fromCBOR(data[0], token), InclusionProof.fromCBOR(data[1]));
    }
    /**
     * Create CertifiedTransferTransaction from mint transaction and inclusion proof.
     *
     * @param {RootTrustBase} trustBase Root trust base used to verify the inclusion certificate.
     * @param {PredicateVerifierService} predicateVerifier Verifier for any embedded predicates.
     * @param {TransferTransaction} transaction Transaction to certify.
     * @param {InclusionProof} inclusionProof Inclusion proof for the transaction.
     * @returns {Promise<CertifiedTransferTransaction>} Verified certified transaction.
     * @throws {VerificationError} If the inclusion proof does not verify.
     */
    static async fromTransaction(trustBase, predicateVerifier, transaction, inclusionProof) {
        const result = await InclusionProofVerificationRule.verify(trustBase, predicateVerifier, inclusionProof, transaction);
        if (result.status !== InclusionProofVerificationStatus.OK) {
            throw new VerificationError('Inclusion proof verification failed', result);
        }
        return new CertifiedTransferTransaction(transaction, inclusionProof);
    }
    /**
     * @inheritDoc
     */
    calculateStateHash() {
        return this.transaction.calculateStateHash();
    }
    /**
     * @inheritDoc
     */
    calculateTransactionHash() {
        return this.transaction.calculateTransactionHash();
    }
    /**
     * @inheritDoc
     */
    toCBOR() {
        return CborSerializer.encodeArray(this.transaction.toCBOR(), this.inclusionProof.toCBOR());
    }
    /**
     * @returns {string} String representation of the certified transaction.
     */
    toString() {
        return dedent `
      CertifiedTransferTransaction
        ${this.transaction.toString()}
        ${this.inclusionProof.toString()}`;
    }
}
