// @unicitylabs/state-transition-sdk/lib/transaction/CertifiedMintTransaction.js
import { MintTransaction } from './MintTransaction.js';
import { InclusionProof } from '../api/InclusionProof.js';
import { CborDeserializer } from '../serialization/cbor/CborDeserializer.js';
import { CborSerializer } from '../serialization/cbor/CborSerializer.js';
import { dedent } from '../util/StringUtils.js';
import { VerificationError } from '../verification/VerificationError.js';
import { InclusionProofVerificationRule, InclusionProofVerificationStatus, } from './verification/rule/InclusionProofVerificationRule.js';
/**
 * Mint transaction bundled with a verified inclusion proof.
 */
export class CertifiedMintTransaction {
    transaction;
    inclusionProof;
    constructor(transaction, inclusionProof) {
        this.transaction = transaction;
        this.inclusionProof = inclusionProof;
    }
    /**
     * @returns {Uint8Array|null} Data payload of the inner transaction.
     */
    get data() {
        return this.transaction.data;
    }
    /**
     * @returns {Uint8Array|null} Mint justification bytes of the inner transaction.
     */
    get justification() {
        return this.transaction.justification;
    }
    /**
     * @returns {EncodedPredicate} Lock script of the inner transaction.
     */
    get lockScript() {
        return this.transaction.lockScript;
    }
    /**
     * @returns {NetworkId} Network identifier of the inner transaction.
     */
    get networkId() {
        return this.transaction.networkId;
    }
    /**
     * @returns {EncodedPredicate} Recipient predicate of the inner transaction.
     */
    get recipient() {
        return this.transaction.recipient;
    }
    /**
     * @returns {TokenSalt} Mint-transaction salt of the inner transaction.
     */
    get salt() {
        return this.transaction.salt;
    }
    /**
     * @returns {DataHash} Source state hash of the inner transaction.
     */
    get sourceStateHash() {
        return this.transaction.sourceStateHash;
    }
    /**
     * @returns {Uint8Array} State mask of the inner transaction.
     */
    get stateMask() {
        return this.transaction.stateMask;
    }
    /**
     * @returns {TokenId} Token id of the inner transaction (derived from networkId and salt).
     */
    get tokenId() {
        return this.transaction.tokenId;
    }
    /**
     * @returns {TokenType} Token type of the inner transaction.
     */
    get tokenType() {
        return this.transaction.tokenType;
    }
    /**
     * Create CertifiedMintTransaction from CBOR bytes.
     *
     * @param {Uint8Array} bytes CBOR bytes.
     * @returns {Promise<CertifiedMintTransaction>} Decoded certified transaction.
     */
    static async fromCBOR(bytes) {
        const data = CborDeserializer.decodeArray(bytes, 2);
        return new CertifiedMintTransaction(await MintTransaction.fromCBOR(data[0]), InclusionProof.fromCBOR(data[1]));
    }
    /**
     * Create CertifiedMintTransaction from mint transaction and inclusion proof.
     *
     * @param {RootTrustBase} trustBase Root trust base used to verify the inclusion certificate.
     * @param {PredicateVerifierService} predicateVerifier Verifier for any embedded predicates.
     * @param {MintTransaction} transaction Transaction to certify.
     * @param {InclusionProof} inclusionProof Inclusion proof for the transaction.
     * @returns {Promise<CertifiedMintTransaction>} Verified certified transaction.
     * @throws {VerificationError} If the inclusion proof does not verify.
     */
    static async fromTransaction(trustBase, predicateVerifier, transaction, inclusionProof) {
        const result = await InclusionProofVerificationRule.verify(trustBase, predicateVerifier, inclusionProof, transaction);
        if (result.status !== InclusionProofVerificationStatus.OK) {
            throw new VerificationError('Inclusion proof verification failed', result);
        }
        return new CertifiedMintTransaction(transaction, inclusionProof);
    }
    /**
     * @inheritDoc
     */
    calculateStateHash() {
        return this.transaction.calculateStateHash();
    }
    /**
     * @inheritDoc
     */
    calculateTransactionHash() {
        return this.transaction.calculateTransactionHash();
    }
    /**
     * @inheritDoc
     */
    toCBOR() {
        return CborSerializer.encodeArray(this.transaction.toCBOR(), this.inclusionProof.toCBOR());
    }
    /**
     * @returns {string} String representation of the certified transaction.
     */
    toString() {
        return dedent `
      CertifiedMintTransaction
        ${this.transaction.toString()}
        ${this.inclusionProof.toString()}`;
    }
}
