// @unicitylabs/state-transition-sdk/lib/transaction/MintTransaction.js
import { CertifiedMintTransaction } from './CertifiedMintTransaction.js';
import { MintTransactionState } from './MintTransactionState.js';
import { TokenId } from './TokenId.js';
import { TokenSalt } from './TokenSalt.js';
import { TokenType } from './TokenType.js';
import { NetworkId } from '../api/NetworkId.js';
import { DataHasher } from '../crypto/hash/DataHasher.js';
import { HashAlgorithm } from '../crypto/hash/HashAlgorithm.js';
import { MintSigningService } from '../crypto/MintSigningService.js';
import { SignaturePredicate } from '../predicate/builtin/SignaturePredicate.js';
import { EncodedPredicate } from '../predicate/EncodedPredicate.js';
import { CborDeserializer } from '../serialization/cbor/CborDeserializer.js';
import { CborError } from '../serialization/cbor/CborError.js';
import { CborSerializer } from '../serialization/cbor/CborSerializer.js';
import { HexConverter } from '../util/HexConverter.js';
import { dedent } from '../util/StringUtils.js';
/**
 * Token mint transaction.
 */
export class MintTransaction {
    sourceStateHash;
    lockScript;
    networkId;
    recipient;
    salt;
    tokenType;
    tokenId;
    _justification;
    _data;
    static CBOR_TAG = 39041n;
    static VERSION = 1n;
    constructor(sourceStateHash, lockScript, networkId, recipient, salt, tokenType, tokenId, _justification, _data) {
        this.sourceStateHash = sourceStateHash;
        this.lockScript = lockScript;
        this.networkId = networkId;
        this.recipient = recipient;
        this.salt = salt;
        this.tokenType = tokenType;
        this.tokenId = tokenId;
        this._justification = _justification;
        this._data = _data;
    }
    /**
     * @returns {Uint8Array|null} Copy of the data payload, or `null` if absent.
     */
    get data() {
        return this._data ? new Uint8Array(this._data) : null;
    }
    /**
     * @returns {Uint8Array|null} Copy of the mint justification bytes, or `null` if absent.
     */
    get justification() {
        return this._justification ? new Uint8Array(this._justification) : null;
    }
    /**
     * @returns {Uint8Array} State mask used when computing the resulting state hash.
     */
    get stateMask() {
        return new Uint8Array(this.tokenId.bytes);
    }
    /**
     * @returns {bigint} Wire-format version of this transaction.
     */
    get version() {
        return MintTransaction.VERSION;
    }
    /**
     * Create a MintTransaction for a fresh token.
     *
     * @param {NetworkId} networkId Network identifier.
     * @param {IPredicate} recipient Predicate that will lock the minted state.
     * @param {Uint8Array|null} data Optional data payload.
     * @param {TokenType} tokenType Token type being minted.
     * @param {TokenSalt} salt Mint-transaction salt; defaults to a random 32-byte salt.
     * @param {Uint8Array|null} justification Optional mint justification bytes.
     * @returns {Promise<MintTransaction>} New mint transaction.
     */
    static async create(networkId, recipient, data = null, tokenType = TokenType.generate(), salt = TokenSalt.generate(), justification = null) {
        justification = justification ? new Uint8Array(justification) : null;
        data = data ? new Uint8Array(data) : null;
        const tokenId = await TokenId.fromSalt(networkId, salt);
        const signingService = await MintSigningService.create(tokenId);
        return new MintTransaction(await MintTransactionState.create(tokenId), EncodedPredicate.fromPredicate(SignaturePredicate.fromSigningService(signingService)), networkId, EncodedPredicate.fromPredicate(recipient), salt, tokenType, tokenId, justification, data);
    }
    /**
     * Create MintTransaction from CBOR bytes.
     *
     * @param {Uint8Array} bytes CBOR bytes.
     * @returns {Promise<MintTransaction>} Decoded transaction.
     * @throws {CborError} On wrong tag or unsupported version.
     */
    static fromCBOR(bytes) {
        const tag = CborDeserializer.decodeTag(bytes);
        if (tag.tag !== MintTransaction.CBOR_TAG) {
            throw new CborError(`Invalid CBOR tag for MintTransaction: ${tag.tag}`);
        }
        const data = CborDeserializer.decodeArray(tag.data, 7);
        const version = CborDeserializer.decodeUnsignedInteger(data[0]);
        if (version !== MintTransaction.VERSION) {
            throw new CborError(`Unsupported MintTransaction version: ${version}`);
        }
        return MintTransaction.create(NetworkId.fromId(CborDeserializer.decodeUnsignedInteger(data[1])), EncodedPredicate.fromCBOR(data[2]), CborDeserializer.decodeNullable(data[6], CborDeserializer.decodeByteString), TokenType.fromCBOR(data[4]), TokenSalt.fromCBOR(data[3]), CborDeserializer.decodeNullable(data[5], CborDeserializer.decodeByteString));
    }
    /**
     * @inheritDoc
     */
    calculateStateHash() {
        return new DataHasher(HashAlgorithm.SHA256)
            .update(CborSerializer.encodeArray(CborSerializer.encodeByteString(this.sourceStateHash.imprint), CborSerializer.encodeByteString(this.stateMask)))
            .digest();
    }
    /**
     * @inheritDoc
     */
    calculateTransactionHash() {
        return new DataHasher(HashAlgorithm.SHA256).update(this.toCBOR()).digest();
    }
    /**
     * @inheritDoc
     */
    toCBOR() {
        return CborSerializer.encodeTag(MintTransaction.CBOR_TAG, CborSerializer.encodeArray(CborSerializer.encodeUnsignedInteger(this.version), CborSerializer.encodeUnsignedInteger(this.networkId.id), this.recipient.toCBOR(), this.salt.toCBOR(), this.tokenType.toCBOR(), CborSerializer.encodeNullable(this._justification, CborSerializer.encodeByteString), CborSerializer.encodeNullable(this._data, CborSerializer.encodeByteString)));
    }
    /**
     * Bundle this transaction with its inclusion proof into a CertifiedMintTransaction.
     *
     * @param {RootTrustBase} trustBase Root trust base used to verify the inclusion certificate.
     * @param {PredicateVerifierService} predicateVerifier Verifier for any predicates.
     * @param {InclusionProof} inclusionProof Inclusion proof for this transaction.
     * @returns {Promise<CertifiedMintTransaction>} Verified certified transaction.
     */
    toCertifiedTransaction(trustBase, predicateVerifier, inclusionProof) {
        return CertifiedMintTransaction.fromTransaction(trustBase, predicateVerifier, this, inclusionProof);
    }
    /**
     * @returns {string} String representation of the transaction.
     */
    toString() {
        return dedent `
      MintTransaction
        Version: ${this.version.toString()}
        Network ID: ${this.networkId.toString()}
        Lock Script:
          ${this.lockScript.toString()}
        Recipient: ${this.recipient.toString()}
        Salt: ${this.salt.toString()}
        Token ID: ${this.tokenId.toString()}
        Token Type: ${this.tokenType.toString()}
        Mint Justification: ${this._justification ? HexConverter.encode(this._justification) : 'null'}
        Data: ${this._data ? HexConverter.encode(this._data) : 'null'}`;
    }
}
