// @unicitylabs/state-transition-sdk/lib/transaction/TokenId.js
import { DataHasher } from '../crypto/hash/DataHasher.js';
import { HashAlgorithm } from '../crypto/hash/HashAlgorithm.js';
import { CborDeserializer } from '../serialization/cbor/CborDeserializer.js';
import { CborSerializer } from '../serialization/cbor/CborSerializer.js';
import { BitString } from '../util/BitString.js';
import { HexConverter } from '../util/HexConverter.js';
import { areUint8ArraysEqual } from '../util/TypedArrayUtils.js';
/**
 * Globally unique identifier of a token.
 */
export class TokenId {
    _bytes;
    constructor(_bytes) {
        this._bytes = _bytes;
        this._bytes = new Uint8Array(_bytes);
    }
    /**
     * @returns {Uint8Array} Copy of the identifier bytes.
     */
    get bytes() {
        return new Uint8Array(this._bytes);
    }
    /**
     * Create TokenId from CBOR bytes.
     *
     * @param {Uint8Array} bytes CBOR bytes.
     * @returns {TokenId} Decoded token id.
     */
    static fromCBOR(bytes) {
        return new TokenId(CborDeserializer.decodeByteString(bytes));
    }
    /**
     * Derive a TokenId from a salt and network identifier.
     *
     * @param {NetworkId} networkId Network identifier.
     * @param {TokenSalt} salt Mint-transaction salt.
     * @returns {Promise<TokenId>} Derived token id.
     */
    static async fromSalt(networkId, salt) {
        const hash = await new DataHasher(HashAlgorithm.SHA256)
            .update(CborSerializer.encodeArray(salt.toCBOR(), CborSerializer.encodeUnsignedInteger(networkId.id)))
            .digest();
        return new TokenId(hash.data);
    }
    /**
     * Equality check against another value.
     *
     * @param {unknown} o Other value.
     * @returns {boolean} True if `o` is a TokenId with the same bytes.
     */
    equals(o) {
        if (this === o) {
            return true;
        }
        if (!(o instanceof TokenId)) {
            return false;
        }
        return areUint8ArraysEqual(this._bytes, o._bytes);
    }
    /**
     * Convert the TokenId to a bit-string representation.
     *
     * @returns {BitString} Bit-string view of the identifier bytes.
     */
    toBitString() {
        return BitString.fromBytes(this._bytes);
    }
    /**
     * Convert TokenId to CBOR bytes.
     *
     * @returns {Uint8Array} CBOR bytes.
     */
    toCBOR() {
        return CborSerializer.encodeByteString(this._bytes);
    }
    /**
     * @returns {string} Human-readable representation of the token id.
     */
    toString() {
        return `TokenId[${HexConverter.encode(this._bytes)}]`;
    }
}
