// @unicitylabs/state-transition-sdk/lib/transaction/TokenType.js
import { CborDeserializer } from '../serialization/cbor/CborDeserializer.js';
import { CborSerializer } from '../serialization/cbor/CborSerializer.js';
import { HexConverter } from '../util/HexConverter.js';
/** Unique identifier describing the type/category of a token. */
export class TokenType {
    _bytes;
    constructor(_bytes) {
        this._bytes = _bytes;
        this._bytes = new Uint8Array(_bytes);
    }
    /**
     * @returns {Uint8Array} Copy of the token type bytes.
     */
    get bytes() {
        return new Uint8Array(this._bytes);
    }
    /**
     * Create TokenType from CBOR bytes.
     *
     * @param {Uint8Array} bytes CBOR bytes.
     * @returns {TokenType} Decoded token type.
     */
    static fromCBOR(bytes) {
        return new TokenType(CborDeserializer.decodeByteString(bytes));
    }
    /**
     * Generate a fresh random TokenType.
     *
     * @returns {TokenType} New token type with random 32-byte payload.
     */
    static generate() {
        return new TokenType(crypto.getRandomValues(new Uint8Array(32)));
    }
    /**
     * Convert TokenType to CBOR bytes.
     *
     * @returns {Uint8Array} CBOR bytes.
     */
    toCBOR() {
        return CborSerializer.encodeByteString(this._bytes);
    }
    /**
     * @returns {string} Human-readable representation of the token type.
     */
    toString() {
        return `TokenType[${HexConverter.encode(this._bytes)}]`;
    }
}
