// @unicitylabs/state-transition-sdk/lib/transaction/Token.js
import { CertifiedMintTransaction } from './CertifiedMintTransaction.js';
import { CertifiedTransferTransaction } from './CertifiedTransferTransaction.js';
import { CborDeserializer } from '../serialization/cbor/CborDeserializer.js';
import { CborError } from '../serialization/cbor/CborError.js';
import { CborSerializer } from '../serialization/cbor/CborSerializer.js';
import { dedent } from '../util/StringUtils.js';
import { VerificationError } from '../verification/VerificationError.js';
import { VerificationResult } from '../verification/VerificationResult.js';
import { VerificationStatus } from '../verification/VerificationStatus.js';
import { CertifiedMintTransactionVerificationRule } from './verification/rule/CertifiedMintTransactionVerificationRule.js';
import { CertifiedTransferTransactionVerificationRule } from './verification/rule/CertifiedTransferTransactionVerificationRule.js';
/**
 * Token representation.
 */
export class Token {
    genesis;
    _transactions;
    static CBOR_TAG = 39040n;
    static VERSION = 1n;
    constructor(genesis, _transactions = []) {
        this.genesis = genesis;
        this._transactions = _transactions;
    }
    /**
     * @returns {TokenId} Token id from the genesis transaction.
     */
    get id() {
        return this.genesis.tokenId;
    }
    /**
     * @returns {ITransaction} Latest transaction, or the genesis if there are no transfers.
     */
    get latestTransaction() {
        return this._transactions.at(-1) ?? this.genesis;
    }
    /**
     * @returns {CertifiedTransferTransaction[]} Copy of the transfer history.
     */
    get transactions() {
        return this._transactions.slice();
    }
    /**
     * @returns {TokenType} Token type from the genesis transaction.
     */
    get type() {
        return this.genesis.tokenType;
    }
    /**
     * @returns {bigint} Wire-format version of this token.
     */
    get version() {
        return Token.VERSION;
    }
    /**
     * Create Token from CBOR bytes.
     *
     * @param {Uint8Array} bytes CBOR bytes.
     * @returns {Promise<Token>} Decoded token.
     * @throws {CborError} On wrong tag or unsupported version.
     */
    static async fromCBOR(bytes) {
        const tag = CborDeserializer.decodeTag(bytes);
        if (tag.tag !== Token.CBOR_TAG) {
            throw new CborError(`Invalid CBOR tag for Token: ${tag.tag}`);
        }
        const data = CborDeserializer.decodeArray(tag.data, 3);
        const version = CborDeserializer.decodeUnsignedInteger(data[0]);
        if (version !== Token.VERSION) {
            throw new CborError(`Unsupported Token version: ${version}`);
        }
        const transactionsBytes = CborDeserializer.decodeArray(data[2]);
        const genesis = await CertifiedMintTransaction.fromCBOR(data[1]);
        const transactions = [];
        const token = new Token(genesis, transactions);
        for (const transaction of transactionsBytes) {
            transactions.push(await CertifiedTransferTransaction.fromCBOR(transaction, token));
        }
        return token;
    }
    /**
     * Create a Token from a verified genesis mint transaction.
     *
     * @param {RootTrustBase} trustBase Root trust base used to verify the inclusion certificate.
     * @param {PredicateVerifierService} predicateVerifier Verifier for embedded predicates.
     * @param {MintJustificationVerifierService} mintJustificationVerifier Verifier for the mint justification.
     * @param {CertifiedMintTransaction} genesis Genesis mint transaction.
     * @returns {Promise<Token>} New token.
     * @throws {VerificationError} If the genesis does not verify.
     */
    static async mint(trustBase, predicateVerifier, mintJustificationVerifier, genesis) {
        const token = new Token(genesis);
        const result = await token.verify(trustBase, predicateVerifier, mintJustificationVerifier);
        if (result.status !== VerificationStatus.OK) {
            throw new VerificationError('Invalid token genesis', result);
        }
        return token;
    }
    /**
     * Convert Token to CBOR bytes.
     *
     * @returns {Uint8Array} CBOR bytes.
     */
    toCBOR() {
        return CborSerializer.encodeTag(Token.CBOR_TAG, CborSerializer.encodeArray(CborSerializer.encodeUnsignedInteger(this.version), this.genesis.toCBOR(), CborSerializer.encodeArray(...this._transactions.map((transaction) => transaction.toCBOR()))));
    }
    /**
     * @returns {string} String representation of the token.
     */
    toString() {
        return dedent `
      Token
        Version: ${this.version.toString()}
        ${this.genesis.toString()}
        Transactions: [
          ${this._transactions.map((transaction) => transaction.toString()).join('\n')}
        ]`;
    }
    /**
     * Append token with certified transfer transaction.
     *
     * @param {RootTrustBase} trustBase Root trust base used to verify the inclusion certificate.
     * @param {PredicateVerifierService} predicateVerifier Verifier for embedded predicates.
     * @param {CertifiedTransferTransaction} transaction Transfer transaction to apply.
     * @returns {Promise<Token>} Updated token including the new transaction.
     * @throws {VerificationError} If the transfer transaction does not verify.
     */
    async transfer(trustBase, predicateVerifier, transaction) {
        const result = await CertifiedTransferTransactionVerificationRule.verify(trustBase, predicateVerifier, transaction);
        if (result.status !== VerificationStatus.OK) {
            throw new VerificationError('Invalid transfer transaction', result);
        }
        const transactions = this.transactions.slice();
        transactions.push(transaction);
        return new Token(this.genesis, transactions);
    }
    /**
     * Verify the genesis and every transfer in this token.
     *
     * @param {RootTrustBase} trustBase Root trust base used to verify inclusion certificates.
     * @param {PredicateVerifierService} predicateVerifier Verifier for embedded predicates.
     * @param {MintJustificationVerifierService} mintJustificationVerifier Verifier for the mint justification.
     * @returns {Promise<VerificationResult<VerificationStatus>>} Aggregated verification result.
     */
    async verify(trustBase, predicateVerifier, mintJustificationVerifier) {
        const results = [];
        const result = await CertifiedMintTransactionVerificationRule.verify(trustBase, predicateVerifier, mintJustificationVerifier, this.genesis);
        results.push(result);
        if (result.status !== VerificationStatus.OK) {
            return new VerificationResult('TokenVerification', VerificationStatus.FAIL, '', results);
        }
        const transferResults = [];
        for (let i = 0; i < this._transactions.length; i++) {
            const transaction = this._transactions[i];
            const result = await CertifiedTransferTransactionVerificationRule.verify(trustBase, predicateVerifier, transaction);
            transferResults.push(result);
            if (result.status !== VerificationStatus.OK) {
                results.push(new VerificationResult('TokenTransferVerification', VerificationStatus.FAIL, `Transaction[${i}] verification failed.`, transferResults));
                return new VerificationResult('TokenVerification', VerificationStatus.FAIL, '', results);
            }
        }
        results.push(new VerificationResult('TokenTransferVerification', VerificationStatus.OK, ``, transferResults));
        return new VerificationResult('TokenVerification', VerificationStatus.OK, '', results);
    }
}
