// @unicitylabs/state-transition-sdk/lib/transaction/TransferTransaction.js
import { CertifiedTransferTransaction } from './CertifiedTransferTransaction.js';
import { DataHasher } from '../crypto/hash/DataHasher.js';
import { HashAlgorithm } from '../crypto/hash/HashAlgorithm.js';
import { EncodedPredicate } from '../predicate/EncodedPredicate.js';
import { CborDeserializer } from '../serialization/cbor/CborDeserializer.js';
import { CborError } from '../serialization/cbor/CborError.js';
import { CborSerializer } from '../serialization/cbor/CborSerializer.js';
import { HexConverter } from '../util/HexConverter.js';
import { dedent } from '../util/StringUtils.js';
/**
 * Token transfer transaction.
 */
export class TransferTransaction {
    sourceStateHash;
    lockScript;
    recipient;
    _stateMask;
    _data;
    static CBOR_TAG = 39045n;
    static VERSION = 1n;
    constructor(sourceStateHash, lockScript, recipient, _stateMask, _data) {
        this.sourceStateHash = sourceStateHash;
        this.lockScript = lockScript;
        this.recipient = recipient;
        this._stateMask = _stateMask;
        this._data = _data;
    }
    /**
     * @returns {Uint8Array|null} Copy of the data payload, or `null` if absent.
     */
    get data() {
        return this._data ? new Uint8Array(this._data) : null;
    }
    /**
     * @returns {Uint8Array} Copy of the state mask.
     */
    get stateMask() {
        return new Uint8Array(this._stateMask);
    }
    /**
     * @returns {bigint} Wire-format version of this transaction.
     */
    get version() {
        return TransferTransaction.VERSION;
    }
    /**
     * Create a TransferTransaction for the given token.
     *
     * @param {Token} token Token being transferred (last transaction is used as the source).
     * @param {IPredicate} recipient Predicate that will lock the new state.
     * @param {Uint8Array} stateMask State mask mixed into the new state hash.
     * @param {Uint8Array|null} data Optional data payload.
     * @returns {Promise<TransferTransaction>} New transfer transaction.
     */
    static async create(token, recipient, stateMask, data = null) {
        stateMask = new Uint8Array(stateMask);
        data = data ? new Uint8Array(data) : null;
        const transaction = token.latestTransaction;
        return new TransferTransaction(await transaction.calculateStateHash(), transaction.recipient, EncodedPredicate.fromPredicate(recipient), stateMask, data);
    }
    /**
     * Create TransferTransaction from CBOR bytes.
     *
     * @param {Uint8Array} bytes CBOR bytes.
     * @param {Token} token Token providing context for the transfer transaction.
     * @returns {Promise<TransferTransaction>} Decoded transaction.
     * @throws {CborError} On wrong tag or unsupported version.
     */
    static fromCBOR(bytes, token) {
        const tag = CborDeserializer.decodeTag(bytes);
        if (tag.tag !== TransferTransaction.CBOR_TAG) {
            throw new CborError(`Invalid CBOR tag for TransferTransaction: ${tag.tag}`);
        }
        const data = CborDeserializer.decodeArray(tag.data, 4);
        const version = CborDeserializer.decodeUnsignedInteger(data[0]);
        if (version !== TransferTransaction.VERSION) {
            throw new CborError(`Unsupported TransferTransaction version: ${version}`);
        }
        return TransferTransaction.create(token, EncodedPredicate.fromCBOR(data[1]), CborDeserializer.decodeByteString(data[2]), CborDeserializer.decodeNullable(data[3], CborDeserializer.decodeByteString));
    }
    /**
     * @inheritDoc
     */
    calculateStateHash() {
        return new DataHasher(HashAlgorithm.SHA256)
            .update(CborSerializer.encodeArray(CborSerializer.encodeByteString(this.sourceStateHash.imprint), CborSerializer.encodeByteString(this._stateMask)))
            .digest();
    }
    /**
     * @inheritDoc
     */
    calculateTransactionHash() {
        return new DataHasher(HashAlgorithm.SHA256).update(this.toCBOR()).digest();
    }
    /**
     * @inheritDoc
     */
    toCBOR() {
        return CborSerializer.encodeTag(TransferTransaction.CBOR_TAG, CborSerializer.encodeArray(CborSerializer.encodeUnsignedInteger(this.version), this.recipient.toCBOR(), CborSerializer.encodeByteString(this._stateMask), CborSerializer.encodeNullable(this._data, CborSerializer.encodeByteString)));
    }
    /**
     * Bundle this transaction with its inclusion proof into a CertifiedTransferTransaction.
     *
     * @param {RootTrustBase} trustBase Root trust base used to verify the inclusion certificate.
     * @param {PredicateVerifierService} predicateVerifier Verifier for any embedded predicates.
     * @param {InclusionProof} inclusionProof Inclusion proof for this transaction.
     * @returns {Promise<CertifiedTransferTransaction>} Verified certified transaction.
     */
    toCertifiedTransaction(trustBase, predicateVerifier, inclusionProof) {
        return CertifiedTransferTransaction.fromTransaction(trustBase, predicateVerifier, this, inclusionProof);
    }
    /**
     * @returns {string} String representation of the transaction.
     */
    toString() {
        return dedent `
      TransferTransaction
        Version: ${this.version.toString()}
        Source State Hash: ${this.sourceStateHash.toString()}
        Lock Script: 
          ${this.lockScript.toString()}
        Recipient: ${this.recipient.toString()}
        StateMask: ${HexConverter.encode(this._stateMask)}
        Data: ${this._data ? HexConverter.encode(this._data) : 'null'}`;
    }
}
