// @unicitylabs/state-transition-sdk/lib/crypto/hash/HashAlgorithm.js
import { HashError } from './HashError.js';
/**
 * Enumeration of supported hash algorithms. Each instance carries its
 * protocol id, human-readable name, and output length in bytes.
 */
export class HashAlgorithm {
    id;
    name;
    length;
    static RIPEMD160 = new HashAlgorithm(4, 'RIPEMD-160', 20);
    static SHA224 = new HashAlgorithm(1, 'SHA-224', 28);
    static SHA256 = new HashAlgorithm(0, 'SHA-256', 32);
    static SHA384 = new HashAlgorithm(2, 'SHA-384', 48);
    static SHA512 = new HashAlgorithm(3, 'SHA-512', 64);
    constructor(id, name, length) {
        this.id = id;
        this.name = name;
        this.length = length;
    }
    /**
     * Look up the algorithm by its protocol id.
     *
     * @param {number} id Numeric algorithm id.
     * @returns {HashAlgorithm} Matching algorithm.
     * @throws {HashError} If no algorithm with the given id is registered.
     */
    static fromId(id) {
        switch (id) {
            case HashAlgorithm.SHA256.id:
                return HashAlgorithm.SHA256;
            case HashAlgorithm.SHA224.id:
                return HashAlgorithm.SHA224;
            case HashAlgorithm.SHA384.id:
                return HashAlgorithm.SHA384;
            case HashAlgorithm.SHA512.id:
                return HashAlgorithm.SHA512;
            case HashAlgorithm.RIPEMD160.id:
                return HashAlgorithm.RIPEMD160;
            default:
                throw new HashError(`Unsupported hash algorithm ID: ${id}`);
        }
    }
    /**
     * @returns {string} Human-readable name of the algorithm (e.g. `SHA-256`).
     */
    toString() {
        return this.name;
    }
}
