// @unicitylabs/sphere-sdk/dist/token-engine/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// token-engine/index.ts
var token_engine_exports = {};
__export(token_engine_exports, {
  SpherePaymentData: () => SpherePaymentData,
  TransferConflictError: () => TransferConflictError,
  createSphereTokenEngine: () => createSphereTokenEngine,
  createUnicityIdMinter: () => createUnicityIdMinter,
  decodeSpherePaymentData: () => decodeSpherePaymentData,
  deriveDirectAddress: () => deriveDirectAddress,
  sphereAssetToSdk: () => sphereAssetToSdk
});
module.exports = __toCommonJS(token_engine_exports);

// token-engine/sdk.ts
var import_StateTransitionClient = require("@unicitylabs/state-transition-sdk/lib/StateTransitionClient.js");
var import_AggregatorClient = require("@unicitylabs/state-transition-sdk/lib/api/AggregatorClient.js");
var import_NetworkId = require("@unicitylabs/state-transition-sdk/lib/api/NetworkId.js");
var import_CertificationData = require("@unicitylabs/state-transition-sdk/lib/api/CertificationData.js");
var import_CertificationResponse = require("@unicitylabs/state-transition-sdk/lib/api/CertificationResponse.js");
var import_StateId = require("@unicitylabs/state-transition-sdk/lib/api/StateId.js");
var import_InclusionProof = require("@unicitylabs/state-transition-sdk/lib/api/InclusionProof.js");
var import_InclusionProofResponse = require("@unicitylabs/state-transition-sdk/lib/api/InclusionProofResponse.js");
var import_RootTrustBase = require("@unicitylabs/state-transition-sdk/lib/api/bft/RootTrustBase.js");
var import_InclusionProofUtils = require("@unicitylabs/state-transition-sdk/lib/util/InclusionProofUtils.js");
var import_InclusionProofVerificationRule = require("@unicitylabs/state-transition-sdk/lib/transaction/verification/rule/InclusionProofVerificationRule.js");
var import_Token = require("@unicitylabs/state-transition-sdk/lib/transaction/Token.js");
var import_MintTransaction = require("@unicitylabs/state-transition-sdk/lib/transaction/MintTransaction.js");
var import_TransferTransaction = require("@unicitylabs/state-transition-sdk/lib/transaction/TransferTransaction.js");
var import_CertifiedMintTransaction = require("@unicitylabs/state-transition-sdk/lib/transaction/CertifiedMintTransaction.js");
var import_CertifiedTransferTransaction = require("@unicitylabs/state-transition-sdk/lib/transaction/CertifiedTransferTransaction.js");
var import_TokenId = require("@unicitylabs/state-transition-sdk/lib/transaction/TokenId.js");
var import_TokenType = require("@unicitylabs/state-transition-sdk/lib/transaction/TokenType.js");
var import_TokenSalt = require("@unicitylabs/state-transition-sdk/lib/transaction/TokenSalt.js");
var import_MintJustificationVerifierService = require("@unicitylabs/state-transition-sdk/lib/transaction/verification/MintJustificationVerifierService.js");
var import_EncodedPredicate = require("@unicitylabs/state-transition-sdk/lib/predicate/EncodedPredicate.js");
var import_SignaturePredicate = require("@unicitylabs/state-transition-sdk/lib/predicate/builtin/SignaturePredicate.js");
var import_SignaturePredicateUnlockScript = require("@unicitylabs/state-transition-sdk/lib/predicate/builtin/SignaturePredicateUnlockScript.js");
var import_BurnPredicate = require("@unicitylabs/state-transition-sdk/lib/predicate/builtin/BurnPredicate.js");
var import_PredicateVerifierService = require("@unicitylabs/state-transition-sdk/lib/predicate/verification/PredicateVerifierService.js");
var import_SigningService = require("@unicitylabs/state-transition-sdk/lib/crypto/secp256k1/SigningService.js");
var import_Signature = require("@unicitylabs/state-transition-sdk/lib/crypto/secp256k1/Signature.js");
var import_MintSigningService = require("@unicitylabs/state-transition-sdk/lib/crypto/MintSigningService.js");
var import_HashAlgorithm = require("@unicitylabs/state-transition-sdk/lib/crypto/hash/HashAlgorithm.js");
var import_DataHash = require("@unicitylabs/state-transition-sdk/lib/crypto/hash/DataHash.js");
var import_DataHasher = require("@unicitylabs/state-transition-sdk/lib/crypto/hash/DataHasher.js");
var import_DataHasherFactory = require("@unicitylabs/state-transition-sdk/lib/crypto/hash/DataHasherFactory.js");
var import_CborSerializer = require("@unicitylabs/state-transition-sdk/lib/serialization/cbor/CborSerializer.js");
var import_CborDeserializer = require("@unicitylabs/state-transition-sdk/lib/serialization/cbor/CborDeserializer.js");
var import_CborError = require("@unicitylabs/state-transition-sdk/lib/serialization/cbor/CborError.js");
var import_Asset = require("@unicitylabs/state-transition-sdk/lib/payment/asset/Asset.js");
var import_AssetId = require("@unicitylabs/state-transition-sdk/lib/payment/asset/AssetId.js");
var import_PaymentAssetCollection = require("@unicitylabs/state-transition-sdk/lib/payment/asset/PaymentAssetCollection.js");
var import_TokenSplit = require("@unicitylabs/state-transition-sdk/lib/payment/TokenSplit.js");
var import_SplitTokenRequest = require("@unicitylabs/state-transition-sdk/lib/payment/SplitTokenRequest.js");
var import_SplitToken = require("@unicitylabs/state-transition-sdk/lib/payment/SplitToken.js");
var import_SplitAssetProof = require("@unicitylabs/state-transition-sdk/lib/payment/SplitAssetProof.js");
var import_SplitMintJustification = require("@unicitylabs/state-transition-sdk/lib/payment/SplitMintJustification.js");
var import_SplitMintJustificationVerifier = require("@unicitylabs/state-transition-sdk/lib/payment/SplitMintJustificationVerifier.js");
var import_VerificationStatus = require("@unicitylabs/state-transition-sdk/lib/verification/VerificationStatus.js");
var import_VerificationResult = require("@unicitylabs/state-transition-sdk/lib/verification/VerificationResult.js");
var import_UnicityId = require("@unicitylabs/state-transition-sdk/lib/unicity-id/UnicityId.js");
var import_UnicityIdMintTransaction = require("@unicitylabs/state-transition-sdk/lib/unicity-id/UnicityIdMintTransaction.js");
var import_CertifiedUnicityIdMintTransaction = require("@unicitylabs/state-transition-sdk/lib/unicity-id/CertifiedUnicityIdMintTransaction.js");
var import_UnicityIdToken = require("@unicitylabs/state-transition-sdk/lib/unicity-id/UnicityIdToken.js");
var import_HexConverter = require("@unicitylabs/state-transition-sdk/lib/util/HexConverter.js");
var import_BigintConverter = require("@unicitylabs/state-transition-sdk/lib/util/BigintConverter.js");
var import_BitString = require("@unicitylabs/state-transition-sdk/lib/util/BitString.js");

// token-engine/identity.ts
var UNICITY_TOKEN_TYPE_HEX = "f8aa13834268d29355ff12183066f0cb902003629bbc5eb9ef0efbe397867509";
var SIGNING_ALGORITHM = "secp256k1";
var EMBEDDED_PREDICATE_UNMASKED = 0;
var HASH_ALGORITHM_SHA256 = 0n;
var SHA256_IMPRINT_PREFIX = new Uint8Array([0, 0]);
function hexToBytes(hex) {
  const bytes = new Uint8Array(hex.length / 2);
  for (let i = 0; i < bytes.length; i++) {
    bytes[i] = parseInt(hex.slice(i * 2, i * 2 + 2), 16);
  }
  return bytes;
}
function sha256(data) {
  return new import_DataHasher.DataHasher(import_HashAlgorithm.HashAlgorithm.SHA256).update(data).digest().then((h) => h.data);
}
async function deriveDirectAddress(publicKey) {
  const tokenTypeCbor = import_CborSerializer.CborSerializer.encodeByteString(hexToBytes(UNICITY_TOKEN_TYPE_HEX));
  const reference = import_CborSerializer.CborSerializer.encodeArray(
    import_CborSerializer.CborSerializer.encodeByteString(new Uint8Array([EMBEDDED_PREDICATE_UNMASKED])),
    import_CborSerializer.CborSerializer.encodeByteString(tokenTypeCbor),
    import_CborSerializer.CborSerializer.encodeTextString(SIGNING_ALGORITHM),
    import_CborSerializer.CborSerializer.encodeUnsignedInteger(HASH_ALGORITHM_SHA256),
    import_CborSerializer.CborSerializer.encodeByteString(publicKey)
  );
  const refHash = await sha256(reference);
  const imprint = new Uint8Array([...SHA256_IMPRINT_PREFIX, ...refHash]);
  const checksum = (await sha256(imprint)).slice(0, 4);
  return `DIRECT://${import_HexConverter.HexConverter.encode(imprint)}${import_HexConverter.HexConverter.encode(checksum)}`;
}

// core/errors.ts
var SphereError = class extends Error {
  code;
  cause;
  constructor(message, code, cause) {
    super(message);
    this.name = "SphereError";
    this.code = code;
    this.cause = cause;
  }
};

// core/logger.ts
var LOGGER_KEY = "__sphere_sdk_logger__";
function getState() {
  const g = globalThis;
  if (!g[LOGGER_KEY]) {
    g[LOGGER_KEY] = { debug: false, tags: {}, handler: null };
  }
  return g[LOGGER_KEY];
}
function isEnabled(tag) {
  const state = getState();
  if (tag in state.tags) return state.tags[tag];
  return state.debug;
}
var logger = {
  /**
   * Configure the logger. Can be called multiple times (last write wins).
   * Typically called by createBrowserProviders(), createNodeProviders(), or Sphere.init().
   */
  configure(config) {
    const state = getState();
    if (config.debug !== void 0) state.debug = config.debug;
    if (config.handler !== void 0) state.handler = config.handler;
  },
  /**
   * Enable/disable debug logging for a specific tag.
   * Per-tag setting overrides the global debug flag.
   *
   * @example
   * ```ts
   * logger.setTagDebug('Nostr', true);  // enable only Nostr logs
   * logger.setTagDebug('Nostr', false); // disable Nostr logs even if global debug=true
   * ```
   */
  setTagDebug(tag, enabled) {
    getState().tags[tag] = enabled;
  },
  /**
   * Clear per-tag override, falling back to global debug flag.
   */
  clearTagDebug(tag) {
    delete getState().tags[tag];
  },
  /** Returns true if debug mode is enabled for the given tag (or globally). */
  isDebugEnabled(tag) {
    if (tag) return isEnabled(tag);
    return getState().debug;
  },
  /**
   * Debug-level log. Only shown when debug is enabled (globally or for this tag).
   * Use for detailed operational information.
   */
  debug(tag, message, ...args) {
    if (!isEnabled(tag)) return;
    const state = getState();
    if (state.handler) {
      state.handler("debug", tag, message, ...args);
    } else {
      console.log(`[${tag}]`, message, ...args);
    }
  },
  /**
   * Warning-level log. ALWAYS shown regardless of debug flag.
   * Use for important but non-critical issues (timeouts, retries, degraded state).
   */
  warn(tag, message, ...args) {
    const state = getState();
    if (state.handler) {
      state.handler("warn", tag, message, ...args);
    } else {
      console.warn(`[${tag}]`, message, ...args);
    }
  },
  /**
   * Error-level log. ALWAYS shown regardless of debug flag.
   * Use for critical failures that should never be silenced.
   */
  error(tag, message, ...args) {
    const state = getState();
    if (state.handler) {
      state.handler("error", tag, message, ...args);
    } else {
      console.error(`[${tag}]`, message, ...args);
    }
  },
  /** Reset all logger state (debug flag, tags, handler). Primarily for tests. */
  reset() {
    const g = globalThis;
    delete g[LOGGER_KEY];
  }
};

// token-engine/SpherePaymentData.ts
var COIN_ID_PATTERN = /^([0-9a-f]{2})+$/;
function assertAsset(coinId, amount) {
  if (!COIN_ID_PATTERN.test(coinId)) {
    throw new SphereError(`Invalid coin id (expected even-length lowercase hex): "${coinId}"`, "VALIDATION_ERROR");
  }
  if (amount < 0n) {
    throw new SphereError(`Asset amount must be non-negative: ${amount.toString()}`, "VALIDATION_ERROR");
  }
}
function sphereAssetToSdk(coinId, amount) {
  assertAsset(coinId, amount);
  return new import_Asset.Asset(new import_AssetId.AssetId(import_HexConverter.HexConverter.decode(coinId)), amount);
}
var SpherePaymentData = class _SpherePaymentData {
  constructor(assets, _memo = null) {
    this.assets = assets;
    this._memo = _memo;
  }
  /** Sphere-private CBOR tag (verified free in the v2 SDK tag space). */
  static CBOR_TAG = 39050n;
  /** Envelope version; bump when the structure changes. */
  static VERSION = 1n;
  /** Opaque, app-defined memo carried alongside the value (e.g. invoice attribution). */
  get memo() {
    return this._memo ? new Uint8Array(this._memo) : null;
  }
  /** Wrap an existing SDK asset collection (+ optional opaque memo). */
  static create(assets, memo = null) {
    return new _SpherePaymentData(assets, memo);
  }
  /** Build from a sphere-domain value (hex coin id → bigint amount) + optional opaque memo. */
  static fromValue(value, memo = null) {
    const assets = value.assets.map((a) => sphereAssetToSdk(a.coinId, a.amount));
    return new _SpherePaymentData(import_PaymentAssetCollection.PaymentAssetCollection.create(...assets), memo);
  }
  /** Decode from the CBOR envelope produced by {@link encode}. */
  static fromCBOR(bytes) {
    const tag = import_CborDeserializer.CborDeserializer.decodeTag(bytes);
    if (tag.tag !== _SpherePaymentData.CBOR_TAG) {
      throw new import_CborError.CborError(`Invalid SpherePaymentData tag: ${tag.tag}`);
    }
    const fields = import_CborDeserializer.CborDeserializer.decodeArray(tag.data, 3);
    const version = import_CborDeserializer.CborDeserializer.decodeUnsignedInteger(fields[0]);
    if (version !== _SpherePaymentData.VERSION) {
      throw new import_CborError.CborError(`Unsupported SpherePaymentData version: ${version}`);
    }
    const memo = import_CborDeserializer.CborDeserializer.decodeNullable(fields[2], import_CborDeserializer.CborDeserializer.decodeByteString);
    return new _SpherePaymentData(import_PaymentAssetCollection.PaymentAssetCollection.fromCBOR(fields[1]), memo);
  }
  /** Deterministic, versioned, tagged CBOR: `tag(39050)[ version, assets, memo? ]`. */
  encode() {
    return Promise.resolve(
      import_CborSerializer.CborSerializer.encodeTag(
        _SpherePaymentData.CBOR_TAG,
        import_CborSerializer.CborSerializer.encodeArray(
          import_CborSerializer.CborSerializer.encodeUnsignedInteger(_SpherePaymentData.VERSION),
          this.assets.toCBOR(),
          import_CborSerializer.CborSerializer.encodeNullable(this._memo, import_CborSerializer.CborSerializer.encodeByteString)
        )
      )
    );
  }
  /** Project to a sphere-domain value (hex coin id + bigint amount), preserving order. */
  toValue() {
    return {
      assets: this.assets.toArray().map((a) => ({
        coinId: import_HexConverter.HexConverter.encode(a.id.bytes),
        amount: a.value
      }))
    };
  }
  /** Balance of a single coin within this payload (0n when absent). */
  balanceOf(coinId) {
    if (!COIN_ID_PATTERN.test(coinId)) {
      throw new SphereError(`Invalid coin id (expected even-length lowercase hex): "${coinId}"`, "VALIDATION_ERROR");
    }
    const asset = this.assets.get(new import_AssetId.AssetId(import_HexConverter.HexConverter.decode(coinId)));
    return asset ? asset.value : 0n;
  }
};
function decodeSpherePaymentData(bytes) {
  return Promise.resolve(SpherePaymentData.fromCBOR(bytes));
}

// core/uuid.ts
function randomUUID() {
  const c = globalThis.crypto;
  if (typeof c !== "undefined" && typeof c.randomUUID === "function") {
    return c.randomUUID();
  }
  const b = c.getRandomValues(new Uint8Array(16));
  b[6] = b[6] & 15 | 64;
  b[8] = b[8] & 63 | 128;
  const h = Array.from(b, (x) => x.toString(16).padStart(2, "0"));
  return `${h[0]}${h[1]}${h[2]}${h[3]}-${h[4]}${h[5]}-${h[6]}${h[7]}-${h[8]}${h[9]}-${h[10]}${h[11]}${h[12]}${h[13]}${h[14]}${h[15]}`;
}

// token-engine/errors.ts
var TransferConflictError = class extends SphereError {
  /**
   * #625: the source token id (the caller's id, e.g. `v2_<genesis>`) whose state was already spent.
   * Set by PaymentsModule at the engine call site so the self-healing retry can demote it and re-plan.
   */
  conflictedSourceId;
  constructor(message, cause) {
    super(message, "TRANSFER_CONFLICT", cause);
    this.name = "TransferConflictError";
  }
};

// token-engine/token-blob.ts
var TOKEN_BLOB_TAG = 39051n;
var TOKEN_BLOB_VERSION = 1;
function decodeTokenBlob(bytes) {
  const tag = import_CborDeserializer.CborDeserializer.decodeTag(bytes);
  if (tag.tag !== TOKEN_BLOB_TAG) {
    throw new import_CborError.CborError(`Invalid TokenBlob tag: ${tag.tag}`);
  }
  const fields = import_CborDeserializer.CborDeserializer.decodeArray(tag.data, 4);
  const v = Number(import_CborDeserializer.CborDeserializer.decodeUnsignedInteger(fields[0]));
  if (v !== TOKEN_BLOB_VERSION) {
    throw new import_CborError.CborError(`Unsupported TokenBlob version: ${v}`);
  }
  return {
    v,
    network: Number(import_CborDeserializer.CborDeserializer.decodeUnsignedInteger(fields[1])),
    tokenId: import_CborDeserializer.CborDeserializer.decodeTextString(fields[2]),
    token: import_CborDeserializer.CborDeserializer.decodeByteString(fields[3])
  };
}
function unwrapTokenBlobBytes(bytes) {
  try {
    return decodeTokenBlob(bytes).token;
  } catch {
    return bytes;
  }
}

// token-engine/blob-keys.ts
async function deriveDeliveryKeys(blobBytes) {
  const token = await import_Token.Token.fromCBOR(unwrapTokenBlobBytes(blobBytes));
  const stateHash = import_HexConverter.HexConverter.encode((await token.latestTransaction.calculateStateHash()).imprint);
  return { tokenId: import_HexConverter.HexConverter.encode(token.id.bytes), stateHash };
}

// node_modules/@noble/hashes/utils.js
function isBytes(a) {
  return a instanceof Uint8Array || ArrayBuffer.isView(a) && a.constructor.name === "Uint8Array" && "BYTES_PER_ELEMENT" in a && a.BYTES_PER_ELEMENT === 1;
}
function anumber(n, title = "") {
  if (typeof n !== "number") {
    const prefix = title && `"${title}" `;
    throw new TypeError(`${prefix}expected number, got ${typeof n}`);
  }
  if (!Number.isSafeInteger(n) || n < 0) {
    const prefix = title && `"${title}" `;
    throw new RangeError(`${prefix}expected integer >= 0, got ${n}`);
  }
}
function abytes(value, length, title = "") {
  const bytes = isBytes(value);
  const len = value?.length;
  const needsLen = length !== void 0;
  if (!bytes || needsLen && len !== length) {
    const prefix = title && `"${title}" `;
    const ofLen = needsLen ? ` of length ${length}` : "";
    const got = bytes ? `length=${len}` : `type=${typeof value}`;
    const message = prefix + "expected Uint8Array" + ofLen + ", got " + got;
    if (!bytes)
      throw new TypeError(message);
    throw new RangeError(message);
  }
  return value;
}
function ahash(h) {
  if (typeof h !== "function" || typeof h.create !== "function")
    throw new TypeError("Hash must wrapped by utils.createHasher");
  anumber(h.outputLen);
  anumber(h.blockLen);
  if (h.outputLen < 1)
    throw new Error('"outputLen" must be >= 1');
  if (h.blockLen < 1)
    throw new Error('"blockLen" must be >= 1');
}
function aexists(instance, checkFinished = true) {
  if (instance.destroyed)
    throw new Error("Hash instance has been destroyed");
  if (checkFinished && instance.finished)
    throw new Error("Hash#digest() has already been called");
}
function aoutput(out, instance) {
  abytes(out, void 0, "digestInto() output");
  const min = instance.outputLen;
  if (out.length < min) {
    throw new RangeError('"digestInto() output" expected to be of length >=' + min);
  }
}
function clean(...arrays) {
  for (let i = 0; i < arrays.length; i++) {
    arrays[i].fill(0);
  }
}
function createView(arr) {
  return new DataView(arr.buffer, arr.byteOffset, arr.byteLength);
}
function rotr(word, shift) {
  return word << 32 - shift | word >>> shift;
}
function createHasher(hashCons, info = {}) {
  const hashC = (msg, opts) => hashCons(opts).update(msg).digest();
  const tmp = hashCons(void 0);
  hashC.outputLen = tmp.outputLen;
  hashC.blockLen = tmp.blockLen;
  hashC.canXOF = tmp.canXOF;
  hashC.create = (opts) => hashCons(opts);
  Object.assign(hashC, info);
  return Object.freeze(hashC);
}
var oidNist = (suffix) => ({
  // Current NIST hashAlgs suffixes used here fit in one DER subidentifier octet.
  // Larger suffix values would need base-128 OID encoding and a different length byte.
  oid: Uint8Array.from([6, 9, 96, 134, 72, 1, 101, 3, 4, 2, suffix])
});

// node_modules/@noble/hashes/hmac.js
var _HMAC = class {
  oHash;
  iHash;
  blockLen;
  outputLen;
  canXOF = false;
  finished = false;
  destroyed = false;
  constructor(hash, key) {
    ahash(hash);
    abytes(key, void 0, "key");
    this.iHash = hash.create();
    if (typeof this.iHash.update !== "function")
      throw new Error("Expected instance of class which extends utils.Hash");
    this.blockLen = this.iHash.blockLen;
    this.outputLen = this.iHash.outputLen;
    const blockLen = this.blockLen;
    const pad = new Uint8Array(blockLen);
    pad.set(key.length > blockLen ? hash.create().update(key).digest() : key);
    for (let i = 0; i < pad.length; i++)
      pad[i] ^= 54;
    this.iHash.update(pad);
    this.oHash = hash.create();
    for (let i = 0; i < pad.length; i++)
      pad[i] ^= 54 ^ 92;
    this.oHash.update(pad);
    clean(pad);
  }
  update(buf) {
    aexists(this);
    this.iHash.update(buf);
    return this;
  }
  digestInto(out) {
    aexists(this);
    aoutput(out, this);
    this.finished = true;
    const buf = out.subarray(0, this.outputLen);
    this.iHash.digestInto(buf);
    this.oHash.update(buf);
    this.oHash.digestInto(buf);
    this.destroy();
  }
  digest() {
    const out = new Uint8Array(this.oHash.outputLen);
    this.digestInto(out);
    return out;
  }
  _cloneInto(to) {
    to ||= Object.create(Object.getPrototypeOf(this), {});
    const { oHash, iHash, finished, destroyed, blockLen, outputLen } = this;
    to = to;
    to.finished = finished;
    to.destroyed = destroyed;
    to.blockLen = blockLen;
    to.outputLen = outputLen;
    to.oHash = oHash._cloneInto(to.oHash);
    to.iHash = iHash._cloneInto(to.iHash);
    return to;
  }
  clone() {
    return this._cloneInto();
  }
  destroy() {
    this.destroyed = true;
    this.oHash.destroy();
    this.iHash.destroy();
  }
};
var hmac = /* @__PURE__ */ (() => {
  const hmac_ = ((hash, key, message) => new _HMAC(hash, key).update(message).digest());
  hmac_.create = (hash, key) => new _HMAC(hash, key);
  return hmac_;
})();

// node_modules/@noble/hashes/hkdf.js
function extract(hash, ikm, salt) {
  ahash(hash);
  if (salt === void 0)
    salt = new Uint8Array(hash.outputLen);
  return hmac(hash, salt, ikm);
}
var HKDF_COUNTER = /* @__PURE__ */ Uint8Array.of(0);
var EMPTY_BUFFER = /* @__PURE__ */ Uint8Array.of();
function expand(hash, prk, info, length = 32) {
  ahash(hash);
  anumber(length, "length");
  abytes(prk, void 0, "prk");
  const olen = hash.outputLen;
  if (prk.length < olen)
    throw new Error('"prk" must be at least HashLen octets');
  if (length > 255 * olen)
    throw new Error("Length must be <= 255*HashLen");
  const blocks = Math.ceil(length / olen);
  if (info === void 0)
    info = EMPTY_BUFFER;
  else
    abytes(info, void 0, "info");
  const okm = new Uint8Array(blocks * olen);
  const HMAC = hmac.create(hash, prk);
  const HMACTmp = HMAC._cloneInto();
  const T = new Uint8Array(HMAC.outputLen);
  for (let counter = 0; counter < blocks; counter++) {
    HKDF_COUNTER[0] = counter + 1;
    HMACTmp.update(counter === 0 ? EMPTY_BUFFER : T).update(info).update(HKDF_COUNTER).digestInto(T);
    okm.set(T, olen * counter);
    HMAC._cloneInto(HMACTmp);
  }
  HMAC.destroy();
  HMACTmp.destroy();
  clean(T, HKDF_COUNTER);
  return okm.slice(0, length);
}
var hkdf = (hash, ikm, salt, info, length) => expand(hash, extract(hash, ikm, salt), info, length);

// node_modules/@noble/hashes/_md.js
function Chi(a, b, c) {
  return a & b ^ ~a & c;
}
function Maj(a, b, c) {
  return a & b ^ a & c ^ b & c;
}
var HashMD = class {
  blockLen;
  outputLen;
  canXOF = false;
  padOffset;
  isLE;
  // For partial updates less than block size
  buffer;
  view;
  finished = false;
  length = 0;
  pos = 0;
  destroyed = false;
  constructor(blockLen, outputLen, padOffset, isLE) {
    this.blockLen = blockLen;
    this.outputLen = outputLen;
    this.padOffset = padOffset;
    this.isLE = isLE;
    this.buffer = new Uint8Array(blockLen);
    this.view = createView(this.buffer);
  }
  update(data) {
    aexists(this);
    abytes(data);
    const { view, buffer, blockLen } = this;
    const len = data.length;
    for (let pos = 0; pos < len; ) {
      const take = Math.min(blockLen - this.pos, len - pos);
      if (take === blockLen) {
        const dataView = createView(data);
        for (; blockLen <= len - pos; pos += blockLen)
          this.process(dataView, pos);
        continue;
      }
      buffer.set(data.subarray(pos, pos + take), this.pos);
      this.pos += take;
      pos += take;
      if (this.pos === blockLen) {
        this.process(view, 0);
        this.pos = 0;
      }
    }
    this.length += data.length;
    this.roundClean();
    return this;
  }
  digestInto(out) {
    aexists(this);
    aoutput(out, this);
    this.finished = true;
    const { buffer, view, blockLen, isLE } = this;
    let { pos } = this;
    buffer[pos++] = 128;
    clean(this.buffer.subarray(pos));
    if (this.padOffset > blockLen - pos) {
      this.process(view, 0);
      pos = 0;
    }
    for (let i = pos; i < blockLen; i++)
      buffer[i] = 0;
    view.setBigUint64(blockLen - 8, BigInt(this.length * 8), isLE);
    this.process(view, 0);
    const oview = createView(out);
    const len = this.outputLen;
    if (len % 4)
      throw new Error("_sha2: outputLen must be aligned to 32bit");
    const outLen = len / 4;
    const state = this.get();
    if (outLen > state.length)
      throw new Error("_sha2: outputLen bigger than state");
    for (let i = 0; i < outLen; i++)
      oview.setUint32(4 * i, state[i], isLE);
  }
  digest() {
    const { buffer, outputLen } = this;
    this.digestInto(buffer);
    const res = buffer.slice(0, outputLen);
    this.destroy();
    return res;
  }
  _cloneInto(to) {
    to ||= new this.constructor();
    to.set(...this.get());
    const { blockLen, buffer, length, finished, destroyed, pos } = this;
    to.destroyed = destroyed;
    to.finished = finished;
    to.length = length;
    to.pos = pos;
    if (length % blockLen)
      to.buffer.set(buffer);
    return to;
  }
  clone() {
    return this._cloneInto();
  }
};
var SHA256_IV = /* @__PURE__ */ Uint32Array.from([
  1779033703,
  3144134277,
  1013904242,
  2773480762,
  1359893119,
  2600822924,
  528734635,
  1541459225
]);

// node_modules/@noble/hashes/sha2.js
var SHA256_K = /* @__PURE__ */ Uint32Array.from([
  1116352408,
  1899447441,
  3049323471,
  3921009573,
  961987163,
  1508970993,
  2453635748,
  2870763221,
  3624381080,
  310598401,
  607225278,
  1426881987,
  1925078388,
  2162078206,
  2614888103,
  3248222580,
  3835390401,
  4022224774,
  264347078,
  604807628,
  770255983,
  1249150122,
  1555081692,
  1996064986,
  2554220882,
  2821834349,
  2952996808,
  3210313671,
  3336571891,
  3584528711,
  113926993,
  338241895,
  666307205,
  773529912,
  1294757372,
  1396182291,
  1695183700,
  1986661051,
  2177026350,
  2456956037,
  2730485921,
  2820302411,
  3259730800,
  3345764771,
  3516065817,
  3600352804,
  4094571909,
  275423344,
  430227734,
  506948616,
  659060556,
  883997877,
  958139571,
  1322822218,
  1537002063,
  1747873779,
  1955562222,
  2024104815,
  2227730452,
  2361852424,
  2428436474,
  2756734187,
  3204031479,
  3329325298
]);
var SHA256_W = /* @__PURE__ */ new Uint32Array(64);
var SHA2_32B = class extends HashMD {
  constructor(outputLen) {
    super(64, outputLen, 8, false);
  }
  get() {
    const { A, B, C, D, E, F, G, H } = this;
    return [A, B, C, D, E, F, G, H];
  }
  // prettier-ignore
  set(A, B, C, D, E, F, G, H) {
    this.A = A | 0;
    this.B = B | 0;
    this.C = C | 0;
    this.D = D | 0;
    this.E = E | 0;
    this.F = F | 0;
    this.G = G | 0;
    this.H = H | 0;
  }
  process(view, offset) {
    for (let i = 0; i < 16; i++, offset += 4)
      SHA256_W[i] = view.getUint32(offset, false);
    for (let i = 16; i < 64; i++) {
      const W15 = SHA256_W[i - 15];
      const W2 = SHA256_W[i - 2];
      const s0 = rotr(W15, 7) ^ rotr(W15, 18) ^ W15 >>> 3;
      const s1 = rotr(W2, 17) ^ rotr(W2, 19) ^ W2 >>> 10;
      SHA256_W[i] = s1 + SHA256_W[i - 7] + s0 + SHA256_W[i - 16] | 0;
    }
    let { A, B, C, D, E, F, G, H } = this;
    for (let i = 0; i < 64; i++) {
      const sigma1 = rotr(E, 6) ^ rotr(E, 11) ^ rotr(E, 25);
      const T1 = H + sigma1 + Chi(E, F, G) + SHA256_K[i] + SHA256_W[i] | 0;
      const sigma0 = rotr(A, 2) ^ rotr(A, 13) ^ rotr(A, 22);
      const T2 = sigma0 + Maj(A, B, C) | 0;
      H = G;
      G = F;
      F = E;
      E = D + T1 | 0;
      D = C;
      C = B;
      B = A;
      A = T1 + T2 | 0;
    }
    A = A + this.A | 0;
    B = B + this.B | 0;
    C = C + this.C | 0;
    D = D + this.D | 0;
    E = E + this.E | 0;
    F = F + this.F | 0;
    G = G + this.G | 0;
    H = H + this.H | 0;
    this.set(A, B, C, D, E, F, G, H);
  }
  roundClean() {
    clean(SHA256_W);
  }
  destroy() {
    this.destroyed = true;
    this.set(0, 0, 0, 0, 0, 0, 0, 0);
    clean(this.buffer);
  }
};
var _SHA256 = class extends SHA2_32B {
  // We cannot use array here since array allows indexing by variable
  // which means optimizer/compiler cannot use registers.
  A = SHA256_IV[0] | 0;
  B = SHA256_IV[1] | 0;
  C = SHA256_IV[2] | 0;
  D = SHA256_IV[3] | 0;
  E = SHA256_IV[4] | 0;
  F = SHA256_IV[5] | 0;
  G = SHA256_IV[6] | 0;
  H = SHA256_IV[7] | 0;
  constructor() {
    super(32);
  }
};
var sha2562 = /* @__PURE__ */ createHasher(
  () => new _SHA256(),
  /* @__PURE__ */ oidNist(1)
);

// token-engine/realization.ts
var REALIZATION_HKDF_INFO_PREFIX = "sphere-realization-v1";
function deriveRealization(privKeyHex, transferId, index, field) {
  const ikm = import_HexConverter.HexConverter.decode(privKeyHex);
  const info = new TextEncoder().encode(`${REALIZATION_HKDF_INFO_PREFIX}:${transferId}:${field}:${index}`);
  return hkdf(sha2562, ikm, void 0, info, 32);
}

// token-engine/SphereTokenEngine.ts
var TRANSFER_ID_RE = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/;
var SphereTokenEngine = class {
  constructor(deps) {
    this.deps = deps;
    this.privateKeyHex = import_HexConverter.HexConverter.encode(deps.privateKey);
  }
  /** Hex form of the wallet key — deriveRealization's ikm input (Part E.1). */
  privateKeyHex;
  // ── identity ────────────────────────────────────────────────────────────────
  getIdentity() {
    return { chainPubkey: new Uint8Array(this.deps.signingService.publicKey) };
  }
  /** Legacy DIRECT:// address (Path A). Async: the derivation hashes via the SDK. */
  deriveIdentityAddress(pubkey) {
    return deriveDirectAddress(pubkey ?? this.deps.signingService.publicKey);
  }
  // ── value (read) ─────────────────────────────────────────────────────────────
  readValue(token) {
    return token.value;
  }
  balanceOf(token, coinId) {
    let sum = 0n;
    for (const asset of token.value?.assets ?? []) {
      if (asset.coinId === coinId) sum += asset.amount;
    }
    return sum;
  }
  tokenId(token) {
    return token.blob.tokenId;
  }
  readMemo(token) {
    const sdkToken = token.sdkToken;
    if (sdkToken.transactions.length > 0) {
      return sdkToken.latestTransaction.data;
    }
    const data = sdkToken.genesis.data;
    if (data && this.isSpherePaymentData(data)) {
      return SpherePaymentData.fromCBOR(data).memo;
    }
    return null;
  }
  readTokenData(token) {
    const data = token.sdkToken.genesis.data;
    return data ? new Uint8Array(data) : null;
  }
  // ── lifecycle ────────────────────────────────────────────────────────────────
  async mint(params, options) {
    const recipient = import_SignaturePredicate.SignaturePredicate.create(params.recipientPubkey);
    const data = params.value ? await SpherePaymentData.fromValue(params.value).encode() : null;
    const mintTx = await import_MintTransaction.MintTransaction.create(this.deps.networkId, recipient, data);
    const certificationData = await import_CertificationData.CertificationData.fromMintTransaction(mintTx);
    const response = await this.deps.client.submitCertificationRequest(certificationData);
    if (response.status !== import_CertificationResponse.CertificationStatus.SUCCESS) {
      throw new SphereError(`Mint certification failed: ${response.status}`, "AGGREGATOR_ERROR");
    }
    const proof = await (0, import_InclusionProofUtils.waitInclusionProof)(
      this.deps.client,
      this.deps.trustBase,
      this.deps.predicateVerifier,
      mintTx,
      options?.signal
    );
    const certified = await mintTx.toCertifiedTransaction(this.deps.trustBase, this.deps.predicateVerifier, proof);
    const token = await import_Token.Token.mint(
      this.deps.trustBase,
      this.deps.predicateVerifier,
      this.deps.mintJustificationVerifier,
      certified
    );
    return this.wrapToken(token);
  }
  async mintDataToken(params, options) {
    const recipient = import_SignaturePredicate.SignaturePredicate.create(params.recipientPubkey);
    const tokenType = params.tokenType ? new import_TokenType.TokenType(params.tokenType) : import_TokenType.TokenType.generate();
    const salt = params.salt ? import_TokenSalt.TokenSalt.fromBytes(params.salt) : import_TokenSalt.TokenSalt.generate();
    const mintTx = await import_MintTransaction.MintTransaction.create(this.deps.networkId, recipient, params.data, tokenType, salt);
    const certificationData = await import_CertificationData.CertificationData.fromMintTransaction(mintTx);
    const response = await this.deps.client.submitCertificationRequest(certificationData);
    if (response.status !== import_CertificationResponse.CertificationStatus.SUCCESS) {
      throw new SphereError(`Data-token mint failed: ${response.status}`, "AGGREGATOR_ERROR");
    }
    const proof = await (0, import_InclusionProofUtils.waitInclusionProof)(
      this.deps.client,
      this.deps.trustBase,
      this.deps.predicateVerifier,
      mintTx,
      options?.signal
    );
    const certified = await mintTx.toCertifiedTransaction(this.deps.trustBase, this.deps.predicateVerifier, proof);
    const token = await import_Token.Token.mint(
      this.deps.trustBase,
      this.deps.predicateVerifier,
      this.deps.mintJustificationVerifier,
      certified
    );
    return this.wrapToken(token);
  }
  async transfer(params, options) {
    this.assertOwned(params.token);
    const transferId = this.resolveTransferId(options);
    const recipient = import_SignaturePredicate.SignaturePredicate.create(params.recipientPubkey);
    const stateMask = deriveRealization(this.privateKeyHex, transferId, this.resolveOpIndex(options), "stateMask");
    const transferTx = await import_TransferTransaction.TransferTransaction.create(params.token.sdkToken, recipient, stateMask, params.data ?? null);
    const unlockScript = await import_SignaturePredicateUnlockScript.SignaturePredicateUnlockScript.create(transferTx, this.deps.signingService);
    const certificationData = await import_CertificationData.CertificationData.fromTransaction(transferTx, unlockScript);
    const proof = await this.submitAndAwaitProof(
      certificationData,
      transferTx,
      "Transfer certification failed",
      "TRANSFER_FAILED",
      options
    );
    const certified = await transferTx.toCertifiedTransaction(this.deps.trustBase, this.deps.predicateVerifier, proof);
    const transferred = await params.token.sdkToken.transfer(this.deps.trustBase, this.deps.predicateVerifier, certified);
    return this.wrapToken(transferred);
  }
  async split(params, options) {
    this.assertOwned(params.token);
    if (params.outputs.length === 0) {
      throw new SphereError("Split requires at least one output", "VALIDATION_ERROR");
    }
    const transferId = this.resolveTransferId(options);
    const tokenType = new import_TokenType.TokenType(import_HexConverter.HexConverter.decode(UNICITY_TOKEN_TYPE_HEX));
    const requests = params.outputs.map(
      (o, i) => import_SplitTokenRequest.SplitTokenRequest.create(
        import_SignaturePredicate.SignaturePredicate.create(o.recipientPubkey),
        import_PaymentAssetCollection.PaymentAssetCollection.create(sphereAssetToSdk(o.coinId, o.amount)),
        tokenType,
        import_TokenSalt.TokenSalt.fromBytes(deriveRealization(this.privateKeyHex, transferId, i, "salt"))
      )
    );
    const split = await import_TokenSplit.TokenSplit.split(
      params.token.sdkToken,
      decodeSpherePaymentData,
      requests,
      deriveRealization(this.privateKeyHex, transferId, this.resolveOpIndex(options), "burn")
    );
    const burnUnlock = await import_SignaturePredicateUnlockScript.SignaturePredicateUnlockScript.create(split.burn.transaction, this.deps.signingService);
    const burnCert = await import_CertificationData.CertificationData.fromTransaction(split.burn.transaction, burnUnlock);
    const burnProof = await this.submitAndAwaitProof(
      burnCert,
      split.burn.transaction,
      "Split burn failed",
      "TRANSFER_FAILED",
      options
    );
    const burnCertified = await split.burn.transaction.toCertifiedTransaction(
      this.deps.trustBase,
      this.deps.predicateVerifier,
      burnProof
    );
    const burntToken = await params.token.sdkToken.transfer(
      this.deps.trustBase,
      this.deps.predicateVerifier,
      burnCertified
    );
    const outputs = [];
    for (let i = 0; i < split.tokens.length; i++) {
      const splitToken = split.tokens[i];
      const data = await SpherePaymentData.create(splitToken.assets, params.outputs[i].data ?? null).encode();
      const justification = import_SplitMintJustification.SplitMintJustification.create(burntToken, splitToken.proofs).toCBOR();
      const mintTx = await import_MintTransaction.MintTransaction.create(
        splitToken.networkId,
        splitToken.recipient,
        data,
        splitToken.tokenType,
        splitToken.salt,
        justification
      );
      const certData = await import_CertificationData.CertificationData.fromMintTransaction(mintTx);
      const proof = await this.submitAndAwaitProof(certData, mintTx, "Split mint failed", "AGGREGATOR_ERROR", options);
      const certified = await mintTx.toCertifiedTransaction(this.deps.trustBase, this.deps.predicateVerifier, proof);
      const token = await import_Token.Token.mint(
        this.deps.trustBase,
        this.deps.predicateVerifier,
        this.deps.mintJustificationVerifier,
        certified
      );
      outputs.push(this.wrapToken(token));
    }
    return { outputs };
  }
  // ── verification ─────────────────────────────────────────────────────────────
  async verify(token, _options) {
    const result = await token.sdkToken.verify(
      this.deps.trustBase,
      this.deps.predicateVerifier,
      this.deps.mintJustificationVerifier
    );
    return result.status === import_VerificationStatus.VerificationStatus.OK ? { ok: true } : { ok: false, reason: String(result.status) };
  }
  isOwnedBy(token, pubkey) {
    const owner = token.sdkToken.latestTransaction.recipient;
    const claimed = import_EncodedPredicate.EncodedPredicate.fromPredicate(import_SignaturePredicate.SignaturePredicate.create(pubkey));
    return import_EncodedPredicate.EncodedPredicate.equals(owner, claimed);
  }
  async isSpent(token, _options) {
    const probe = await import_TransferTransaction.TransferTransaction.create(
      token.sdkToken,
      import_SignaturePredicate.SignaturePredicate.create(this.deps.signingService.publicKey),
      new Uint8Array(32)
    );
    const stateId = await import_StateId.StateId.fromTransaction(probe);
    const response = await this.deps.client.getInclusionProof(stateId);
    return response.inclusionProof.inclusionCertificate !== null;
  }
  // ── serialization ────────────────────────────────────────────────────────────
  /** @inheritDoc — SDK-derived (Token.fromCBOR → latestTransaction.calculateStateHash().imprint). */
  deliveryKeys(blobBytes) {
    return deriveDeliveryKeys(blobBytes);
  }
  encodeToken(token) {
    return token.blob;
  }
  async decodeToken(blob) {
    const sdkToken = await import_Token.Token.fromCBOR(blob.token);
    if (sdkToken.genesis.networkId.id !== this.deps.networkId.id) {
      throw new SphereError(
        `Token network mismatch: token is on network ${sdkToken.genesis.networkId.id}, engine on ${this.deps.networkId.id}`,
        "VALIDATION_ERROR"
      );
    }
    return this.wrapToken(sdkToken);
  }
  // ── internals ────────────────────────────────────────────────────────────────
  /**
   * The effective realization seed for one transfer/split (Part E). A caller
   * that wants resumability supplies a pre-persisted UUIDv4 (sdk-changes E.3);
   * otherwise one is generated here — same derivation path, just non-resumable.
   * The canonical-lowercase-UUID check also keeps the HKDF info string
   * unambiguous (no ':' injection, no case-variant double derivations).
   */
  /** Validated op ordinal (§8.1 mask domain separation across ops of one send). */
  resolveOpIndex(options) {
    const opIndex = options?.opIndex ?? 0;
    if (!Number.isInteger(opIndex) || opIndex < 0) {
      throw new SphereError(`opIndex must be a non-negative integer, got ${String(opIndex)}`, "VALIDATION_ERROR");
    }
    return opIndex;
  }
  resolveTransferId(options) {
    const transferId = options?.transferId ?? randomUUID();
    if (!TRANSFER_ID_RE.test(transferId)) {
      throw new SphereError(
        `transferId must be a canonical lowercase UUID string, got "${transferId}"`,
        "VALIDATION_ERROR"
      );
    }
    return transferId;
  }
  /**
   * E.2 — status-agnostic idempotent submit with conflict detection
   * (sdk-changes E.2, ARCHITECTURE §8.1).
   *
   * Submit the certification request and treat ANY outcome — success status,
   * error status, or a response-parse throw (the live gateway may still emit
   * statuses this SDK no longer parses, e.g. the transitional STATE_ID_EXISTS)
   * — as "a certification for this state may already exist". Then ALWAYS fetch
   * the inclusion proof and match-verify it against the rebuilt transaction
   * (`waitInclusionProof` only returns an OK-verified proof):
   *
   * - match (OK)                    → proceed; this is how an interrupted
   *                                   attempt with the same transferId resumes;
   * - mismatch (hash ≠ rebuilt tx)  → `TransferConflictError`: the source was
   *                                   consumed by a DIFFERENT transaction (lost
   *                                   race) — never applied, never retried;
   * - no proof & the submit failed  → the original submit error (a genuine
   *                                   failure, not a resume case).
   */
  async submitAndAwaitProof(certificationData, transaction, failLabel, failCode, options) {
    let submitError = null;
    try {
      const response = await this.deps.client.submitCertificationRequest(certificationData);
      if (response.status !== import_CertificationResponse.CertificationStatus.SUCCESS) {
        submitError = new SphereError(`${failLabel}: ${response.status}`, failCode);
      }
    } catch (err) {
      submitError = err ?? new SphereError(`${failLabel}: submit failed`, failCode);
    }
    try {
      return await (0, import_InclusionProofUtils.waitInclusionProof)(
        this.deps.client,
        this.deps.trustBase,
        this.deps.predicateVerifier,
        transaction,
        options?.signal
      );
    } catch (err) {
      if (err instanceof Error && err.message === `Invalid inclusion proof status: ${import_InclusionProofVerificationRule.InclusionProofVerificationStatus.TRANSACTION_HASH_MISMATCH}`) {
        throw new TransferConflictError(
          `${failLabel}: the source state was already consumed by a different transaction (lost race \u2014 abort this intent and re-plan under a new transferId)`,
          err
        );
      }
      if (submitError !== null) throw submitError;
      throw err;
    }
  }
  /** Fail fast if this engine's key does not own the token's current state. */
  assertOwned(token) {
    if (!this.isOwnedBy(token, this.deps.signingService.publicKey)) {
      throw new SphereError("Cannot transfer a token not owned by this engine identity", "VALIDATION_ERROR");
    }
  }
  /** Wrap an SDK token into a SphereToken: cache its blob (incl. stable tokenId) + decoded value. */
  wrapToken(sdkToken) {
    const data = sdkToken.genesis.data;
    let value = null;
    if (data && this.isSpherePaymentData(data)) {
      try {
        value = SpherePaymentData.fromCBOR(data).toValue();
      } catch (err) {
        throw new SphereError(
          `Failed to decode token payment data: ${err instanceof Error ? err.message : String(err)}`,
          "VALIDATION_ERROR"
        );
      }
    }
    const blob = {
      v: TOKEN_BLOB_VERSION,
      network: sdkToken.genesis.networkId.id,
      tokenId: import_HexConverter.HexConverter.encode(sdkToken.id.bytes),
      token: sdkToken.toCBOR()
    };
    return { sdkToken, blob, value };
  }
  /** True if the bytes are a SpherePaymentData envelope (value token) vs a raw data token. */
  isSpherePaymentData(data) {
    try {
      return import_CborDeserializer.CborDeserializer.decodeTag(data).tag === SpherePaymentData.CBOR_TAG;
    } catch {
      return false;
    }
  }
};

// token-engine/factory.ts
async function createSphereTokenEngine(config) {
  if (config.trustBaseJson == null) {
    throw new SphereError("Engine config requires a trust base (trustBaseJson)", "INVALID_CONFIG");
  }
  if (!config.apiKey) {
    logger.warn(
      "TokenEngine",
      "No aggregator apiKey \u2014 pass config.oracle.apiKey (testnet2 value in .env.example; mainnet from a secret env var). Gateway requests will be unauthenticated."
    );
  }
  const trustBase = import_RootTrustBase.RootTrustBase.fromJSON(config.trustBaseJson);
  const predicateVerifier = import_PredicateVerifierService.PredicateVerifierService.create();
  const mintJustificationVerifier = new import_MintJustificationVerifierService.MintJustificationVerifierService();
  mintJustificationVerifier.register(
    new import_SplitMintJustificationVerifier.SplitMintJustificationVerifier(trustBase, predicateVerifier, decodeSpherePaymentData)
  );
  const deps = {
    client: new import_StateTransitionClient.StateTransitionClient(new import_AggregatorClient.AggregatorClient(config.aggregatorUrl, config.apiKey ?? null)),
    trustBase,
    predicateVerifier,
    mintJustificationVerifier,
    signingService: new import_SigningService.SigningService(config.privateKey),
    // Also the HKDF ikm for deterministic realization (Part E.1) — the
    // SigningService wraps the key but does not expose it back.
    privateKey: config.privateKey,
    // The trust base is the single source of truth for the network id (it carries
    // NetworkId.fromId, so any id works — e.g. testnet2 = 4 — with no enum entry).
    networkId: trustBase.networkId
  };
  return new SphereTokenEngine(deps);
}

// token-engine/unicity-id.ts
var SelfIssuedUnicityIdMinter = class {
  constructor(client, trustBase, predicateVerifier, signingService) {
    this.client = client;
    this.trustBase = trustBase;
    this.predicateVerifier = predicateVerifier;
    this.signingService = signingService;
  }
  async mintUnicityIdToken(name, options) {
    const unicityId = new import_UnicityId.UnicityId(name);
    const self = import_SignaturePredicate.SignaturePredicate.fromSigningService(this.signingService);
    const mintTx = await import_UnicityIdMintTransaction.UnicityIdMintTransaction.create(
      self,
      // issuer lock script (self-issued)
      self,
      // recipient — locks the minted token state to this wallet
      unicityId,
      new import_TokenType.TokenType(import_HexConverter.HexConverter.decode(UNICITY_TOKEN_TYPE_HEX)),
      self
      // target predicate the unicity id resolves to
    );
    const certificationData = await import_CertificationData.CertificationData.fromTransaction(
      mintTx,
      await import_SignaturePredicateUnlockScript.SignaturePredicateUnlockScript.create(mintTx, this.signingService)
    );
    const response = await this.client.submitCertificationRequest(certificationData);
    if (response.status !== import_CertificationResponse.CertificationStatus.SUCCESS) {
      logger.warn(
        "UnicityIdMinter",
        `Certification request returned ${response.status} for "@${name}" \u2014 attempting proof recovery (idempotent re-mint).`
      );
    }
    try {
      const proof = await (0, import_InclusionProofUtils.waitInclusionProof)(
        this.client,
        this.trustBase,
        this.predicateVerifier,
        mintTx,
        options?.signal
      );
      const certified = await mintTx.toCertifiedTransaction(this.trustBase, this.predicateVerifier, proof);
      const token = await import_UnicityIdToken.UnicityIdToken.mint(this.trustBase, this.predicateVerifier, certified);
      return {
        tokenCborHex: import_HexConverter.HexConverter.encode(token.toCBOR()),
        tokenId: import_HexConverter.HexConverter.encode(token.id.bytes)
      };
    } catch (err) {
      const detail = err instanceof Error ? err.message : String(err);
      throw new SphereError(
        `Unicity ID mint failed for "@${name}": ${detail}` + (response.status !== import_CertificationResponse.CertificationStatus.SUCCESS ? ` (certification status: ${response.status})` : ""),
        "AGGREGATOR_ERROR"
      );
    }
  }
};
function createUnicityIdMinter(config) {
  if (config.trustBaseJson == null) {
    throw new SphereError("Unicity ID minter requires a trust base (trustBaseJson)", "INVALID_CONFIG");
  }
  const trustBase = import_RootTrustBase.RootTrustBase.fromJSON(config.trustBaseJson);
  return new SelfIssuedUnicityIdMinter(
    new import_StateTransitionClient.StateTransitionClient(new import_AggregatorClient.AggregatorClient(config.aggregatorUrl, config.apiKey ?? null)),
    trustBase,
    import_PredicateVerifierService.PredicateVerifierService.create(),
    new import_SigningService.SigningService(config.privateKey)
  );
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  SpherePaymentData,
  TransferConflictError,
  createSphereTokenEngine,
  createUnicityIdMinter,
  decodeSpherePaymentData,
  deriveDirectAddress,
  sphereAssetToSdk
});
