// @unicitylabs/nostr-js-sdk/dist/cjs/client/WebSocketAdapter.browser.js
"use strict";
/**
 * WebSocketAdapter - Browser implementation.
 * Uses the native WebSocket API.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CLOSED = exports.CLOSING = exports.OPEN = exports.CONNECTING = void 0;
exports.createWebSocket = createWebSocket;
exports.extractMessageData = extractMessageData;
/** WebSocket ready state constants */
exports.CONNECTING = 0;
exports.OPEN = 1;
exports.CLOSING = 2;
exports.CLOSED = 3;
/**
 * Create a WebSocket connection using native browser WebSocket.
 * @param url WebSocket URL (ws:// or wss://)
 * @returns WebSocket instance
 */
async function createWebSocket(url) {
    if (typeof WebSocket === 'undefined') {
        throw new Error('WebSocket not available in this environment');
    }
    return new WebSocket(url);
}
/**
 * Extract string data from WebSocket message event.
 * @param event WebSocket message event
 * @returns String message data
 */
function extractMessageData(event) {
    if (typeof event.data === 'string') {
        return event.data;
    }
    if (event.data instanceof ArrayBuffer) {
        return new TextDecoder().decode(event.data);
    }
    if (typeof Blob !== 'undefined' && event.data instanceof Blob) {
        throw new Error('Blob messages are not supported');
    }
    return String(event.data);
}
