// @unicitylabs/nostr-js-sdk/dist/cjs/client/WebSocketAdapter.js
"use strict";
/**
 * WebSocketAdapter - Cross-platform WebSocket implementation.
 * Provides a unified interface for WebSocket connections in Node.js and browsers.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.CLOSED = exports.CLOSING = exports.OPEN = exports.CONNECTING = void 0;
exports.createWebSocket = createWebSocket;
exports.extractMessageData = extractMessageData;
/** WebSocket ready state constants */
exports.CONNECTING = 0;
exports.OPEN = 1;
exports.CLOSING = 2;
exports.CLOSED = 3;
/**
 * Create a WebSocket connection that works in both Node.js and browsers.
 * @param url WebSocket URL (ws:// or wss://)
 * @returns WebSocket instance
 */
async function createWebSocket(url) {
    // Check if we're in a browser environment
    if (typeof WebSocket !== 'undefined') {
        return new WebSocket(url);
    }
    // Node.js environment - dynamically import ws
    try {
        const { default: WS } = await Promise.resolve().then(() => __importStar(require('ws')));
        return new WS(url);
    }
    catch {
        throw new Error('WebSocket not available. In Node.js, install the "ws" package: npm install ws');
    }
}
/**
 * Extract string data from WebSocket message event.
 * Handles different message types across platforms.
 * @param event WebSocket message event
 * @returns String message data
 */
function extractMessageData(event) {
    if (typeof event.data === 'string') {
        return event.data;
    }
    if (event.data instanceof ArrayBuffer) {
        return new TextDecoder().decode(event.data);
    }
    if (typeof Blob !== 'undefined' && event.data instanceof Blob) {
        // This shouldn't happen in normal Nostr relay communication
        throw new Error('Blob messages are not supported');
    }
    // Node.js Buffer case
    if (Buffer && Buffer.isBuffer(event.data)) {
        return event.data.toString('utf-8');
    }
    return String(event.data);
}
