// @unicitylabs/nostr-js-sdk/dist/cjs/protocol/Event.js
"use strict";
/**
 * Event - Represents a Nostr event (NIP-01).
 * Provides serialization, signing, and verification functionality.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Event = void 0;
const sha256_1 = require("@noble/hashes/sha256");
const utils_1 = require("@noble/hashes/utils");
const Schnorr = __importStar(require("../crypto/schnorr.js"));
/**
 * Event class representing a Nostr event (NIP-01).
 */
class Event {
    /** Event ID - SHA-256 hash of the serialized event data (hex) */
    id;
    /** Creator's x-only public key (hex) */
    pubkey;
    /** Unix timestamp in seconds */
    created_at;
    /** Event kind (type) */
    kind;
    /** Event tags */
    tags;
    /** Event content */
    content;
    /** Schnorr signature (hex) */
    sig;
    /**
     * Create an Event instance.
     * @param data Signed event data
     */
    constructor(data) {
        this.id = data.id;
        this.pubkey = data.pubkey;
        this.created_at = data.created_at;
        this.kind = data.kind;
        this.tags = data.tags;
        this.content = data.content;
        this.sig = data.sig;
    }
    /**
     * Create and sign a new event.
     * @param keyManager Key manager with signing key
     * @param data Unsigned event data
     * @returns Signed Event instance
     */
    static create(keyManager, data) {
        const pubkey = keyManager.getPublicKeyHex();
        const created_at = data.created_at ?? Math.floor(Date.now() / 1000);
        // Calculate event ID
        const id = Event.calculateId(pubkey, created_at, data.kind, data.tags, data.content);
        // Sign the event ID
        const idBytes = (0, utils_1.hexToBytes)(id);
        const sig = keyManager.signHex(idBytes);
        return new Event({
            id,
            pubkey,
            created_at,
            kind: data.kind,
            tags: data.tags,
            content: data.content,
            sig,
        });
    }
    /**
     * Calculate the event ID from event data.
     * ID = SHA-256(serialized event data)
     * Serialized format: [0, pubkey, created_at, kind, tags, content]
     */
    static calculateId(pubkey, created_at, kind, tags, content) {
        const serialized = JSON.stringify([0, pubkey, created_at, kind, tags, content]);
        const hash = (0, sha256_1.sha256)(new TextEncoder().encode(serialized));
        return (0, utils_1.bytesToHex)(hash);
    }
    /**
     * Verify the event signature.
     * @returns true if the signature is valid
     */
    verify() {
        try {
            // Verify the ID
            const calculatedId = Event.calculateId(this.pubkey, this.created_at, this.kind, this.tags, this.content);
            if (calculatedId !== this.id) {
                return false;
            }
            // Verify the signature
            const idBytes = (0, utils_1.hexToBytes)(this.id);
            const sigBytes = (0, utils_1.hexToBytes)(this.sig);
            const pubkeyBytes = (0, utils_1.hexToBytes)(this.pubkey);
            return Schnorr.verify(sigBytes, idBytes, pubkeyBytes);
        }
        catch {
            return false;
        }
    }
    /**
     * Parse an event from JSON data.
     * @param json JSON object or string
     * @returns Event instance
     */
    static fromJSON(json) {
        const data = typeof json === 'string' ? JSON.parse(json) : json;
        if (!Event.isValidEventData(data)) {
            throw new Error('Invalid event data');
        }
        return new Event(data);
    }
    /**
     * Check if data has valid event structure.
     */
    static isValidEventData(data) {
        if (typeof data !== 'object' || data === null) {
            return false;
        }
        const obj = data;
        return (typeof obj.id === 'string' &&
            typeof obj.pubkey === 'string' &&
            typeof obj.created_at === 'number' &&
            typeof obj.kind === 'number' &&
            Array.isArray(obj.tags) &&
            typeof obj.content === 'string' &&
            typeof obj.sig === 'string');
    }
    /**
     * Convert the event to a plain object.
     * @returns Plain object representation
     */
    toJSON() {
        return {
            id: this.id,
            pubkey: this.pubkey,
            created_at: this.created_at,
            kind: this.kind,
            tags: this.tags,
            content: this.content,
            sig: this.sig,
        };
    }
    /**
     * Get the first value of a tag by name.
     * @param tagName Tag name to find
     * @returns First value of the tag, or undefined if not found
     */
    getTagValue(tagName) {
        const tag = this.tags.find((t) => t[0] === tagName);
        return tag?.[1];
    }
    /**
     * Get all values of a tag by name.
     * @param tagName Tag name to find
     * @returns Array of tag values
     */
    getTagValues(tagName) {
        return this.tags.filter((t) => t[0] === tagName).map((t) => t[1] ?? '');
    }
    /**
     * Check if a tag exists.
     * @param tagName Tag name to check
     * @returns true if the tag exists
     */
    hasTag(tagName) {
        return this.tags.some((t) => t[0] === tagName);
    }
    /**
     * Get all values from a single tag entry (excluding the tag name).
     * @param tagName Tag name to find
     * @returns Array of values from the first matching tag, or empty array
     */
    getTagEntryValues(tagName) {
        const tag = this.tags.find((t) => t[0] === tagName);
        return tag ? tag.slice(1) : [];
    }
}
exports.Event = Event;
