// @unicitylabs/nostr-js-sdk/dist/cjs/protocol/EventKinds.js
"use strict";
/**
 * EventKinds - Standardized Nostr event kind definitions.
 * Includes both standard NIP kinds and Unicity custom kinds.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PAYMENT_REQUEST_RESPONSE = exports.PAYMENT_REQUEST = exports.FILE_METADATA = exports.TOKEN_TRANSFER = exports.AGENT_LOCATION = exports.AGENT_PROFILE = exports.APP_DATA = exports.AUTH = exports.RELAY_LIST = exports.GIFT_WRAP = exports.REACTION = exports.READ_RECEIPT = exports.CHAT_MESSAGE = exports.SEAL = exports.DELETION = exports.ENCRYPTED_DM = exports.CONTACTS = exports.RECOMMEND_RELAY = exports.TEXT_NOTE = exports.PROFILE = void 0;
exports.isReplaceable = isReplaceable;
exports.isEphemeral = isEphemeral;
exports.isParameterizedReplaceable = isParameterizedReplaceable;
exports.getName = getName;
// ============================================================================
// Standard NIP Event Kinds
// ============================================================================
/** NIP-01: User profile metadata */
exports.PROFILE = 0;
/** NIP-01: Short text note */
exports.TEXT_NOTE = 1;
/** NIP-01: Recommend relay to followers */
exports.RECOMMEND_RELAY = 2;
/** NIP-02: Contact list (follows) */
exports.CONTACTS = 3;
/** NIP-04: Encrypted direct messages */
exports.ENCRYPTED_DM = 4;
/** NIP-09: Event deletion */
exports.DELETION = 5;
/** NIP-17: Seal (signed, encrypted rumor) */
exports.SEAL = 13;
/** NIP-17: Private chat message (rumor - unsigned inner event) */
exports.CHAT_MESSAGE = 14;
/** NIP-17: Read receipt (rumor kind) */
exports.READ_RECEIPT = 15;
/** NIP-25: Reactions (likes, etc.) */
exports.REACTION = 7;
/** NIP-59: Gift wrap for private events */
exports.GIFT_WRAP = 1059;
/** NIP-65: Relay list metadata */
exports.RELAY_LIST = 10002;
/** NIP-42: Client authentication to relay */
exports.AUTH = 22242;
/** NIP-78: Application-specific data (parameterized replaceable) */
exports.APP_DATA = 30078;
// ============================================================================
// Unicity Custom Event Kinds
// ============================================================================
/** Unicity: Agent profile information */
exports.AGENT_PROFILE = 31111;
/** Unicity: Agent GPS location */
exports.AGENT_LOCATION = 31112;
/** Unicity: Token transfer event */
exports.TOKEN_TRANSFER = 31113;
/** Unicity: File metadata */
exports.FILE_METADATA = 31114;
/** Unicity: Payment request */
exports.PAYMENT_REQUEST = 31115;
/** Unicity: Payment request response (accept/decline) */
exports.PAYMENT_REQUEST_RESPONSE = 31116;
// ============================================================================
// Event Kind Classification Functions
// ============================================================================
/**
 * Check if an event kind is replaceable.
 * Replaceable events (kinds 0, 3, or 10000-19999) are replaced when a new
 * event with the same kind and pubkey is published.
 * @param kind Event kind number
 * @returns true if the event kind is replaceable
 */
function isReplaceable(kind) {
    return kind === 0 || kind === 3 || (kind >= 10000 && kind < 20000);
}
/**
 * Check if an event kind is ephemeral.
 * Ephemeral events (kinds 20000-29999) are not stored by relays.
 * @param kind Event kind number
 * @returns true if the event kind is ephemeral
 */
function isEphemeral(kind) {
    return kind >= 20000 && kind < 30000;
}
/**
 * Check if an event kind is parameterized replaceable.
 * Parameterized replaceable events (kinds 30000-39999) are replaced when a
 * new event with the same kind, pubkey, and "d" tag value is published.
 * @param kind Event kind number
 * @returns true if the event kind is parameterized replaceable
 */
function isParameterizedReplaceable(kind) {
    return kind >= 30000 && kind < 40000;
}
/**
 * Get a human-readable name for an event kind.
 * @param kind Event kind number
 * @returns Human-readable name for the event kind
 */
function getName(kind) {
    switch (kind) {
        case exports.PROFILE:
            return 'Profile';
        case exports.TEXT_NOTE:
            return 'Text Note';
        case exports.RECOMMEND_RELAY:
            return 'Recommend Relay';
        case exports.CONTACTS:
            return 'Contacts';
        case exports.ENCRYPTED_DM:
            return 'Encrypted DM';
        case exports.DELETION:
            return 'Deletion';
        case exports.SEAL:
            return 'Seal';
        case exports.CHAT_MESSAGE:
            return 'Chat Message';
        case exports.READ_RECEIPT:
            return 'Read Receipt';
        case exports.REACTION:
            return 'Reaction';
        case exports.GIFT_WRAP:
            return 'Gift Wrap';
        case exports.RELAY_LIST:
            return 'Relay List';
        case exports.APP_DATA:
            return 'App Data';
        case exports.AGENT_PROFILE:
            return 'Agent Profile';
        case exports.AGENT_LOCATION:
            return 'Agent Location';
        case exports.TOKEN_TRANSFER:
            return 'Token Transfer';
        case exports.FILE_METADATA:
            return 'File Metadata';
        case exports.PAYMENT_REQUEST:
            return 'Payment Request';
        case exports.PAYMENT_REQUEST_RESPONSE:
            return 'Payment Request Response';
        default:
            if (isReplaceable(kind)) {
                return `Replaceable (${kind})`;
            }
            if (isEphemeral(kind)) {
                return `Ephemeral (${kind})`;
            }
            if (isParameterizedReplaceable(kind)) {
                return `Parameterized Replaceable (${kind})`;
            }
            return `Unknown (${kind})`;
    }
}
