// @unicitylabs/nostr-js-sdk/dist/cjs/protocol/Filter.js
"use strict";
/**
 * Filter - Nostr subscription filter (NIP-01).
 * Provides a builder pattern for constructing filters.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.FilterBuilder = exports.Filter = void 0;
/**
 * Filter class for Nostr subscription queries.
 * Use the builder pattern to construct filters.
 */
class Filter {
    ids;
    authors;
    kinds;
    '#e';
    '#p';
    '#t';
    '#d';
    '#h';
    since;
    until;
    limit;
    /**
     * Create a Filter instance.
     * @param data Optional filter data
     */
    constructor(data) {
        if (data) {
            if (data.ids)
                this.ids = [...data.ids];
            if (data.authors)
                this.authors = [...data.authors];
            if (data.kinds)
                this.kinds = [...data.kinds];
            if (data['#e'])
                this['#e'] = [...data['#e']];
            if (data['#p'])
                this['#p'] = [...data['#p']];
            if (data['#t'])
                this['#t'] = [...data['#t']];
            if (data['#d'])
                this['#d'] = [...data['#d']];
            if (data['#h'])
                this['#h'] = [...data['#h']];
            if (data.since !== undefined)
                this.since = data.since;
            if (data.until !== undefined)
                this.until = data.until;
            if (data.limit !== undefined)
                this.limit = data.limit;
        }
    }
    /**
     * Create a new Filter builder.
     * @returns Filter builder instance
     */
    static builder() {
        return new FilterBuilder();
    }
    /**
     * Convert the filter to a plain object for JSON serialization.
     * Only includes defined properties.
     * @returns Plain object representation
     */
    toJSON() {
        const result = {};
        if (this.ids && this.ids.length > 0)
            result.ids = this.ids;
        if (this.authors && this.authors.length > 0)
            result.authors = this.authors;
        if (this.kinds && this.kinds.length > 0)
            result.kinds = this.kinds;
        if (this['#e'] && this['#e'].length > 0)
            result['#e'] = this['#e'];
        if (this['#p'] && this['#p'].length > 0)
            result['#p'] = this['#p'];
        if (this['#t'] && this['#t'].length > 0)
            result['#t'] = this['#t'];
        if (this['#d'] && this['#d'].length > 0)
            result['#d'] = this['#d'];
        if (this['#h'] && this['#h'].length > 0)
            result['#h'] = this['#h'];
        if (this.since !== undefined)
            result.since = this.since;
        if (this.until !== undefined)
            result.until = this.until;
        if (this.limit !== undefined)
            result.limit = this.limit;
        return result;
    }
    /**
     * Parse a filter from JSON data.
     * @param json JSON object or string
     * @returns Filter instance
     */
    static fromJSON(json) {
        const data = typeof json === 'string' ? JSON.parse(json) : json;
        return new Filter(data);
    }
}
exports.Filter = Filter;
/**
 * Builder class for constructing Filter instances.
 */
class FilterBuilder {
    data = {};
    ids(idsOrFirst, ...rest) {
        if (Array.isArray(idsOrFirst)) {
            this.data.ids = [...idsOrFirst];
        }
        else {
            this.data.ids = [idsOrFirst, ...rest];
        }
        return this;
    }
    authors(authorsOrFirst, ...rest) {
        if (Array.isArray(authorsOrFirst)) {
            this.data.authors = [...authorsOrFirst];
        }
        else {
            this.data.authors = [authorsOrFirst, ...rest];
        }
        return this;
    }
    kinds(kindsOrFirst, ...rest) {
        if (Array.isArray(kindsOrFirst)) {
            this.data.kinds = [...kindsOrFirst];
        }
        else {
            this.data.kinds = [kindsOrFirst, ...rest];
        }
        return this;
    }
    eTags(eTagsOrFirst, ...rest) {
        if (Array.isArray(eTagsOrFirst)) {
            this.data['#e'] = [...eTagsOrFirst];
        }
        else {
            this.data['#e'] = [eTagsOrFirst, ...rest];
        }
        return this;
    }
    pTags(pTagsOrFirst, ...rest) {
        if (Array.isArray(pTagsOrFirst)) {
            this.data['#p'] = [...pTagsOrFirst];
        }
        else {
            this.data['#p'] = [pTagsOrFirst, ...rest];
        }
        return this;
    }
    tTags(tTagsOrFirst, ...rest) {
        if (Array.isArray(tTagsOrFirst)) {
            this.data['#t'] = [...tTagsOrFirst];
        }
        else {
            this.data['#t'] = [tTagsOrFirst, ...rest];
        }
        return this;
    }
    dTags(dTagsOrFirst, ...rest) {
        if (Array.isArray(dTagsOrFirst)) {
            this.data['#d'] = [...dTagsOrFirst];
        }
        else {
            this.data['#d'] = [dTagsOrFirst, ...rest];
        }
        return this;
    }
    hTags(hTagsOrFirst, ...rest) {
        if (Array.isArray(hTagsOrFirst)) {
            this.data['#h'] = [...hTagsOrFirst];
        }
        else {
            this.data['#h'] = [hTagsOrFirst, ...rest];
        }
        return this;
    }
    /**
     * Set minimum timestamp (inclusive).
     * @param since Unix timestamp in seconds
     * @returns This builder for chaining
     */
    since(since) {
        this.data.since = since;
        return this;
    }
    /**
     * Set maximum timestamp (inclusive).
     * @param until Unix timestamp in seconds
     * @returns This builder for chaining
     */
    until(until) {
        this.data.until = until;
        return this;
    }
    /**
     * Set maximum number of events to return.
     * @param limit Maximum number of events
     * @returns This builder for chaining
     */
    limit(limit) {
        this.data.limit = limit;
        return this;
    }
    /**
     * Build the Filter instance.
     * @returns Filter instance
     */
    build() {
        return new Filter(this.data);
    }
}
exports.FilterBuilder = FilterBuilder;
