// @unicitylabs/sphere-sdk/dist/impl/nodejs/connect/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// impl/nodejs/connect/index.ts
var connect_exports = {};
__export(connect_exports, {
  WebSocketClientTransport: () => WebSocketClientTransport,
  WebSocketServerTransport: () => WebSocketServerTransport,
  WebSocketTransport: () => WebSocketTransport
});
module.exports = __toCommonJS(connect_exports);

// core/logger.ts
var LOGGER_KEY = "__sphere_sdk_logger__";
function getState() {
  const g = globalThis;
  if (!g[LOGGER_KEY]) {
    g[LOGGER_KEY] = { debug: false, tags: {}, handler: null };
  }
  return g[LOGGER_KEY];
}
function isEnabled(tag) {
  const state = getState();
  if (tag in state.tags) return state.tags[tag];
  return state.debug;
}
var logger = {
  /**
   * Configure the logger. Can be called multiple times (last write wins).
   * Typically called by createBrowserProviders(), createNodeProviders(), or Sphere.init().
   */
  configure(config) {
    const state = getState();
    if (config.debug !== void 0) state.debug = config.debug;
    if (config.handler !== void 0) state.handler = config.handler;
  },
  /**
   * Enable/disable debug logging for a specific tag.
   * Per-tag setting overrides the global debug flag.
   *
   * @example
   * ```ts
   * logger.setTagDebug('Nostr', true);  // enable only Nostr logs
   * logger.setTagDebug('Nostr', false); // disable Nostr logs even if global debug=true
   * ```
   */
  setTagDebug(tag, enabled) {
    getState().tags[tag] = enabled;
  },
  /**
   * Clear per-tag override, falling back to global debug flag.
   */
  clearTagDebug(tag) {
    delete getState().tags[tag];
  },
  /** Returns true if debug mode is enabled for the given tag (or globally). */
  isDebugEnabled(tag) {
    if (tag) return isEnabled(tag);
    return getState().debug;
  },
  /**
   * Debug-level log. Only shown when debug is enabled (globally or for this tag).
   * Use for detailed operational information.
   */
  debug(tag, message, ...args) {
    if (!isEnabled(tag)) return;
    const state = getState();
    if (state.handler) {
      state.handler("debug", tag, message, ...args);
    } else {
      console.log(`[${tag}]`, message, ...args);
    }
  },
  /**
   * Warning-level log. ALWAYS shown regardless of debug flag.
   * Use for important but non-critical issues (timeouts, retries, degraded state).
   */
  warn(tag, message, ...args) {
    const state = getState();
    if (state.handler) {
      state.handler("warn", tag, message, ...args);
    } else {
      console.warn(`[${tag}]`, message, ...args);
    }
  },
  /**
   * Error-level log. ALWAYS shown regardless of debug flag.
   * Use for critical failures that should never be silenced.
   */
  error(tag, message, ...args) {
    const state = getState();
    if (state.handler) {
      state.handler("error", tag, message, ...args);
    } else {
      console.error(`[${tag}]`, message, ...args);
    }
  },
  /** Reset all logger state (debug flag, tags, handler). Primarily for tests. */
  reset() {
    const g = globalThis;
    delete g[LOGGER_KEY];
  }
};

// connect/semver.ts
function majorOf(v) {
  return parseInt(String(v).split(".")[0], 10);
}

// constants.ts
var STORAGE_KEYS_GLOBAL = {
  /** Encrypted BIP39 mnemonic */
  MNEMONIC: "mnemonic",
  /** Encrypted master private key */
  MASTER_KEY: "master_key",
  /** BIP32 chain code */
  CHAIN_CODE: "chain_code",
  /** HD derivation path (full path like m/44'/0'/0'/0/0) */
  DERIVATION_PATH: "derivation_path",
  /** Base derivation path (like m/44'/0'/0' without chain/index) */
  BASE_PATH: "base_path",
  /** Derivation mode: bip32, wif_hmac, legacy_hmac */
  DERIVATION_MODE: "derivation_mode",
  /** Wallet source: mnemonic, file, unknown */
  WALLET_SOURCE: "wallet_source",
  /** Wallet existence flag */
  WALLET_EXISTS: "wallet_exists",
  /** Current active address index */
  CURRENT_ADDRESS_INDEX: "current_address_index",
  /** Nametag cache per address (separate from tracked addresses registry) */
  ADDRESS_NAMETAGS: "address_nametags",
  /** Active addresses registry (JSON: TrackedAddressesStorage) */
  TRACKED_ADDRESSES: "tracked_addresses",
  /** Last processed Nostr wallet event timestamp (unix seconds), keyed per pubkey */
  LAST_WALLET_EVENT_TS: "last_wallet_event_ts",
  /** Last processed Nostr DM (gift-wrap) event timestamp (unix seconds), keyed per pubkey */
  LAST_DM_EVENT_TS: "last_dm_event_ts",
  /** Group chat: last used relay URL (stale data detection) — global, same relay for all addresses */
  GROUP_CHAT_RELAY_URL: "group_chat_relay_url",
  /** Cached token registry JSON (fetched from remote) */
  TOKEN_REGISTRY_CACHE: "token_registry_cache",
  /** Timestamp of last token registry cache update (ms since epoch) */
  TOKEN_REGISTRY_CACHE_TS: "token_registry_cache_ts",
  /** Cached price data JSON (from CoinGecko or other provider) */
  PRICE_CACHE: "price_cache",
  /** Timestamp of last price cache update (ms since epoch) */
  PRICE_CACHE_TS: "price_cache_ts"
};
var STORAGE_KEYS_ADDRESS = {
  /** Pending transfers for this address */
  PENDING_TRANSFERS: "pending_transfers",
  /** Transfer outbox for this address */
  OUTBOX: "outbox",
  /** Conversations for this address */
  CONVERSATIONS: "conversations",
  /** Messages for this address */
  MESSAGES: "messages",
  /** Transaction history for this address */
  TRANSACTION_HISTORY: "transaction_history",
  /** Pending V5 finalization tokens (unconfirmed instant split tokens) */
  PENDING_V5_TOKENS: "pending_v5_tokens",
  /**
   * FINISHED v2 token blobs awaiting transport delivery. Written the moment a
   * transfer/split output is certified on-chain (the source is already spent),
   * removed after successful delivery — survives transport failures + crashes
   * so the recipient's token is never lost with the process.
   */
  PENDING_V2_DELIVERIES: "pending_v2_deliveries",
  /** Group chat: joined groups for this address */
  GROUP_CHAT_GROUPS: "group_chat_groups",
  /** Group chat: messages for this address */
  GROUP_CHAT_MESSAGES: "group_chat_messages",
  /** Group chat: members for this address */
  GROUP_CHAT_MEMBERS: "group_chat_members",
  /** Group chat: processed event IDs for deduplication */
  GROUP_CHAT_PROCESSED_EVENTS: "group_chat_processed_events",
  /** Processed V5 split group IDs for Nostr re-delivery dedup */
  PROCESSED_SPLIT_GROUP_IDS: "processed_split_group_ids",
  /** Processed V6 combined transfer IDs for Nostr re-delivery dedup */
  PROCESSED_COMBINED_TRANSFER_IDS: "processed_combined_transfer_ids",
  // Invoice / Accounting storage keys
  /** Set of cancelled invoice IDs (JSON string array) */
  CANCELLED_INVOICES: "cancelled_invoices",
  /** Set of closed invoice IDs (JSON string array) */
  CLOSED_INVOICES: "closed_invoices",
  /** Frozen balances for terminated invoices (JSON map: invoiceId → FrozenInvoiceBalances) */
  FROZEN_BALANCES: "frozen_balances",
  /** Auto-return settings (JSON: AutoReturnSettings) */
  AUTO_RETURN: "auto_return",
  /** Auto-return dedup ledger (JSON: AutoReturnLedger) */
  AUTO_RETURN_LEDGER: "auto_return_ledger",
  /** Invoice-transfer index metadata (JSON: Record<invoiceId, { terminated, frozenAt? }>) */
  INV_LEDGER_INDEX: "inv_ledger_index",
  /** Token scan state watermarks (JSON: Record<tokenId, txCount>) */
  TOKEN_SCAN_STATE: "token_scan_state",
  // Swap storage keys
  /** Per-swap key: swap:{swapId} */
  SWAP_RECORD_PREFIX: "swap:",
  /** Lightweight index array for listing */
  SWAP_INDEX: "swap_index"
};
var NETWORK_SCOPED_ADDRESS_KEYS = [
  STORAGE_KEYS_ADDRESS.PENDING_TRANSFERS,
  STORAGE_KEYS_ADDRESS.OUTBOX,
  STORAGE_KEYS_ADDRESS.TRANSACTION_HISTORY,
  STORAGE_KEYS_ADDRESS.PENDING_V5_TOKENS,
  STORAGE_KEYS_ADDRESS.PENDING_V2_DELIVERIES,
  STORAGE_KEYS_ADDRESS.PROCESSED_SPLIT_GROUP_IDS,
  STORAGE_KEYS_ADDRESS.PROCESSED_COMBINED_TRANSFER_IDS,
  STORAGE_KEYS_ADDRESS.CANCELLED_INVOICES,
  STORAGE_KEYS_ADDRESS.CLOSED_INVOICES,
  STORAGE_KEYS_ADDRESS.FROZEN_BALANCES,
  STORAGE_KEYS_ADDRESS.AUTO_RETURN,
  STORAGE_KEYS_ADDRESS.AUTO_RETURN_LEDGER,
  STORAGE_KEYS_ADDRESS.INV_LEDGER_INDEX,
  STORAGE_KEYS_ADDRESS.TOKEN_SCAN_STATE,
  STORAGE_KEYS_ADDRESS.SWAP_INDEX
];
var NETWORK_SCOPED_ADDRESS_PREFIXES = [
  STORAGE_KEYS_ADDRESS.SWAP_RECORD_PREFIX,
  // 'swap:'
  "inv_ledger:"
  // AccountingModule INV_LEDGER_PREFIX
];
var STORAGE_KEYS = {
  ...STORAGE_KEYS_GLOBAL,
  ...STORAGE_KEYS_ADDRESS
};
var DEFAULT_NOSTR_RELAYS = [
  "wss://relay.unicity.network",
  "wss://relay.damus.io",
  "wss://nos.lol",
  "wss://relay.nostr.band"
];
var DEFAULT_AGGREGATOR_URL = "https://aggregator.unicity.network/rpc";
var DEV_AGGREGATOR_URL = "https://dev-aggregator.dyndns.org/rpc";
var DEFAULT_IPFS_GATEWAYS = [
  "https://unicity-ipfs1.dyndns.org"
];
var DEFAULT_BASE_PATH = "m/44'/0'/0'";
var DEFAULT_DERIVATION_PATH = `${DEFAULT_BASE_PATH}/0/0`;
var TOKEN_REGISTRY_URL = "https://raw.githubusercontent.com/unicitynetwork/unicity-ids/refs/heads/main/unicity-ids.testnet.json";
var TEST_NOSTR_RELAYS = [
  "wss://nostr-relay.testnet.unicity.network"
];
var DEFAULT_GROUP_RELAYS = [
  "wss://sphere-relay.unicity.network"
];
var NETWORKS = {
  mainnet: {
    name: "Mainnet",
    aggregatorUrl: DEFAULT_AGGREGATOR_URL,
    nostrRelays: DEFAULT_NOSTR_RELAYS,
    ipfsGateways: DEFAULT_IPFS_GATEWAYS,
    groupRelays: DEFAULT_GROUP_RELAYS,
    tokenRegistryUrl: TOKEN_REGISTRY_URL
  },
  // v1 cutover: 'testnet' now POINTS AT TESTNET2 (the v2 gateway network). The
  // old goggregator testnet spoke the removed v1 protocol — a v2 engine cannot
  // run against it. 'testnet2' stays as an alias of the same configuration.
  testnet: {
    name: "Testnet2",
    networkId: 4,
    // v2 state-transition gateway (networkId 4 comes from the trust base). apiKey is env-injected.
    aggregatorUrl: "https://gateway.testnet2.unicity.network",
    nostrRelays: TEST_NOSTR_RELAYS,
    // reuse testnet infra (shared relays/ipfs)
    ipfsGateways: DEFAULT_IPFS_GATEWAYS,
    groupRelays: DEFAULT_GROUP_RELAYS,
    tokenRegistryUrl: "https://raw.githubusercontent.com/unicitynetwork/unicity-ids/refs/heads/main/unicity-ids.testnet2.json"
  },
  testnet2: {
    name: "Testnet2",
    networkId: 4,
    // v2 state-transition gateway (networkId 4 comes from the trust base). apiKey is env-injected.
    aggregatorUrl: "https://gateway.testnet2.unicity.network",
    nostrRelays: TEST_NOSTR_RELAYS,
    // reuse testnet infra (shared relays/ipfs)
    ipfsGateways: DEFAULT_IPFS_GATEWAYS,
    groupRelays: DEFAULT_GROUP_RELAYS,
    tokenRegistryUrl: "https://raw.githubusercontent.com/unicitynetwork/unicity-ids/refs/heads/main/unicity-ids.testnet2.json"
  },
  // NOTE: mainnet/dev still point at v1-era aggregators. The v2 engine cannot
  // operate against them until their gateways are cut over to the v2 protocol —
  // wallet operations on these networks fail loudly (AGGREGATOR_ERROR) until then.
  dev: {
    name: "Development",
    aggregatorUrl: DEV_AGGREGATOR_URL,
    nostrRelays: TEST_NOSTR_RELAYS,
    ipfsGateways: DEFAULT_IPFS_GATEWAYS,
    groupRelays: DEFAULT_GROUP_RELAYS,
    tokenRegistryUrl: TOKEN_REGISTRY_URL
  }
};
var SPHERE_NETWORKS = {
  testnet2: { id: NETWORKS.testnet2.networkId, name: "testnet2" }
};

// connect/protocol.ts
var SPHERE_CONNECT_NAMESPACE = "sphere-connect";
var SPHERE_CONNECT_VERSION = "2.0";
var RPC_METHODS = {
  GET_IDENTITY: "sphere_getIdentity",
  GET_BALANCE: "sphere_getBalance",
  GET_ASSETS: "sphere_getAssets",
  GET_FIAT_BALANCE: "sphere_getFiatBalance",
  GET_TOKENS: "sphere_getTokens",
  GET_HISTORY: "sphere_getHistory",
  RESOLVE: "sphere_resolve",
  SUBSCRIBE: "sphere_subscribe",
  UNSUBSCRIBE: "sphere_unsubscribe",
  DISCONNECT: "sphere_disconnect",
  GET_CONVERSATIONS: "sphere_getConversations",
  GET_MESSAGES: "sphere_getMessages",
  GET_DM_UNREAD_COUNT: "sphere_getDMUnreadCount",
  MARK_AS_READ: "sphere_markAsRead",
  GET_INVOICES: "sphere_getInvoices",
  GET_INVOICE_STATUS: "sphere_getInvoiceStatus"
};
var INTENT_ACTIONS = {
  SEND: "send",
  DM: "dm",
  PAYMENT_REQUEST: "payment_request",
  RECEIVE: "receive",
  SIGN_MESSAGE: "sign_message",
  CREATE_INVOICE: "create_invoice",
  CLOSE_INVOICE: "close_invoice",
  CANCEL_INVOICE: "cancel_invoice",
  PAY_INVOICE: "pay_invoice",
  RETURN_INVOICE_PAYMENT: "return_invoice_payment",
  IMPORT_INVOICE: "import_invoice",
  SEND_INVOICE_RECEIPTS: "send_invoice_receipts",
  SEND_CANCELLATION_NOTICES: "send_cancellation_notices",
  SET_AUTO_RETURN: "set_auto_return",
  MINT: "mint"
};
function isSphereConnectMessage(msg) {
  if (!msg || typeof msg !== "object") return false;
  const m = msg;
  if (m.ns !== SPHERE_CONNECT_NAMESPACE) return false;
  if (m.type === "handshake") return true;
  if (typeof m.v !== "string") return false;
  return majorOf(m.v) === majorOf(SPHERE_CONNECT_VERSION);
}

// connect/permissions.ts
var PERMISSION_SCOPES = {
  IDENTITY_READ: "identity:read",
  BALANCE_READ: "balance:read",
  TOKENS_READ: "tokens:read",
  HISTORY_READ: "history:read",
  EVENTS_SUBSCRIBE: "events:subscribe",
  RESOLVE_PEER: "resolve:peer",
  TRANSFER_REQUEST: "transfer:request",
  DM_REQUEST: "dm:request",
  DM_READ: "dm:read",
  DM_MANAGE: "dm:manage",
  PAYMENT_REQUEST: "payment:request",
  SIGN_REQUEST: "sign:request",
  MINT_REQUEST: "mint:request",
  INVOICE_READ: "invoice:read",
  INVOICE_WRITE: "invoice:write"
};
var ALL_PERMISSIONS = Object.values(PERMISSION_SCOPES);
var DEFAULT_PERMISSIONS = [
  PERMISSION_SCOPES.IDENTITY_READ
];
var METHOD_PERMISSIONS = {
  [RPC_METHODS.GET_IDENTITY]: PERMISSION_SCOPES.IDENTITY_READ,
  [RPC_METHODS.GET_BALANCE]: PERMISSION_SCOPES.BALANCE_READ,
  [RPC_METHODS.GET_ASSETS]: PERMISSION_SCOPES.BALANCE_READ,
  [RPC_METHODS.GET_FIAT_BALANCE]: PERMISSION_SCOPES.BALANCE_READ,
  [RPC_METHODS.GET_TOKENS]: PERMISSION_SCOPES.TOKENS_READ,
  [RPC_METHODS.GET_HISTORY]: PERMISSION_SCOPES.HISTORY_READ,
  [RPC_METHODS.RESOLVE]: PERMISSION_SCOPES.RESOLVE_PEER,
  [RPC_METHODS.SUBSCRIBE]: PERMISSION_SCOPES.EVENTS_SUBSCRIBE,
  [RPC_METHODS.UNSUBSCRIBE]: PERMISSION_SCOPES.EVENTS_SUBSCRIBE,
  [RPC_METHODS.GET_CONVERSATIONS]: PERMISSION_SCOPES.DM_READ,
  [RPC_METHODS.GET_MESSAGES]: PERMISSION_SCOPES.DM_READ,
  [RPC_METHODS.GET_DM_UNREAD_COUNT]: PERMISSION_SCOPES.DM_READ,
  [RPC_METHODS.MARK_AS_READ]: PERMISSION_SCOPES.DM_MANAGE,
  [RPC_METHODS.GET_INVOICES]: PERMISSION_SCOPES.INVOICE_READ,
  [RPC_METHODS.GET_INVOICE_STATUS]: PERMISSION_SCOPES.INVOICE_READ
};
var INTENT_PERMISSIONS = {
  [INTENT_ACTIONS.SEND]: PERMISSION_SCOPES.TRANSFER_REQUEST,
  [INTENT_ACTIONS.DM]: PERMISSION_SCOPES.DM_REQUEST,
  [INTENT_ACTIONS.PAYMENT_REQUEST]: PERMISSION_SCOPES.PAYMENT_REQUEST,
  [INTENT_ACTIONS.RECEIVE]: PERMISSION_SCOPES.IDENTITY_READ,
  [INTENT_ACTIONS.SIGN_MESSAGE]: PERMISSION_SCOPES.SIGN_REQUEST,
  [INTENT_ACTIONS.CREATE_INVOICE]: PERMISSION_SCOPES.INVOICE_WRITE,
  [INTENT_ACTIONS.CLOSE_INVOICE]: PERMISSION_SCOPES.INVOICE_WRITE,
  [INTENT_ACTIONS.CANCEL_INVOICE]: PERMISSION_SCOPES.INVOICE_WRITE,
  [INTENT_ACTIONS.PAY_INVOICE]: PERMISSION_SCOPES.TRANSFER_REQUEST,
  [INTENT_ACTIONS.RETURN_INVOICE_PAYMENT]: PERMISSION_SCOPES.TRANSFER_REQUEST,
  [INTENT_ACTIONS.IMPORT_INVOICE]: PERMISSION_SCOPES.INVOICE_WRITE,
  [INTENT_ACTIONS.SEND_INVOICE_RECEIPTS]: PERMISSION_SCOPES.INVOICE_WRITE,
  [INTENT_ACTIONS.SEND_CANCELLATION_NOTICES]: PERMISSION_SCOPES.INVOICE_WRITE,
  [INTENT_ACTIONS.SET_AUTO_RETURN]: PERMISSION_SCOPES.INVOICE_WRITE,
  [INTENT_ACTIONS.MINT]: PERMISSION_SCOPES.MINT_REQUEST
};

// transport/websocket.ts
var WebSocketReadyState = {
  CONNECTING: 0,
  OPEN: 1,
  CLOSING: 2,
  CLOSED: 3
};

// impl/nodejs/connect/WebSocketTransport.ts
var WebSocketServerTransport = class {
  server = null;
  // WebSocketServer from 'ws' package
  clientSocket = null;
  handlers = /* @__PURE__ */ new Set();
  config;
  constructor(config) {
    this.config = config;
  }
  /** Start the WebSocket server. Must be called before use. */
  async start() {
    const { WebSocketServer } = await import("ws");
    const wss = new WebSocketServer({
      port: this.config.port,
      host: this.config.host ?? "0.0.0.0"
    });
    this.server = wss;
    wss.on("connection", (ws) => {
      if (this.clientSocket) {
        ws.close(4e3, "Another client is already connected");
        return;
      }
      this.clientSocket = ws;
      ws.onmessage = (event) => {
        try {
          const msg = JSON.parse(typeof event.data === "string" ? event.data : String(event.data));
          if (isSphereConnectMessage(msg)) {
            for (const handler of this.handlers) {
              try {
                handler(msg);
              } catch (err) {
                logger.debug("WebSocket", "Message handler error", err);
              }
            }
          }
        } catch (err) {
          logger.debug("WebSocket", "Malformed message received", err);
        }
      };
      ws.onclose = () => {
        if (this.clientSocket === ws) {
          this.clientSocket = null;
        }
      };
    });
    await new Promise((resolve, reject) => {
      wss.on("listening", resolve);
      wss.on("error", reject);
    });
  }
  send(message) {
    if (this.clientSocket && this.clientSocket.readyState === WebSocketReadyState.OPEN) {
      this.clientSocket.send(JSON.stringify(message));
    }
  }
  onMessage(handler) {
    this.handlers.add(handler);
    return () => {
      this.handlers.delete(handler);
    };
  }
  destroy() {
    if (this.clientSocket) {
      this.clientSocket.close();
      this.clientSocket = null;
    }
    if (this.server) {
      this.server.close();
      this.server = null;
    }
    this.handlers.clear();
  }
};
var WebSocketClientTransport = class {
  ws = null;
  handlers = /* @__PURE__ */ new Set();
  config;
  reconnectAttempts = 0;
  reconnectTimer = null;
  destroyed = false;
  constructor(config) {
    this.config = {
      autoReconnect: true,
      reconnectDelayMs: 2e3,
      maxReconnectDelayMs: 3e4,
      maxReconnectAttempts: 10,
      ...config
    };
  }
  /** Connect to the WebSocket server. Must be called before use. */
  async connect() {
    return this.doConnect();
  }
  send(message) {
    if (this.ws && this.ws.readyState === WebSocketReadyState.OPEN) {
      this.ws.send(JSON.stringify(message));
    }
  }
  onMessage(handler) {
    this.handlers.add(handler);
    return () => {
      this.handlers.delete(handler);
    };
  }
  destroy() {
    this.destroyed = true;
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
    this.handlers.clear();
  }
  // ===========================================================================
  // Private
  // ===========================================================================
  doConnect() {
    return new Promise((resolve, reject) => {
      try {
        this.ws = this.config.createWebSocket(this.config.url);
      } catch (err) {
        reject(err);
        return;
      }
      this.ws.onopen = () => {
        this.reconnectAttempts = 0;
        resolve();
      };
      this.ws.onmessage = (event) => {
        try {
          const msg = JSON.parse(event.data);
          if (isSphereConnectMessage(msg)) {
            for (const handler of this.handlers) {
              try {
                handler(msg);
              } catch (err) {
                logger.debug("WebSocket", "Message handler error", err);
              }
            }
          }
        } catch (err) {
          logger.debug("WebSocket", "Malformed message received", err);
        }
      };
      this.ws.onerror = (err) => {
        reject(err);
      };
      this.ws.onclose = () => {
        this.ws = null;
        if (!this.destroyed && this.config.autoReconnect) {
          this.scheduleReconnect();
        }
      };
    });
  }
  scheduleReconnect() {
    const maxAttempts = this.config.maxReconnectAttempts;
    if (maxAttempts > 0 && this.reconnectAttempts >= maxAttempts) {
      return;
    }
    this.reconnectAttempts++;
    const baseDelay = this.config.reconnectDelayMs;
    const maxDelay = this.config.maxReconnectDelayMs;
    const delay = Math.min(baseDelay * Math.pow(2, this.reconnectAttempts - 1), maxDelay);
    this.reconnectTimer = setTimeout(() => {
      this.reconnectTimer = null;
      this.doConnect().catch((err) => logger.debug("WebSocket", "Reconnect attempt failed", err));
    }, delay);
  }
};
var WebSocketTransport = {
  /** Create a WebSocket server transport (wallet side) */
  createServer(config) {
    return new WebSocketServerTransport(config);
  },
  /** Create a WebSocket client transport (dApp side) */
  createClient(config) {
    return new WebSocketClientTransport(config);
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  WebSocketClientTransport,
  WebSocketServerTransport,
  WebSocketTransport
});
