// @unicitylabs/state-transition-sdk/lib/predicate/builtin/SignaturePredicate.js
import { SigningService } from '../../crypto/secp256k1/SigningService.js';
import { CborDeserializer } from '../../serialization/cbor/CborDeserializer.js';
import { CborSerializer } from '../../serialization/cbor/CborSerializer.js';
import { HexConverter } from '../../util/HexConverter.js';
import { dedent } from '../../util/StringUtils.js';
import { PredicateEngine } from '../PredicateEngine.js';
import { BuiltInPredicateType } from './BuiltInPredicateType.js';
/**
 * Built-in predicate that locks a state to a single secp256k1 public key.
 * Spending requires a {@link SignaturePredicateUnlockScript} signed by the
 * matching private key.
 */
export class SignaturePredicate {
    _publicKey;
    constructor(_publicKey) {
        this._publicKey = _publicKey;
        this._publicKey = new Uint8Array(_publicKey);
    }
    /**
     * @returns {PredicateEngine} Built-in predicate engine.
     */
    get engine() {
        return PredicateEngine.BUILT_IN;
    }
    /**
     * @returns {Uint8Array} Copy of the compressed public key.
     */
    get publicKey() {
        return new Uint8Array(this._publicKey);
    }
    /**
     * @returns {BuiltInPredicateType} Signature predicate type id.
     */
    get type() {
        return BuiltInPredicateType.Signature;
    }
    /**
     * Create a SignaturePredicate from a compressed secp256k1 public key.
     *
     * @param {Uint8Array} publicKey Compressed public key bytes.
     * @returns {SignaturePredicate} New predicate.
     * @throws {Error} If `publicKey` is not a valid compressed point.
     */
    static create(publicKey) {
        if (!SigningService.isPublicKeyValid(publicKey)) {
            throw new Error('Invalid public key.');
        }
        return new SignaturePredicate(publicKey);
    }
    /**
     * Decode a SignaturePredicate from an EncodedPredicate.
     *
     * @param {EncodedPredicate} predicate Encoded predicate.
     * @returns {SignaturePredicate} Decoded predicate.
     * @throws {Error} If the engine or type does not match.
     */
    static fromPredicate(predicate) {
        if (predicate.engine !== PredicateEngine.BUILT_IN) {
            throw new Error(`Predicate engine must be ${PredicateEngine.BUILT_IN}.`);
        }
        const type = CborDeserializer.decodeUnsignedInteger(predicate.encodeCode());
        if (type !== BigInt(BuiltInPredicateType.Signature)) {
            throw new Error(`Predicate type must be ${BuiltInPredicateType.Signature}.`);
        }
        return new SignaturePredicate(predicate.encodeParameters());
    }
    /**
     * Create a SignaturePredicate from the public key of a SigningService.
     *
     * @param {SigningService} signingService Service whose public key is used.
     * @returns {SignaturePredicate} New predicate.
     */
    static fromSigningService(signingService) {
        return new SignaturePredicate(signingService.publicKey);
    }
    /**
     * @inheritDoc
     */
    encodeCode() {
        return CborSerializer.encodeUnsignedInteger(this.type);
    }
    /**
     * @inheritDoc
     */
    encodeParameters() {
        return this.publicKey;
    }
    /**
     * @returns {string} String representation of the predicate.
     */
    toString() {
        return dedent `
      SignaturePredicate
        Public Key: ${HexConverter.encode(this._publicKey)}`;
    }
}
