// @unicitylabs/state-transition-sdk/lib/predicate/builtin/UnicityIdPredicate.js
import { CborDeserializer } from '../../serialization/cbor/CborDeserializer.js';
import { CborSerializer } from '../../serialization/cbor/CborSerializer.js';
import { dedent } from '../../util/StringUtils.js';
import { PredicateEngine } from '../PredicateEngine.js';
import { BuiltInPredicateType } from './BuiltInPredicateType.js';
import { UnicityId } from '../../unicity-id/UnicityId.js';
/**
 * Built-in predicate that locks a state to a specific {@link UnicityId}.
 * Spending requires presenting a matching unicity-id token in the unlock
 * script.
 */
export class UnicityIdPredicate {
    unicityId;
    constructor(unicityId) {
        this.unicityId = unicityId;
    }
    /**
     * @returns {PredicateEngine} Built-in predicate engine.
     */
    get engine() {
        return PredicateEngine.BUILT_IN;
    }
    /**
     * @returns {BuiltInPredicateType} UnicityId predicate type id.
     */
    get type() {
        return BuiltInPredicateType.UnicityId;
    }
    /**
     * Create a UnicityIdPredicate for the given UnicityId.
     *
     * @param {UnicityId} unicityId Unicity id the state is locked to.
     * @returns {UnicityIdPredicate} New predicate.
     */
    static create(unicityId) {
        return new UnicityIdPredicate(unicityId);
    }
    /**
     * Decode a UnicityIdPredicate from an EncodedPredicate.
     *
     * @param {EncodedPredicate} predicate Encoded predicate.
     * @returns {UnicityIdPredicate} Decoded predicate.
     * @throws {Error} If the engine or type does not match.
     */
    static fromPredicate(predicate) {
        if (predicate.engine !== PredicateEngine.BUILT_IN) {
            throw new Error(`Predicate engine must be ${PredicateEngine.BUILT_IN}.`);
        }
        const type = CborDeserializer.decodeUnsignedInteger(predicate.encodeCode());
        if (type !== BigInt(BuiltInPredicateType.UnicityId)) {
            throw new Error(`Predicate type must be ${BuiltInPredicateType.UnicityId}.`);
        }
        return new UnicityIdPredicate(UnicityId.fromCBOR(predicate.encodeParameters()));
    }
    /**
     * @inheritDoc
     */
    encodeCode() {
        return CborSerializer.encodeUnsignedInteger(this.type);
    }
    /**
     * @inheritDoc
     */
    encodeParameters() {
        return this.unicityId.toCBOR();
    }
    /**
     * @returns {string} String representation of the predicate.
     */
    toString() {
        return dedent `
      UnicityIdPredicate
        UnicityId: 
          ${this.unicityId.toString()}`;
    }
}
