// @unicitylabs/state-transition-sdk/lib/predicate/builtin/UnicityIdPredicateUnlockScript.js
import { SignaturePredicateUnlockScript } from './SignaturePredicateUnlockScript.js';
import { UnicityIdPredicate } from './UnicityIdPredicate.js';
import { CborDeserializer } from '../../serialization/cbor/CborDeserializer.js';
import { CborSerializer } from '../../serialization/cbor/CborSerializer.js';
import { UnicityIdToken } from '../../unicity-id/UnicityIdToken.js';
/**
 * Unlock script for {@link UnicityIdPredicate}: a signed unlock script
 * bundled with the unicity-id token that authorizes the spend.
 */
export class UnicityIdPredicateUnlockScript {
    _unlockScript;
    token;
    constructor(_unlockScript, token) {
        this._unlockScript = _unlockScript;
        this.token = token;
        this._unlockScript = new Uint8Array(_unlockScript);
    }
    /**
     * @returns {Uint8Array} Copy of the inner signature unlock script bytes.
     */
    get unlockScript() {
        return new Uint8Array(this._unlockScript);
    }
    /**
     * Build an unlock script that proves the spender owns `token`, the
     * unicity-id token bound to the predicate.
     *
     * @param {UnicityIdToken} token Unicity-id token authorizing the spend.
     * @param {ITransaction} transaction Transaction being unlocked.
     * @param {SigningService} signingService Service holding the private key.
     * @returns {Promise<UnicityIdPredicateUnlockScript>} Signed unlock script.
     * @throws {Error} If the token's id does not match the predicate's UnicityId.
     */
    static async create(token, transaction, signingService) {
        const predicate = UnicityIdPredicate.fromPredicate(transaction.lockScript);
        if (!token.id.equals(await predicate.unicityId.toTokenId())) {
            throw new Error('Invalid Unicity ID for transaction');
        }
        const unlockScript = await SignaturePredicateUnlockScript.create(transaction, signingService);
        return new UnicityIdPredicateUnlockScript(unlockScript.encode(), token);
    }
    /**
     * Decode a UnicityIdPredicateUnlockScript from CBOR bytes.
     *
     * @param {Uint8Array} bytes CBOR bytes.
     * @returns {Promise<UnicityIdPredicateUnlockScript>} Decoded unlock script.
     */
    static async decode(bytes) {
        const [unlockScriptBytes, tokenBytes] = CborDeserializer.decodeArray(bytes);
        return new UnicityIdPredicateUnlockScript(CborDeserializer.decodeByteString(unlockScriptBytes), await UnicityIdToken.fromCBOR(tokenBytes));
    }
    /**
     * @inheritDoc
     */
    encode() {
        return CborSerializer.encodeArray(CborSerializer.encodeByteString(this.unlockScript), this.token.toCBOR());
    }
}
