// @unicitylabs/state-transition-sdk/lib/payment/asset/PaymentAssetCollection.js
import { Asset } from './Asset.js';
import { CborDeserializer } from '../../serialization/cbor/CborDeserializer.js';
import { CborSerializer } from '../../serialization/cbor/CborSerializer.js';
import { HexConverter } from '../../util/HexConverter.js';
/**
 * Asset-id-keyed collection of {@link Asset} values used in a payment.
 */
export class PaymentAssetCollection {
    _assets;
    constructor(_assets) {
        this._assets = _assets;
    }
    /**
     * Create a PaymentAssetCollection.
     *
     * @param {...Asset} data Assets to include.
     * @returns {PaymentAssetCollection} New collection.
     * @throws {Error} If any asset id appears more than once.
     */
    static create(...data) {
        const assets = new Map();
        for (const asset of data) {
            const key = HexConverter.encode(asset.id.bytes);
            if (assets.has(key)) {
                throw new Error('Invalid payment asset collection. Duplicate assets found.');
            }
            assets.set(key, asset);
        }
        return new PaymentAssetCollection(assets);
    }
    /**
     * Create PaymentAssetCollection from CBOR bytes.
     *
     * @param {Uint8Array} bytes CBOR bytes.
     * @returns {PaymentAssetCollection} Decoded collection.
     */
    static fromCBOR(bytes) {
        const data = CborDeserializer.decodeArray(bytes);
        const assets = [];
        for (const asset of data) {
            assets.push(Asset.fromCBOR(asset));
        }
        return PaymentAssetCollection.create(...assets);
    }
    /**
     * Look up the asset with the given id.
     *
     * @param {AssetId} id Asset id.
     * @returns {Asset|null} Matching asset, or `null`.
     */
    get(id) {
        return this._assets.get(HexConverter.encode(id.bytes)) ?? null;
    }
    /**
     * @returns {number} Number of assets in this collection.
     */
    size() {
        return this._assets.size;
    }
    /**
     * @returns {Asset[]} Assets in insertion order.
     */
    toArray() {
        return Array.from(this._assets.values());
    }
    /**
     * Convert PaymentAssetCollection to CBOR bytes.
     *
     * @returns {Uint8Array} CBOR bytes.
     */
    toCBOR() {
        return CborSerializer.encodeArray(...this.toArray().map((asset) => asset.toCBOR()));
    }
}
