// @multiformats/dns/dist/src/dns.js
import { CustomProgressEvent } from 'progress-events';
import { DNSQueryFailedError, EmptyDNSAnswerError } from "./errors.js";
import { defaultResolver } from "./resolvers/default.js";
import { cache } from "./utils/cache.js";
import { getTypes } from "./utils/get-types.js";
const DEFAULT_ANSWER_CACHE_SIZE = 1000;
const RANDOM = () => (Math.random() > 0.5) ? -1 : 1;
export class DNS {
    resolvers;
    cache;
    logger;
    sorter;
    constructor(init) {
        this.resolvers = {};
        this.cache = cache(init.cacheSize ?? DEFAULT_ANSWER_CACHE_SIZE);
        this.logger = init.logger;
        this.sorter = init.sorter ?? RANDOM;
        Object.entries(init.resolvers ?? {}).forEach(([tld, resolver]) => {
            if (!Array.isArray(resolver)) {
                resolver = [resolver];
            }
            // convert `com` -> `com.`
            if (!tld.endsWith('.')) {
                tld = `${tld}.`;
            }
            this.resolvers[tld] = resolver;
        });
        // configure default resolver if none specified
        if (this.resolvers['.'] == null) {
            this.resolvers['.'] = defaultResolver();
        }
    }
    /**
     * Queries DNS resolvers for the passed record types for the passed domain.
     *
     * If cached records exist for all desired types they will be returned
     * instead.
     *
     * Any new responses will be added to the cache for subsequent requests.
     */
    async query(domain, options = {}) {
        const types = getTypes(options.types);
        const cached = options.cached !== false ? this.cache.get(domain, types) : undefined;
        if (cached != null) {
            options.onProgress?.(new CustomProgressEvent('dns:cache', cached));
            return cached;
        }
        const tld = `${domain.split('.').pop()}.`;
        const resolvers = (this.resolvers[tld] ?? this.resolvers['.']).sort(this.sorter);
        const errors = [];
        for (const resolver of resolvers) {
            // skip further resolutions if the user aborted the signal
            if (options.signal?.aborted === true) {
                break;
            }
            try {
                const result = await resolver(domain, {
                    ...options,
                    logger: this.logger,
                    types
                });
                if (result.Answer.length === 0) {
                    throw new EmptyDNSAnswerError('Query result had no answers');
                }
                for (const answer of result.Answer) {
                    this.cache.add(domain, answer);
                }
                return result;
            }
            catch (err) {
                errors.push(err);
                options.onProgress?.(new CustomProgressEvent('dns:error', err));
            }
        }
        throw new DNSQueryFailedError(errors, `DNS lookup of ${domain} ${types} failed`);
    }
}
