// @unicitylabs/nostr-js-sdk/dist/esm/client/NostrClient.js
/**
 * NostrClient - Main entry point for Nostr protocol operations.
 * Handles relay connections, event publishing, and subscriptions.
 */
import { Event } from '../protocol/Event.js';
import * as EventKinds from '../protocol/EventKinds.js';
import { createWebSocket, extractMessageData, OPEN, CLOSED, } from './WebSocketAdapter.js';
import * as NIP17 from '../messaging/nip17.js';
import { createBindingEvent, createNametagToPubkeyFilter, createAddressToBindingFilter, createIdentityBindingEvent, parseBindingInfo, hasNametagOwnershipMarker, } from '../nametag/NametagBinding.js';
/** Connection timeout in milliseconds */
const CONNECTION_TIMEOUT_MS = 30000;
/** Default options */
const DEFAULT_QUERY_TIMEOUT_MS = 5000;
const DEFAULT_RECONNECT_INTERVAL_MS = 1000;
const DEFAULT_MAX_RECONNECT_INTERVAL_MS = 30000;
const DEFAULT_PING_INTERVAL_MS = 30000;
/**
 * Internal sub_id reserved for the keepalive REQ. Namespaced with a
 * `__nostr-sdk-` prefix so that user code calling
 * {@link NostrClient.subscribe} with an explicit `subscriptionId`
 * cannot collide — a user choosing the literal `"ping"` would
 * otherwise have their subscription forcibly CLOSE/REQ'd every
 * ping interval. The leading `__` is a stable convention for
 * "do not pick this name."
 */
const PING_SUB_ID = '__nostr-sdk-keepalive__';
/**
 * Filter id used by the keepalive REQ. We need a filter the relay can
 * resolve immediately (so EOSE comes back fast = relay is alive), but
 * which can NOT match any real event past EOSE (so the live tail stays
 * empty).
 *
 * Scoping by `authors:[selfPubkey]` was the original approach but it
 * matched every event the wallet itself published — including kind-31113
 * token transfers — which the relay then echoed back on the keepalive
 * sub. Some relays dedupe events across overlapping subs, so the
 * wallet's own consumer subscription would not receive its own echo and
 * any flow waiting on that echo would time out.
 *
 * The filter `{ ids: ['00...00'] }` asks the relay for a single event
 * whose id is exactly the all-zero hash. Real Nostr event ids are
 * SHA-256 over a canonical JSON serialization, so the all-zero hash is
 * unreachable in practice. Result: instant EOSE, empty live tail.
 */
const KEEPALIVE_NEVER_MATCH_ID = '0'.repeat(64);
/**
 * Delay before resubscribing after NIP-42 authentication.
 * This gives the relay time to process the AUTH response before we send
 * subscription requests. Without this delay, some relays may still reject
 * the subscriptions as the AUTH hasn't been fully processed yet.
 */
const AUTH_RESUBSCRIBE_DELAY_MS = 100;
/**
 * NostrClient provides the main interface for Nostr protocol operations.
 */
export class NostrClient {
    keyManager;
    relays = new Map();
    subscriptions = new Map();
    eventQueue = [];
    pendingOks = new Map();
    subscriptionCounter = 0;
    closed = false;
    // Configuration options
    queryTimeoutMs;
    autoReconnect;
    reconnectIntervalMs;
    maxReconnectIntervalMs;
    pingIntervalMs;
    // Connection event listeners
    connectionListeners = [];
    /**
     * Create a NostrClient instance.
     * @param keyManager Key manager with signing keys
     * @param options Optional configuration options
     */
    constructor(keyManager, options) {
        this.keyManager = keyManager;
        this.queryTimeoutMs = options?.queryTimeoutMs ?? DEFAULT_QUERY_TIMEOUT_MS;
        this.autoReconnect = options?.autoReconnect ?? true;
        this.reconnectIntervalMs = options?.reconnectIntervalMs ?? DEFAULT_RECONNECT_INTERVAL_MS;
        this.maxReconnectIntervalMs = options?.maxReconnectIntervalMs ?? DEFAULT_MAX_RECONNECT_INTERVAL_MS;
        this.pingIntervalMs = options?.pingIntervalMs ?? DEFAULT_PING_INTERVAL_MS;
    }
    /**
     * Replace the key manager used for signing and encryption.
     *
     * The connection stays alive — but every operation that consults the
     * key manager from this point on uses the new key, including:
     *   - signing future published events,
     *   - signing NIP-42 AUTH challenge responses,
     *   - the `authors:[selfPubkey]` filter on the keepalive ping REQ
     *     (computed each ping interval),
     *   - any other code path that calls `getPublicKeyHex()` on the
     *     stored manager.
     *
     * Existing in-flight subscriptions are not re-issued or re-keyed.
     * @param keyManager New key manager
     */
    setKeyManager(keyManager) {
        this.keyManager = keyManager;
    }
    /**
     * Get the current key manager.
     */
    getKeyManager() {
        return this.keyManager;
    }
    /**
     * Add a connection event listener.
     * @param listener Listener for connection events
     */
    addConnectionListener(listener) {
        this.connectionListeners.push(listener);
    }
    /**
     * Remove a connection event listener.
     * @param listener Listener to remove
     */
    removeConnectionListener(listener) {
        const index = this.connectionListeners.indexOf(listener);
        if (index !== -1) {
            this.connectionListeners.splice(index, 1);
        }
    }
    /**
     * Emit a connection event to all listeners.
     */
    emitConnectionEvent(eventType, relayUrl, extra) {
        for (const listener of this.connectionListeners) {
            try {
                switch (eventType) {
                    case 'connect':
                        listener.onConnect?.(relayUrl);
                        break;
                    case 'disconnect':
                        listener.onDisconnect?.(relayUrl, extra);
                        break;
                    case 'reconnecting':
                        listener.onReconnecting?.(relayUrl, extra);
                        break;
                    case 'reconnected':
                        listener.onReconnected?.(relayUrl);
                        break;
                }
            }
            catch {
                // Ignore listener errors
            }
        }
    }
    /**
     * Get the current query timeout in milliseconds.
     * @returns Query timeout in milliseconds
     */
    getQueryTimeout() {
        return this.queryTimeoutMs;
    }
    /**
     * Set the query timeout for nametag lookups and other queries.
     * @param timeoutMs Timeout in milliseconds
     */
    setQueryTimeout(timeoutMs) {
        this.queryTimeoutMs = timeoutMs;
    }
    /**
     * Connect to one or more relay WebSocket URLs.
     * @param relayUrls Relay URLs to connect to
     * @returns Promise that resolves when all connections are established
     */
    async connect(...relayUrls) {
        if (this.closed) {
            throw new Error('Client has been disconnected');
        }
        const connectionPromises = relayUrls.map((url) => this.connectToRelay(url));
        await Promise.all(connectionPromises);
    }
    /**
     * Connect to a single relay.
     * @param isReconnect Whether this is a reconnection attempt
     */
    async connectToRelay(url, isReconnect = false) {
        const existingRelay = this.relays.get(url);
        if (existingRelay?.connected) {
            return;
        }
        return new Promise((resolve, reject) => {
            // The connection-setup timeout has three races to defend
            // against:
            //   A) createWebSocket resolves AFTER the timeout fired.
            //   B) createWebSocket resolves BEFORE the timeout, but
            //      `onopen` fires AFTER the timeout fired.
            //   C) createWebSocket resolves and `onopen` fires BEFORE the
            //      timeout (the success path).
            // `pendingSocket` lets the timeout proactively close any
            // socket that's already been created but hasn't fired
            // `onopen` yet. The `timedOut` flag covers (A) inside `.then`
            // and (B) inside `socket.onopen`.
            let timedOut = false;
            let pendingSocket = null;
            const timeoutId = setTimeout(() => {
                timedOut = true;
                if (pendingSocket) {
                    try {
                        pendingSocket.close(1000, 'Connection setup timed out');
                    }
                    catch { /* ignore */ }
                }
                reject(new Error(`Connection to ${url} timed out`));
            }, CONNECTION_TIMEOUT_MS);
            createWebSocket(url)
                .then((socket) => {
                if (timedOut) {
                    // Caller already saw the rejection. Discard the late
                    // socket so we don't leak it.
                    try {
                        socket.close(1000, 'Connection setup timed out');
                    }
                    catch { /* ignore */ }
                    return;
                }
                pendingSocket = socket;
                const relay = {
                    url,
                    socket,
                    connected: false,
                    reconnecting: false,
                    reconnectAttempts: 0,
                    reconnectTimer: null,
                    pingTimer: null,
                    lastPongTime: Date.now(),
                    unansweredPings: 0,
                    wasConnected: existingRelay?.wasConnected ?? false,
                    // Reset on every new connection: a relay's per-connection
                    // sub-slot accounting is fresh, so previously-rejected REQs
                    // should be re-issued on the new socket.
                    closedSubIds: new Set(),
                    eosedSubIds: new Set(),
                };
                socket.onopen = () => {
                    // The `.then` block already guards against a socket
                    // arriving after the connection timeout, but the socket
                    // can also be created BEFORE the timeout while
                    // `onopen` fires AFTER the timeout has rejected the
                    // outer promise. Without this second guard we'd register
                    // the relay, start a pingTimer, and resubscribe — orphan
                    // background resources the caller can't see or clean up
                    // because their connect() call already saw a rejection.
                    if (timedOut) {
                        try {
                            socket.close(1000, 'Connection setup timed out');
                        }
                        catch { /* ignore */ }
                        return;
                    }
                    clearTimeout(timeoutId);
                    relay.connected = true;
                    relay.reconnectAttempts = 0; // Reset on successful connection
                    relay.lastPongTime = Date.now();
                    this.relays.set(url, relay);
                    // Emit appropriate connection event
                    if (isReconnect && relay.wasConnected) {
                        this.emitConnectionEvent('reconnected', url);
                    }
                    else {
                        this.emitConnectionEvent('connect', url);
                    }
                    relay.wasConnected = true;
                    // Start ping health check
                    this.startPingTimer(url);
                    // Re-establish subscriptions
                    this.resubscribeAll(url);
                    // Flush queued events
                    this.flushEventQueue();
                    resolve();
                };
                socket.onmessage = (event) => {
                    try {
                        const data = extractMessageData(event);
                        // Update last pong time and reset unanswered pings on any message (relay is alive)
                        const r = this.relays.get(url);
                        if (r) {
                            r.lastPongTime = Date.now();
                            r.unansweredPings = 0;
                        }
                        this.handleRelayMessage(url, data);
                    }
                    catch (error) {
                        console.error(`Error handling message from ${url}:`, error);
                    }
                };
                socket.onclose = (event) => {
                    const wasConnected = relay.connected;
                    relay.connected = false;
                    this.stopPingTimer(url);
                    // Pre-onopen close: TCP handshake failure or relay
                    // immediately closed the WS during the upgrade. Without
                    // this, the connectToRelay promise stays pending until
                    // CONNECTION_TIMEOUT_MS (30s) expires; surfacing it now
                    // lets the caller see the failure promptly and retry.
                    if (!wasConnected && !timedOut) {
                        timedOut = true;
                        clearTimeout(timeoutId);
                        reject(new Error(`Connection to ${url} closed during handshake: ${event?.reason || 'no reason'}`));
                    }
                    if (wasConnected) {
                        const reason = event?.reason || 'Connection closed';
                        this.emitConnectionEvent('disconnect', url, reason);
                        // Re-trigger the all-done check on every active sub.
                        // queryWithFirstSeenWins.allRelaysDoneFor only runs
                        // from listener callbacks (EOSE / CLOSED via onError);
                        // a socket that drops without sending either would
                        // otherwise leave the query hanging until
                        // queryTimeoutMs even though the disconnected relay no
                        // longer counts toward "still pending" relays. Firing
                        // a synthetic onError gives every active sub a chance
                        // to re-evaluate now that the relay set has shrunk.
                        // Include the relay URL so listeners in a multi-relay
                        // client can attribute which relay dropped.
                        const inflight = Array.from(this.subscriptions.entries());
                        for (const [subId, sub] of inflight) {
                            try {
                                sub.listener.onError?.(subId, `Relay disconnected (${url}): ${reason}`);
                            }
                            catch {
                                // Ignore listener errors — we're notifying
                                // best-effort.
                            }
                        }
                    }
                    if (!this.closed && this.autoReconnect && !relay.reconnecting) {
                        this.scheduleReconnect(url);
                    }
                };
                socket.onerror = (error) => {
                    if (!relay.connected) {
                        clearTimeout(timeoutId);
                        reject(new Error(`Failed to connect to ${url}: ${error.message || 'Unknown error'}`));
                    }
                };
                // Note: we do NOT register the relay in `this.relays` here —
                // only after `onopen` fires successfully. Registering eagerly
                // (before onopen) would leak the relay into the global map
                // even when the connection setup times out and the caller's
                // promise has already rejected.
            })
                .catch((error) => {
                clearTimeout(timeoutId);
                reject(error);
            });
        });
    }
    /**
     * Schedule a reconnection attempt for a relay with exponential backoff.
     */
    scheduleReconnect(url) {
        const relay = this.relays.get(url);
        if (!relay || this.closed || !this.autoReconnect)
            return;
        // Clear any existing reconnect timer
        if (relay.reconnectTimer) {
            clearTimeout(relay.reconnectTimer);
        }
        relay.reconnecting = true;
        relay.reconnectAttempts++;
        // Calculate delay with exponential backoff
        const baseDelay = this.reconnectIntervalMs;
        const exponentialDelay = baseDelay * Math.pow(2, relay.reconnectAttempts - 1);
        const delay = Math.min(exponentialDelay, this.maxReconnectIntervalMs);
        this.emitConnectionEvent('reconnecting', url, relay.reconnectAttempts);
        relay.reconnectTimer = setTimeout(async () => {
            if (this.closed)
                return;
            relay.reconnectTimer = null;
            try {
                relay.reconnecting = false;
                await this.connectToRelay(url, true);
            }
            catch {
                // Connection failed, schedule another attempt
                if (!this.closed && this.autoReconnect) {
                    this.scheduleReconnect(url);
                }
            }
        }, delay);
    }
    /**
     * Start the ping timer for a relay to detect stale connections.
     */
    startPingTimer(url) {
        if (this.pingIntervalMs <= 0)
            return;
        const relay = this.relays.get(url);
        if (!relay)
            return;
        // Stop existing timer if any
        this.stopPingTimer(url);
        relay.pingTimer = setInterval(() => {
            if (!relay.connected || !relay.socket) {
                this.stopPingTimer(url);
                return;
            }
            const timeSinceLastPong = Date.now() - relay.lastPongTime;
            if (timeSinceLastPong > this.pingIntervalMs * 2 && relay.unansweredPings >= 2) {
                // No inbound message for 2x the ping interval AND we've sent at least 2 pings
                // without any response — the connection is truly stale.
                // The unanswered pings gate handles browser tab throttling: on the first tick
                // after waking, unansweredPings is 0, so we send a ping and wait. If the relay
                // is alive it responds (resetting the counter). If dead, subsequent ticks
                // increment the counter until it reaches the threshold, even under sustained
                // throttling where intervals are irregular.
                console.warn(`Relay ${url} appears stale (no response for ${timeSinceLastPong}ms, ${relay.unansweredPings} unanswered pings), reconnecting...`);
                this.stopPingTimer(url);
                try {
                    relay.socket.close();
                }
                catch {
                    // Ignore close errors
                }
                return;
            }
            // Send a subscription request as a ping (relays respond with EOSE).
            // The filter MUST be tightly scoped to something no real event can
            // match — both for the initial query AND the post-EOSE live tail.
            //
            // Earlier iterations used `authors:[self]` reasoning that "the
            // relay would only forward our own future events". That reasoning
            // was wrong: it precisely DOES forward every event the wallet
            // publishes, including kind-31113 token transfers. Some relays
            // dedupe events across overlapping subs, so the wallet's own
            // consumer subscription would miss its echo and any flow waiting
            // on it would time out.
            //
            // {@link KEEPALIVE_NEVER_MATCH_ID} (the all-zero SHA-256 hash) is
            // unreachable in real event-id space, so the relay returns EOSE
            // immediately and the live tail never matches.
            try {
                // First close any existing ping subscription to ensure we don't accumulate
                const closeMessage = JSON.stringify(['CLOSE', PING_SUB_ID]);
                relay.socket.send(closeMessage);
                // Then send the new ping request (limit:1 ensures relay sends EOSE)
                const pingMessage = JSON.stringify([
                    'REQ',
                    PING_SUB_ID,
                    { ids: [KEEPALIVE_NEVER_MATCH_ID], limit: 1 },
                ]);
                relay.socket.send(pingMessage);
                relay.unansweredPings++;
            }
            catch {
                // Send failed, connection likely dead
                console.warn(`Ping to ${url} failed, reconnecting...`);
                this.stopPingTimer(url);
                try {
                    relay.socket.close();
                }
                catch {
                    // Ignore close errors
                }
            }
        }, this.pingIntervalMs);
    }
    /**
     * Stop the ping timer for a relay.
     */
    stopPingTimer(url) {
        const relay = this.relays.get(url);
        if (relay?.pingTimer) {
            clearInterval(relay.pingTimer);
            relay.pingTimer = null;
        }
    }
    /**
     * Re-establish all subscriptions for a relay.
     */
    resubscribeAll(url) {
        const relay = this.relays.get(url);
        if (!relay?.socket || !relay.connected)
            return;
        for (const [subId, info] of this.subscriptions) {
            // Skip subs this relay has previously CLOSED — re-issuing them
            // just triggers the same rejection in a loop. Other healthy
            // relays still resubscribe.
            if (relay.closedSubIds.has(subId))
                continue;
            const message = JSON.stringify(['REQ', subId, info.filter.toJSON()]);
            relay.socket.send(message);
        }
    }
    /**
     * Flush queued events to connected relays.
     */
    flushEventQueue() {
        const queue = [...this.eventQueue];
        this.eventQueue = [];
        for (const item of queue) {
            this.broadcastEvent(item.event)
                .then(() => item.resolve(item.event.id))
                .catch(item.reject);
        }
    }
    /**
     * Handle a message from a relay.
     */
    handleRelayMessage(relayUrl, message) {
        try {
            const json = JSON.parse(message);
            if (!Array.isArray(json) || json.length < 2)
                return;
            const messageType = json[0];
            switch (messageType) {
                case 'EVENT':
                    this.handleEventMessage(json);
                    break;
                case 'OK':
                    this.handleOkMessage(json);
                    break;
                case 'EOSE':
                    this.handleEOSEMessage(relayUrl, json);
                    break;
                case 'NOTICE':
                    this.handleNoticeMessage(json);
                    break;
                case 'CLOSED':
                    this.handleClosedMessage(relayUrl, json);
                    break;
                case 'AUTH':
                    this.handleAuthMessage(relayUrl, json);
                    break;
            }
        }
        catch {
            // Ignore malformed messages
        }
    }
    /**
     * Handle EVENT message from relay.
     */
    handleEventMessage(json) {
        if (json.length < 3 || typeof json[1] !== 'string')
            return;
        const subscriptionId = json[1];
        const eventData = json[2];
        const subscription = this.subscriptions.get(subscriptionId);
        if (!subscription)
            return;
        try {
            const event = Event.fromJSON(eventData);
            subscription.listener.onEvent(event);
        }
        catch {
            // Ignore invalid events
        }
    }
    /**
     * Handle OK message from relay.
     */
    handleOkMessage(json) {
        if (json.length < 4)
            return;
        const eventId = json[1];
        const accepted = json[2];
        const message = json[3];
        const pending = this.pendingOks.get(eventId);
        if (!pending)
            return;
        clearTimeout(pending.timer);
        this.pendingOks.delete(eventId);
        if (accepted) {
            pending.resolve(eventId);
        }
        else {
            pending.reject(new Error(`Event rejected: ${message}`));
        }
    }
    /**
     * Handle EOSE (End of Stored Events) message from relay.
     *
     * Records the per-relay EOSE marker (mirroring closedSubIds) so
     * queryWithFirstSeenWins can decide when ALL connected relays have
     * finished — either streamed EOSE or rejected with CLOSED — instead
     * of settling off the first fast relay's EOSE while a slower relay
     * is still about to deliver matching events.
     */
    handleEOSEMessage(relayUrl, json) {
        if (json.length < 2 || typeof json[1] !== 'string')
            return;
        const subscriptionId = json[1];
        if (!this.subscriptions.has(subscriptionId))
            return;
        const relay = this.relays.get(relayUrl);
        if (relay) {
            relay.eosedSubIds.add(subscriptionId);
        }
        const subscription = this.subscriptions.get(subscriptionId);
        if (subscription?.listener.onEndOfStoredEvents) {
            subscription.listener.onEndOfStoredEvents(subscriptionId);
        }
    }
    /**
     * Handle NOTICE message from relay.
     */
    handleNoticeMessage(json) {
        if (json.length < 2)
            return;
        const message = json[1];
        console.warn('Relay notice:', message);
    }
    /**
     * Handle CLOSED message from relay (subscription closed by relay).
     *
     * NIP-01 CLOSED frames are terminal for the named subscription **on
     * the sending relay**. In a multi-relay client the same sub_id may
     * still be alive on a healthy relay, so we must NOT delete the
     * global `this.subscriptions` entry here — that would silently drop
     * EVENT/EOSE frames from the still-healthy relays in
     * `handleEventMessage` (which consults the global map).
     *
     * Instead we record the rejection on the sending relay's
     * `closedSubIds` set so `resubscribeAll` and post-AUTH resubscribe
     * skip it on this relay only. The listener is notified via
     * `onError` so callers (e.g. queryWithFirstSeenWins) can decide to
     * settle and explicitly `unsubscribe()` if they want to give up
     * across all relays.
     */
    handleClosedMessage(relayUrl, json) {
        // NIP-01 makes the message field optional: `["CLOSED", <sub>]` is
        // valid. Dropping such frames was exactly the leak this PR sets out
        // to fix — no closedSubIds marker and no onError notification means
        // queries hang until timeout and resubscribe loops persist.
        if (json.length < 2 || typeof json[1] !== 'string')
            return;
        const subscriptionId = json[1];
        // Ignore CLOSED for sub_ids we don't know about. A misbehaving or
        // malicious relay could otherwise spam us with arbitrary sub_ids
        // and grow `closedSubIds` unbounded over a long-lived connection,
        // and could pre-emptively block sub_ids we might use later.
        if (!this.subscriptions.has(subscriptionId))
            return;
        const message = typeof json[2] === 'string' ? json[2] : 'no reason provided';
        // NIP-42 transient case: relays that require AUTH typically reject
        // pre-auth REQs with `CLOSED("auth-required:...")` and then send
        // an AUTH challenge. resubscribeAfterAuth re-issues the sub, so
        // this rejection is NOT terminal. If we marked closedSubIds here,
        // queryWithFirstSeenWins.onError would settle the future
        // prematurely (single-relay → allRelaysDoneFor=true), unsubscribe
        // the sub, and the post-AUTH retry would find nothing to retry.
        // Listener still gets onError so callers see the reason; we just
        // don't poison the per-relay state with a transient marker.
        //
        // We accept three on-the-wire shapes: `auth-required:...`
        // (NIP-42 standard with reason), `auth-required ...` (whitespace
        // separator), and bare `auth-required` (no suffix at all — some
        // relays / tests).
        const isAuthRequired = message === 'auth-required'
            || message.startsWith('auth-required:')
            || message.startsWith('auth-required ');
        const relay = this.relays.get(relayUrl);
        if (relay && !isAuthRequired) {
            relay.closedSubIds.add(subscriptionId);
        }
        const subscription = this.subscriptions.get(subscriptionId);
        if (subscription?.listener.onError) {
            // Pass the relay's reason through verbatim so callers can
            // pattern-match on standard prefixes (`auth-required:`,
            // `rate-limited:`, `blocked:`, etc.) without parsing through
            // a wrapper string.
            subscription.listener.onError(subscriptionId, message);
        }
    }
    /**
     * Handle AUTH message from relay (NIP-42 authentication challenge).
     */
    handleAuthMessage(relayUrl, json) {
        if (json.length < 2)
            return;
        const challenge = json[1];
        const relay = this.relays.get(relayUrl);
        if (!relay?.socket || !relay.connected)
            return;
        // Create and sign the auth event (kind 22242)
        const authEvent = Event.create(this.keyManager, {
            kind: EventKinds.AUTH,
            tags: [
                ['relay', relayUrl],
                ['challenge', challenge],
            ],
            content: '',
        });
        // Send AUTH response
        const message = JSON.stringify(['AUTH', authEvent.toJSON()]);
        relay.socket.send(message);
        // Re-send subscriptions after auth (relay may have ignored pre-auth
        // requests). Two separate per-relay markers, two separate decisions:
        //
        //  - `closedSubIds`: do NOT clear. handleClosedMessage already
        //    skips the auth-required transient case, so anything in this
        //    set is a TERMINAL rejection (rate-limited, blocked, etc.)
        //    that AUTH does not relax. The resubscribeAll guard then
        //    correctly skips terminal-rejected subs on this relay. They
        //    will be retried on the next reconnect, when onopen creates a
        //    fresh RelayConnection with empty markers.
        //
        //  - `eosedSubIds`: clear. A relay may have EOSE'd a pre-auth sub
        //    with zero events (filter unsatisfiable without auth context);
        //    post-auth the same filter might match. We must re-arm the
        //    local "still waiting" state so any in-flight
        //    queryWithFirstSeenWins doesn't see this relay as already-done
        //    from a stale marker.
        setTimeout(() => {
            const r = this.relays.get(relayUrl);
            if (r)
                r.eosedSubIds.clear();
            this.resubscribeAll(relayUrl);
        }, AUTH_RESUBSCRIBE_DELAY_MS);
    }
    /**
     * Disconnect from all relays.
     */
    disconnect() {
        this.closed = true;
        // Clear pending OKs
        for (const [, pending] of this.pendingOks) {
            clearTimeout(pending.timer);
            pending.reject(new Error('Client disconnected'));
        }
        this.pendingOks.clear();
        // Clear queued events
        for (const item of this.eventQueue) {
            item.reject(new Error('Client disconnected'));
        }
        this.eventQueue = [];
        // Close all relay connections and clean up timers. Mark every
        // relay disconnected synchronously BEFORE we notify subscriptions
        // below, so any listener that consults `allRelaysDoneFor` sees
        // zero connected relays and settles immediately.
        for (const [url, relay] of this.relays) {
            relay.connected = false;
            if (relay.pingTimer) {
                clearInterval(relay.pingTimer);
                relay.pingTimer = null;
            }
            if (relay.reconnectTimer) {
                clearTimeout(relay.reconnectTimer);
                relay.reconnectTimer = null;
            }
            if (relay.socket && relay.socket.readyState !== CLOSED) {
                relay.socket.close(1000, 'Client disconnected');
            }
            this.emitConnectionEvent('disconnect', url, 'Client disconnected');
        }
        // Notify in-flight subscriptions that we're shutting down.
        // queryWithFirstSeenWins.onError re-checks allRelaysDoneFor (now
        // 0 connected → trivially true) and settles immediately, sparing
        // callers the full queryTimeoutMs wait. Snapshot keys first
        // because the listener may call unsubscribe(), which mutates
        // this.subscriptions while we iterate.
        const inflightSubs = Array.from(this.subscriptions.entries());
        for (const [subId, sub] of inflightSubs) {
            try {
                sub.listener.onError?.(subId, 'Client disconnected');
            }
            catch {
                // Ignore listener errors — we're tearing down anyway.
            }
        }
        this.relays.clear();
        this.subscriptions.clear();
    }
    /**
     * Check if connected to at least one relay.
     * @returns true if connected to at least one relay
     */
    isConnected() {
        for (const [, relay] of this.relays) {
            if (relay.connected)
                return true;
        }
        return false;
    }
    /**
     * Get the set of connected relay URLs.
     * @returns Set of connected relay URLs
     */
    getConnectedRelays() {
        const connected = new Set();
        for (const [url, relay] of this.relays) {
            if (relay.connected) {
                connected.add(url);
            }
        }
        return connected;
    }
    /**
     * Publish an event to all connected relays.
     * @param event Event to publish
     * @returns Promise that resolves with the event ID
     */
    async publishEvent(event) {
        if (this.closed) {
            throw new Error('Client has been disconnected');
        }
        if (!this.isConnected()) {
            // Queue the event for later
            return new Promise((resolve, reject) => {
                this.eventQueue.push({
                    event,
                    timestamp: Date.now(),
                    resolve,
                    reject,
                });
            });
        }
        return this.broadcastEvent(event);
    }
    /**
     * Broadcast an event to all connected relays.
     */
    async broadcastEvent(event) {
        const message = JSON.stringify(['EVENT', event.toJSON()]);
        let sent = false;
        for (const [, relay] of this.relays) {
            if (relay.connected && relay.socket?.readyState === OPEN) {
                relay.socket.send(message);
                sent = true;
            }
        }
        if (!sent) {
            throw new Error('No connected relays');
        }
        // Wait for at least one OK response
        return new Promise((resolve, reject) => {
            const timer = setTimeout(() => {
                this.pendingOks.delete(event.id);
                // Consider it successful if we sent it (some relays don't send OK)
                resolve(event.id);
            }, 5000);
            this.pendingOks.set(event.id, { resolve, reject, timer });
        });
    }
    /**
     * Publish an encrypted direct message (NIP-04).
     * @param recipientPubkeyHex Recipient's public key (hex)
     * @param message Message to send
     * @returns Promise that resolves with the event ID
     */
    async publishEncryptedMessage(recipientPubkeyHex, message) {
        const encryptedContent = await this.keyManager.encryptHex(message, recipientPubkeyHex);
        const event = Event.create(this.keyManager, {
            kind: EventKinds.ENCRYPTED_DM,
            tags: [['p', recipientPubkeyHex]],
            content: encryptedContent,
        });
        return this.publishEvent(event);
    }
    /**
     * Send a token transfer (encrypted).
     * @param recipientPubkeyHex Recipient's public key (hex)
     * @param tokenJson Token JSON string
     * @param options Optional parameters (amount, symbol, replyToEventId)
     * @returns Promise that resolves with the event ID
     */
    async sendTokenTransfer(recipientPubkeyHex, tokenJson, options) {
        const TokenTransferProtocol = await import('../token/TokenTransferProtocol.js');
        const event = await TokenTransferProtocol.createTokenTransferEvent(this.keyManager, recipientPubkeyHex, tokenJson, options);
        return this.publishEvent(event);
    }
    /**
     * Send a payment request to a target (encrypted).
     * @param targetPubkeyHex Target's public key (who should pay)
     * @param request Payment request details
     * @returns Promise that resolves with the event ID
     */
    async sendPaymentRequest(targetPubkeyHex, request) {
        const PaymentRequestProtocol = await import('../payment/PaymentRequestProtocol.js');
        const event = await PaymentRequestProtocol.createPaymentRequestEvent(this.keyManager, targetPubkeyHex, request);
        return this.publishEvent(event);
    }
    /**
     * Send a payment request response (decline/expiration notification).
     * @param targetPubkeyHex Original requester's public key
     * @param response Response details
     * @returns Promise that resolves with the event ID
     */
    async sendPaymentRequestResponse(targetPubkeyHex, response) {
        const PaymentRequestProtocol = await import('../payment/PaymentRequestProtocol.js');
        const event = await PaymentRequestProtocol.createPaymentRequestResponseEvent(this.keyManager, targetPubkeyHex, {
            requestId: response.requestId,
            originalEventId: response.originalEventId,
            status: response.status === 'DECLINED'
                ? PaymentRequestProtocol.ResponseStatus.DECLINED
                : PaymentRequestProtocol.ResponseStatus.EXPIRED,
            reason: response.reason,
        });
        return this.publishEvent(event);
    }
    /**
     * Send a payment request decline response.
     * Convenience method for declining a payment request.
     * @param originalRequestSenderPubkey Pubkey of who sent the original payment request
     * @param originalEventId Event ID of the original payment request
     * @param requestId Request ID from the original payment request
     * @param reason Optional reason for declining
     * @returns Promise that resolves with the event ID
     */
    async sendPaymentRequestDecline(originalRequestSenderPubkey, originalEventId, requestId, reason) {
        return this.sendPaymentRequestResponse(originalRequestSenderPubkey, {
            requestId,
            originalEventId,
            status: 'DECLINED',
            reason,
        });
    }
    /**
     * Publish a nametag binding.
     * Checks for existing claims by other pubkeys before publishing.
     * @param nametagId Nametag identifier
     * @param unicityAddress Unicity address
     * @returns Promise that resolves with success status
     * @throws Error if nametag is invalid or already claimed by another pubkey
     */
    async publishNametagBinding(nametagId, unicityAddress, identity) {
        // Check if already claimed by another pubkey
        const existingOwner = await this.queryPubkeyByNametag(nametagId);
        if (existingOwner && existingOwner !== this.keyManager.getPublicKeyHex()) {
            throw new Error(`Nametag "${nametagId}" is already claimed by another pubkey`);
        }
        const event = await createBindingEvent(this.keyManager, nametagId, unicityAddress, undefined, identity);
        try {
            await this.publishEvent(event);
            return true;
        }
        catch (err) {
            // UNIP-01: a relay-enforced single-owner rejection (NIP-20 `blocked:`)
            // means the nametag is owned by another key — surface it as a hard
            // failure rather than a silent `false`, which a caller could mistake for
            // a transient publish error. Other (transient) errors keep returning
            // false, preserving the prior best-effort behavior.
            const msg = err instanceof Error ? err.message : String(err);
            // Match only the ownership-conflict message, not the generic NIP-20
            // `blocked:` prefix (which also covers spam/rate-limit/policy rejections).
            if (/owned by another key|already claimed/i.test(msg)) {
                throw new Error(`Nametag "${nametagId}" is already claimed by another key`);
            }
            return false;
        }
    }
    /**
     * Publish a base identity binding (no nametag).
     * Uses d-tag = SHA256('unicity:identity:' + nostrPubkey) so each wallet
     * has exactly one identity binding. Subsequent calls replace the previous event.
     * @param identity Identity parameters (publicKey, l1Address, directAddress)
     * @returns true if published successfully
     */
    async publishIdentityBinding(identity) {
        const event = createIdentityBindingEvent(this.keyManager, identity);
        try {
            await this.publishEvent(event);
            return true;
        }
        catch {
            return false;
        }
    }
    subscribe(filterOrSubId, listenerOrFilter, maybeListener) {
        let subscriptionId;
        let filter;
        let listener;
        if (typeof filterOrSubId === 'string') {
            subscriptionId = filterOrSubId;
            filter = listenerOrFilter;
            listener = maybeListener;
        }
        else {
            subscriptionId = `sub_${++this.subscriptionCounter}`;
            filter = filterOrSubId;
            listener = listenerOrFilter;
        }
        // Reserved prefix for SDK-internal sub_ids (currently just the
        // keepalive `PING_SUB_ID`). Reject explicit caller use so the
        // keepalive timer's CLOSE/REQ cycle can't stomp on user state.
        if (subscriptionId.startsWith('__nostr-sdk-')) {
            throw new Error(`Subscription ID "${subscriptionId}" uses the reserved "__nostr-sdk-" prefix — pick a different id.`);
        }
        this.subscriptions.set(subscriptionId, { filter, listener });
        // Wipe any stale per-relay EOSE/CLOSED markers for this sub_id
        // before issuing the REQ — otherwise a fresh subscribe with a
        // sub_id that was previously CLOSED (or was just freshly
        // EOSE'd) would be skipped or treated as "already done" on
        // those relays.
        for (const [, relay] of this.relays) {
            relay.closedSubIds.delete(subscriptionId);
            relay.eosedSubIds.delete(subscriptionId);
        }
        // Send subscription request to all connected relays
        const message = JSON.stringify(['REQ', subscriptionId, filter.toJSON()]);
        for (const [, relay] of this.relays) {
            if (relay.connected && relay.socket?.readyState === OPEN) {
                relay.socket.send(message);
            }
        }
        return subscriptionId;
    }
    /**
     * Unsubscribe from a subscription.
     * @param subscriptionId Subscription ID to unsubscribe
     */
    unsubscribe(subscriptionId) {
        if (!this.subscriptions.has(subscriptionId))
            return;
        this.subscriptions.delete(subscriptionId);
        // Send CLOSE to all connected relays — except those that already
        // CLOSED the sub themselves (no point telling the relay something
        // it told us).
        const message = JSON.stringify(['CLOSE', subscriptionId]);
        for (const [, relay] of this.relays) {
            if (relay.connected && relay.socket?.readyState === OPEN
                && !relay.closedSubIds.has(subscriptionId)) {
                relay.socket.send(message);
            }
            // Drop both per-relay markers now that the sub is gone from
            // the global map.
            relay.closedSubIds.delete(subscriptionId);
            relay.eosedSubIds.delete(subscriptionId);
        }
    }
    /**
     * Query binding events and select the owner.
     *
     * UNIP-01 (preferred): if any returned binding carries the single-owner
     * namespace marker (NIP-32 ["L", "unicity:nametag"]), ownership is taken from
     * the marked binding. These are vetted by UNIP-01 relays for single ownership
     * (first author by relay receive order owns the identifier), so the
     * self-asserted `created_at` is ignored entirely. With a UNIP-01 relay set
     * there is exactly one marked owner; observing more than one distinct marked
     * author (e.g. across relays in differing states) is treated as ambiguous and
     * resolves to null rather than guessing.
     *
     * Legacy fallback (no marked binding present — e.g. an identifier not yet
     * migrated): first-seen-wins across authors by `created_at`, latest-wins for
     * the same author, lexicographic pubkey tie-break. Retained for backward
     * compatibility during migration; superseded as soon as the owner republishes
     * a marked binding. (Self-asserted timestamps are not authoritative — this
     * path exists only so un-migrated identifiers keep resolving.)
     *
     * Events with invalid signatures are silently skipped to prevent relay
     * injection of forged events.
     *
     * @param filter Subscription filter
     * @param extractResult Callback to extract the desired result from the winning event
     * @returns Promise resolving to the extracted result, or null
     */
    queryWithFirstSeenWins(filter, extractResult) {
        return new Promise((resolve) => {
            let subscriptionId = '';
            let settled = false;
            // Declared as `let` and initialized lazily so `finishWith` can be
            // invoked before the setTimeout call below without hitting the
            // TDZ on `clearTimeout(timeoutId)`. (The same comment on the
            // listener anticipates synchronous-callback hypothetical paths.)
            let timeoutId;
            // Accept an explicit `id` so callers from inside the listener can
            // pass the sub_id the relay echoed back. This guards against any
            // future change to subscribe() that would invoke listener
            // callbacks before its return value is bound to `subscriptionId`
            // — the closure-captured value would still be `''` and we'd skip
            // the CLOSE frame, leaking the slot on the relay.
            const finishWith = (result, id) => {
                if (settled)
                    return;
                settled = true;
                if (timeoutId !== undefined)
                    clearTimeout(timeoutId);
                const subId = id || subscriptionId;
                if (subId)
                    this.unsubscribe(subId);
                resolve(result);
            };
            timeoutId = setTimeout(() => finishWith(null), this.queryTimeoutMs);
            const authors = new Map();
            const allRelaysDone = (id) => this.allRelaysDoneFor(id);
            const pickWinner = () => {
                // UNIP-01: prefer marker-carrying (relay-vetted single-owner) bindings,
                // ignoring the self-asserted created_at. The marker is read from each
                // author's CURRENT (latest, per NIP-33 replaceable) event, so a newer
                // unmarked update correctly drops the author from the marked set, and a
                // marked owner resolves to their freshest binding.
                const marked = [...authors.values()].filter((e) => hasNametagOwnershipMarker(e.latestEvent));
                if (marked.length > 0) {
                    // Exactly one marked owner is the relay-enforced norm. More than one
                    // distinct marked author means cross-relay disagreement — do not guess.
                    const owner = marked.length === 1 ? marked[0] : undefined;
                    return owner ? extractResult(owner.latestEvent) : null;
                }
                // Legacy fallback: first-seen-wins by self-asserted created_at.
                let winnerEntry = null;
                let winnerPubkey = '';
                for (const [pubkey, entry] of authors) {
                    if (!winnerEntry
                        || entry.firstSeen < winnerEntry.firstSeen
                        || (entry.firstSeen === winnerEntry.firstSeen && pubkey < winnerPubkey)) {
                        winnerEntry = entry;
                        winnerPubkey = pubkey;
                    }
                }
                return winnerEntry ? extractResult(winnerEntry.latestEvent) : null;
            };
            subscriptionId = this.subscribe(filter, {
                onEvent: (event) => {
                    // Verify signature to prevent relay injection of forged events (#4)
                    if (!event.verify())
                        return;
                    const existing = authors.get(event.pubkey);
                    if (!existing) {
                        authors.set(event.pubkey, { firstSeen: event.created_at, latestEvent: event });
                    }
                    else {
                        if (event.created_at < existing.firstSeen) {
                            existing.firstSeen = event.created_at;
                        }
                        if (event.created_at > existing.latestEvent.created_at) {
                            existing.latestEvent = event;
                        }
                    }
                },
                // EOSE means *this relay* has finished delivering stored
                // events. In a multi-relay client we must not settle yet — a
                // slower relay may still be about to deliver matching events.
                // Settle only when every connected relay has either EOSE'd
                // OR CLOSED'd this sub. (Single-relay clients are unaffected:
                // allDone is trivially true with one relay.)
                onEndOfStoredEvents: (id) => {
                    if (allRelaysDone(id)) {
                        finishWith(pickWinner(), id);
                    }
                },
                // Subscription error from the SDK — fires from three paths
                // that all need the same "is it time to settle?" check:
                //   1. Relay sent CLOSED for this sub. In a multi-relay
                //      client the same sub_id may still be alive on a
                //      healthy relay; settling on the first CLOSED would
                //      prematurely abort a query other relays could
                //      satisfy. handleClosedMessage records the rejection
                //      on the sending relay's closedSubIds before invoking
                //      us, so we can decide via allRelaysDoneFor.
                //   2. Relay disconnected mid-query (socket.onclose →
                //      synthetic onError). The relay no longer counts as
                //      connected, so allRelaysDoneFor excludes it.
                //   3. Client disconnected (disconnect() → synthetic
                //      onError). All relays are torn down, allRelaysDoneFor
                //      sees zero connected and settles.
                onError: (id, message) => {
                    console.warn(`Subscription error on ${id}: ${message}`);
                    if (allRelaysDone(id)) {
                        finishWith(pickWinner(), id);
                    }
                    // else: keep waiting for EOSE / CLOSED from remaining
                    // relays or the overall query timeout.
                },
            });
        });
    }
    /**
     * True if every currently-connected relay has finished delivering
     * for the given sub_id (either EOSE'd or CLOSED'd it). Used by
     * queryWithFirstSeenWins to coordinate multi-relay settlement.
     */
    allRelaysDoneFor(subscriptionId) {
        const connected = Array.from(this.relays.values()).filter((r) => r.connected);
        // No connected relays at all → nothing to wait for; settle.
        if (connected.length === 0)
            return true;
        return connected.every((r) => r.eosedSubIds.has(subscriptionId) || r.closedSubIds.has(subscriptionId));
    }
    /**
     * Query for a public key by nametag.
     * Uses first-seen-wins anti-hijacking resolution.
     * @param nametagId Nametag identifier
     * @returns Promise that resolves with the public key hex, or null if not found
     */
    async queryPubkeyByNametag(nametagId) {
        return this.queryWithFirstSeenWins(createNametagToPubkeyFilter(nametagId), (event) => event.pubkey);
    }
    /**
     * Query for full binding info by nametag.
     * Returns extended identity fields (chain pubkey, addresses, etc.) when available.
     * Uses first-seen-wins across authors, latest-wins for same author.
     * @param nametagId Nametag identifier
     * @returns Promise that resolves with BindingInfo, or null if not found
     */
    async queryBindingByNametag(nametagId) {
        return this.queryWithFirstSeenWins(createNametagToPubkeyFilter(nametagId), parseBindingInfo);
    }
    /**
     * Query for binding info by address (reverse lookup).
     * Supports DIRECT://, PROXY://, alpha1..., or chain pubkey lookups.
     * Uses first-seen-wins across authors, latest-wins for same author.
     * @param address Address string
     * @returns Promise that resolves with BindingInfo, or null if not found
     */
    async queryBindingByAddress(address) {
        return this.queryWithFirstSeenWins(createAddressToBindingFilter(address), parseBindingInfo);
    }
    /**
     * Create and publish a signed event.
     * @param data Unsigned event data
     * @returns Promise that resolves with the event ID
     */
    async createAndPublishEvent(data) {
        const event = Event.create(this.keyManager, data);
        return this.publishEvent(event);
    }
    // ========== NIP-17 Private Messages ==========
    /**
     * Send a private message using NIP-17 gift-wrapping.
     * @param recipientPubkeyHex Recipient's public key (hex)
     * @param message Message content
     * @param options Optional message options (reply-to, etc.)
     * @returns Promise that resolves with the gift wrap event ID
     */
    async sendPrivateMessage(recipientPubkeyHex, message, options) {
        const giftWrap = NIP17.createGiftWrap(this.keyManager, recipientPubkeyHex, message, options);
        return this.publishEvent(giftWrap);
    }
    /**
     * Send a private message to a recipient identified by their nametag.
     * Resolves the nametag to a pubkey automatically.
     * @param recipientNametag Recipient's nametag (Unicity ID)
     * @param message Message content
     * @param options Optional message options (reply-to, etc.)
     * @returns Promise that resolves with the gift wrap event ID
     */
    async sendPrivateMessageToNametag(recipientNametag, message, options) {
        const pubkey = await this.queryPubkeyByNametag(recipientNametag);
        if (!pubkey) {
            throw new Error(`Nametag not found: ${recipientNametag}`);
        }
        return this.sendPrivateMessage(pubkey, message, options);
    }
    /**
     * Send a read receipt for a message using NIP-17 gift-wrapping.
     * @param recipientPubkeyHex Recipient (original sender) public key
     * @param messageEventId Event ID of the message being acknowledged
     * @returns Promise that resolves with the gift wrap event ID
     */
    async sendReadReceipt(recipientPubkeyHex, messageEventId) {
        const giftWrap = NIP17.createReadReceipt(this.keyManager, recipientPubkeyHex, messageEventId);
        return this.publishEvent(giftWrap);
    }
    /**
     * Unwrap a gift-wrapped private message.
     * @param giftWrap Gift wrap event (kind 1059)
     * @returns Parsed private message
     */
    unwrapPrivateMessage(giftWrap) {
        return NIP17.unwrap(giftWrap, this.keyManager);
    }
}
