// @unicitylabs/nostr-js-sdk/dist/esm/client/WebSocketAdapter.browser.js
/**
 * WebSocketAdapter - Browser implementation.
 * Uses the native WebSocket API.
 */
/** WebSocket ready state constants */
export const CONNECTING = 0;
export const OPEN = 1;
export const CLOSING = 2;
export const CLOSED = 3;
/**
 * Create a WebSocket connection using native browser WebSocket.
 * @param url WebSocket URL (ws:// or wss://)
 * @returns WebSocket instance
 */
export async function createWebSocket(url) {
    if (typeof WebSocket === 'undefined') {
        throw new Error('WebSocket not available in this environment');
    }
    return new WebSocket(url);
}
/**
 * Extract string data from WebSocket message event.
 * @param event WebSocket message event
 * @returns String message data
 */
export function extractMessageData(event) {
    if (typeof event.data === 'string') {
        return event.data;
    }
    if (event.data instanceof ArrayBuffer) {
        return new TextDecoder().decode(event.data);
    }
    if (typeof Blob !== 'undefined' && event.data instanceof Blob) {
        throw new Error('Blob messages are not supported');
    }
    return String(event.data);
}
