// @unicitylabs/state-transition-sdk/lib/api/bft/InputRecord.js
import { CborDeserializer } from '../../serialization/cbor/CborDeserializer.js';
import { CborError } from '../../serialization/cbor/CborError.js';
import { CborSerializer } from '../../serialization/cbor/CborSerializer.js';
import { HexConverter } from '../../util/HexConverter.js';
import { dedent } from '../../util/StringUtils.js';
/**
 * Input record for UnicityCertificate.
 */
export class InputRecord {
    roundNumber;
    epoch;
    _previousHash;
    _hash;
    _summaryValue;
    timestamp;
    _blockHash;
    sumOfEarnedFees;
    _executedTransactionsHash;
    static CBOR_TAG = 39002n;
    static VERSION = 1n;
    constructor(roundNumber, epoch, _previousHash, _hash, _summaryValue, timestamp, _blockHash, sumOfEarnedFees, _executedTransactionsHash) {
        this.roundNumber = roundNumber;
        this.epoch = epoch;
        this._previousHash = _previousHash;
        this._hash = _hash;
        this._summaryValue = _summaryValue;
        this.timestamp = timestamp;
        this._blockHash = _blockHash;
        this.sumOfEarnedFees = sumOfEarnedFees;
        this._executedTransactionsHash = _executedTransactionsHash;
        this._previousHash = _previousHash ? new Uint8Array(_previousHash) : null;
        this._hash = new Uint8Array(_hash);
        this._summaryValue = new Uint8Array(_summaryValue);
        this._blockHash = _blockHash ? new Uint8Array(_blockHash) : null;
        this._executedTransactionsHash = _executedTransactionsHash ? new Uint8Array(_executedTransactionsHash) : null;
    }
    /**
     * Get the block hash.
     */
    get blockHash() {
        return this._blockHash ? new Uint8Array(this._blockHash) : null;
    }
    /**
     * Get the executed transactions hash.
     */
    get executedTransactionsHash() {
        return this._executedTransactionsHash ? new Uint8Array(this._executedTransactionsHash) : null;
    }
    /**
     * Get the hash.
     */
    get hash() {
        return new Uint8Array(this._hash);
    }
    /**
     * Get the previous hash.
     */
    get previousHash() {
        return this._previousHash ? new Uint8Array(this._previousHash) : null;
    }
    /**
     * Get the summary value.
     */
    get summaryValue() {
        return new Uint8Array(this._summaryValue);
    }
    /**
     * @returns {bigint} Wire-format version of this input record.
     */
    get version() {
        return InputRecord.VERSION;
    }
    /**
     * Create InputRecord from CBOR bytes.
     *
     * @param bytes CBOR bytes
     * @return input record
     */
    static fromCBOR(bytes) {
        const tag = CborDeserializer.decodeTag(bytes);
        if (tag.tag !== InputRecord.CBOR_TAG) {
            throw new CborError(`Invalid CBOR tag for InputRecord: ${tag.tag}`);
        }
        const data = CborDeserializer.decodeArray(tag.data, 10);
        const version = CborDeserializer.decodeUnsignedInteger(data[0]);
        if (version !== InputRecord.VERSION) {
            throw new CborError(`Unsupported InputRecord version: ${version}`);
        }
        return new InputRecord(CborDeserializer.decodeUnsignedInteger(data[1]), CborDeserializer.decodeUnsignedInteger(data[2]), CborDeserializer.decodeNullable(data[3], CborDeserializer.decodeByteString), CborDeserializer.decodeByteString(data[4]), CborDeserializer.decodeByteString(data[5]), CborDeserializer.decodeUnsignedInteger(data[6]), CborDeserializer.decodeNullable(data[7], CborDeserializer.decodeByteString), CborDeserializer.decodeUnsignedInteger(data[8]), CborDeserializer.decodeNullable(data[9], CborDeserializer.decodeByteString));
    }
    /**
     * Convert InputRecord to CBOR bytes.
     *
     * @return CBOR bytes
     */
    toCBOR() {
        return CborSerializer.encodeTag(InputRecord.CBOR_TAG, CborSerializer.encodeArray(CborSerializer.encodeUnsignedInteger(this.version), CborSerializer.encodeUnsignedInteger(this.roundNumber), CborSerializer.encodeUnsignedInteger(this.epoch), CborSerializer.encodeNullable(this.previousHash, CborSerializer.encodeByteString), CborSerializer.encodeByteString(this.hash), CborSerializer.encodeByteString(this.summaryValue), CborSerializer.encodeUnsignedInteger(this.timestamp), CborSerializer.encodeNullable(this.blockHash, CborSerializer.encodeByteString), CborSerializer.encodeUnsignedInteger(this.sumOfEarnedFees), CborSerializer.encodeNullable(this.executedTransactionsHash, CborSerializer.encodeByteString)));
    }
    /**
     * Returns a string representation of the InputRecord.
     * @returns The string representation.
     */
    toString() {
        return dedent `
      Input Record
        Version: ${this.version.toString()}
        Round Number: ${this.roundNumber.toString()}
        Epoch: ${this.epoch.toString()}
        Previous Hash: ${this._previousHash ? HexConverter.encode(this._previousHash) : 'null'}
        Hash: ${HexConverter.encode(this._hash)}
        Summary Value: ${HexConverter.encode(this._summaryValue)}
        Timestamp: ${this.timestamp.toString()}
        Block Hash: ${this._blockHash ? HexConverter.encode(this._blockHash) : 'null'}
        Sum of Earned Fees: ${this.sumOfEarnedFees.toString()}
        Executed Transactions Hash: ${this._executedTransactionsHash ? HexConverter.encode(this._executedTransactionsHash) : 'null'}`;
    }
}
