// @unicitylabs/state-transition-sdk/lib/api/bft/RootTrustBase.js
import { InvalidJsonStructureError } from '../../InvalidJsonStructureError.js';
import { HexConverter } from '../../util/HexConverter.js';
import { dedent } from '../../util/StringUtils.js';
import { NetworkId } from '../NetworkId.js';
/**
 * Root trust base node information.
 */
export class RootTrustBaseNodeInfo {
    nodeId;
    _signingKey;
    stakedAmount;
    /**
     * RootTrustBaseNodeInfo constructor.
     * @param nodeId node id.
     * @param _signingKey signing key.
     * @param stakedAmount staked amount.
     */
    constructor(nodeId, _signingKey, stakedAmount) {
        this.nodeId = nodeId;
        this._signingKey = _signingKey;
        this.stakedAmount = stakedAmount;
        this._signingKey = new Uint8Array(_signingKey);
    }
    /**
     * Get the signing key.
     */
    get signingKey() {
        return new Uint8Array(this._signingKey);
    }
    /**
     * Create a RootTrustBaseNodeInfo from JSON representation.
     * @param input JSON representation.
     * @return RootTrustBaseNodeInfo instance.
     */
    static fromJSON(input) {
        if (!RootTrustBaseNodeInfo.isJSON(input)) {
            throw new InvalidJsonStructureError();
        }
        return new RootTrustBaseNodeInfo(input.nodeId, HexConverter.decode(input.sigKey), BigInt(input.stake));
    }
    /**
     * Check whether the input is a JSON representation of RootTrustBaseNodeInfo.
     * @param input input to check.
     * @return true if the input is a JSON representation of RootTrustBaseNodeInfo.
     */
    static isJSON(input) {
        return typeof input === 'object' && input !== null && 'nodeId' in input && 'sigKey' in input && 'stake' in input;
    }
}
/**
 * Root trust base information.
 */
export class RootTrustBase {
    version;
    networkId;
    epoch;
    epochStartRound;
    _rootNodes;
    quorumThreshold;
    _stateHash;
    _changeRecordHash;
    _previousEntryHash;
    _signatures;
    constructor(version, networkId, epoch, epochStartRound, _rootNodes, quorumThreshold, _stateHash, _changeRecordHash, _previousEntryHash, _signatures) {
        this.version = version;
        this.networkId = networkId;
        this.epoch = epoch;
        this.epochStartRound = epochStartRound;
        this._rootNodes = _rootNodes;
        this.quorumThreshold = quorumThreshold;
        this._stateHash = _stateHash;
        this._changeRecordHash = _changeRecordHash;
        this._previousEntryHash = _previousEntryHash;
        this._signatures = _signatures;
        this._rootNodes = _rootNodes.slice();
        this._stateHash = new Uint8Array(_stateHash);
        this._changeRecordHash = _changeRecordHash ? new Uint8Array(_changeRecordHash) : null;
        this._previousEntryHash = _previousEntryHash ? new Uint8Array(_previousEntryHash) : null;
        this._signatures = new Map(Array.from(_signatures.entries()).map((_signature) => [_signature[0], new Uint8Array(_signature[1])]));
    }
    /**
     * Get the change record hash.
     */
    get changeRecordHash() {
        return this._changeRecordHash ? new Uint8Array(this._changeRecordHash) : null;
    }
    /**
     * Get the previous entry hash.
     */
    get previousEntryHash() {
        return this._previousEntryHash ? new Uint8Array(this._previousEntryHash) : null;
    }
    /**
     * Get the root nodes.
     */
    get rootNodes() {
        return this._rootNodes.slice();
    }
    /**
     * Get the signatures.
     */
    get signatures() {
        return new Map(Array.from(this._signatures.entries()).map((_signature) => [_signature[0], new Uint8Array(_signature[1])]));
    }
    /**
     * Get the state hash.
     */
    get stateHash() {
        return new Uint8Array(this._stateHash);
    }
    /**
     * Create a root trust base from JSON string.
     *
     * @param input JSON string
     * @return root trust base
     */
    static fromJSON(input) {
        if (!RootTrustBase.isJSON(input)) {
            throw new InvalidJsonStructureError();
        }
        return new RootTrustBase(BigInt(input.version), NetworkId.fromId(input.networkId), BigInt(input.epoch), BigInt(input.epochStartRound), input.rootNodes.map((node) => RootTrustBaseNodeInfo.fromJSON(node)), BigInt(input.quorumThreshold), HexConverter.decode(input.stateHash), input.changeRecordHash ? HexConverter.decode(input.changeRecordHash) : null, input.previousEntryHash ? HexConverter.decode(input.previousEntryHash) : null, new Map(Object.entries(input.signatures).map(([id, signature]) => [id, HexConverter.decode(signature)])));
    }
    /**
     * Check whether the input is a JSON representation of RootTrustBase.
     * @param input input to check.
     * @return true if the input is a JSON representation of RootTrustBase.
     */
    static isJSON(input) {
        return (typeof input === 'object' &&
            input !== null &&
            'version' in input &&
            'networkId' in input &&
            'epoch' in input &&
            'epochStartRound' in input &&
            'rootNodes' in input &&
            Array.isArray(input.rootNodes) &&
            'quorumThreshold' in input &&
            'stateHash' in input &&
            'changeRecordHash' in input &&
            'previousEntryHash' in input &&
            'signatures' in input &&
            typeof input.signatures == 'object' &&
            input.signatures !== null);
    }
    /**
     * @returns {string} String representation of the trust base.
     */
    toString() {
        return dedent `
    RootTrustBase:
      Version: ${this.version},
      NetworkId: ${this.networkId.toString()},
      Epoch: ${this.epoch},
      EpochStartRound: ${this.epochStartRound},
      RootNodes: [${this.rootNodes.map((node) => node.nodeId).join(', ')}],
      QuorumThreshold: ${this.quorumThreshold},
      StateHash: ${HexConverter.encode(this.stateHash)},
      ChangeRecordHash: ${this.changeRecordHash ? HexConverter.encode(this.changeRecordHash) : 'null'},
      PreviousEntryHash: ${this.previousEntryHash ? HexConverter.encode(this.previousEntryHash) : 'null'},
      Signatures: [
        ${Array.from(this.signatures.entries())
            .map(([id, sig]) => `${id}: ${HexConverter.encode(sig)}`)
            .join(', ')}
      ]`;
    }
}
