// @unicitylabs/state-transition-sdk/lib/api/bft/ShardTreeCertificate.js
import { ShardId } from './ShardId.js';
import { CborDeserializer } from '../../serialization/cbor/CborDeserializer.js';
import { CborError } from '../../serialization/cbor/CborError.js';
import { CborSerializer } from '../../serialization/cbor/CborSerializer.js';
import { HexConverter } from '../../util/HexConverter.js';
import { dedent } from '../../util/StringUtils.js';
/**
 * Shard tree certificate.
 */
export class ShardTreeCertificate {
    _shard;
    _siblingHashList;
    static CBOR_TAG = 39003n;
    static VERSION = 1n;
    constructor(_shard, _siblingHashList) {
        this._shard = _shard;
        this._siblingHashList = _siblingHashList;
        this._siblingHashList = _siblingHashList.map((hash) => new Uint8Array(hash));
    }
    /**
     * Get the shard.
     */
    get shard() {
        return this._shard;
    }
    /**
     * Get the sibling hash list.
     */
    get siblingHashList() {
        return this._siblingHashList.map((hash) => new Uint8Array(hash));
    }
    /**
     * @returns {bigint} Wire-format version of this certificate.
     */
    get version() {
        return ShardTreeCertificate.VERSION;
    }
    /**
     * Create shard tree certificate from CBOR bytes.
     *
     * @param bytes CBOR bytes
     * @return shard tree certificate
     */
    static fromCBOR(bytes) {
        const tag = CborDeserializer.decodeTag(bytes);
        if (tag.tag !== ShardTreeCertificate.CBOR_TAG) {
            throw new CborError(`Invalid CBOR tag for ShardTreeCertificate: ${tag.tag}`);
        }
        const data = CborDeserializer.decodeArray(tag.data, 3);
        const version = CborDeserializer.decodeUnsignedInteger(data[0]);
        if (version !== ShardTreeCertificate.VERSION) {
            throw new CborError(`Unsupported ShardTreeCertificate version: ${version}`);
        }
        return new ShardTreeCertificate(ShardId.decode(CborDeserializer.decodeByteString(data[1])), CborDeserializer.decodeArray(data[2]).map((hash) => CborDeserializer.decodeByteString(hash)));
    }
    /**
     * Convert shard tree certificate to CBOR bytes.
     *
     * @return CBOR bytes
     */
    toCBOR() {
        return CborSerializer.encodeTag(ShardTreeCertificate.CBOR_TAG, CborSerializer.encodeArray(CborSerializer.encodeUnsignedInteger(this.version), CborSerializer.encodeByteString(this._shard.encode()), CborSerializer.encodeArray(...this._siblingHashList.map((hash) => CborSerializer.encodeByteString(hash)))));
    }
    /**
     * Returns a string representation of the ShardTreeCertificate.
     * @returns The string representation.
     */
    toString() {
        return dedent `
      Shard Tree Certificate
        Shard: ${this._shard.toString()}
        Sibling Hash List: [
          ${this._siblingHashList.map((hash) => HexConverter.encode(hash)).join('\n')}
        ]`;
    }
}
