// @unicitylabs/state-transition-sdk/lib/api/bft/UnicityCertificate.js
import { InputRecord } from './InputRecord.js';
import { ShardTreeCertificate } from './ShardTreeCertificate.js';
import { UnicitySeal } from './UnicitySeal.js';
import { UnicityTreeCertificate } from './UnicityTreeCertificate.js';
import { DataHasher } from '../../crypto/hash/DataHasher.js';
import { HashAlgorithm } from '../../crypto/hash/HashAlgorithm.js';
import { InvalidJsonStructureError } from '../../InvalidJsonStructureError.js';
import { CborDeserializer } from '../../serialization/cbor/CborDeserializer.js';
import { CborError } from '../../serialization/cbor/CborError.js';
import { CborSerializer } from '../../serialization/cbor/CborSerializer.js';
import { HexConverter } from '../../util/HexConverter.js';
import { dedent } from '../../util/StringUtils.js';
/**
 * Unicity certificate.
 */
export class UnicityCertificate {
    inputRecord;
    _technicalRecordHash;
    _shardConfigurationHash;
    shardTreeCertificate;
    unicityTreeCertificate;
    unicitySeal;
    static CBOR_TAG = 39001n;
    static VERSION = 1n;
    constructor(inputRecord, _technicalRecordHash, _shardConfigurationHash, shardTreeCertificate, unicityTreeCertificate, unicitySeal) {
        this.inputRecord = inputRecord;
        this._technicalRecordHash = _technicalRecordHash;
        this._shardConfigurationHash = _shardConfigurationHash;
        this.shardTreeCertificate = shardTreeCertificate;
        this.unicityTreeCertificate = unicityTreeCertificate;
        this.unicitySeal = unicitySeal;
        this._technicalRecordHash = _technicalRecordHash ? new Uint8Array(_technicalRecordHash) : null;
        this._shardConfigurationHash = new Uint8Array(_shardConfigurationHash);
    }
    /**
     * @returns {Uint8Array} Copy of the shard configuration hash.
     */
    get shardConfigurationHash() {
        return new Uint8Array(this._shardConfigurationHash);
    }
    /**
     * @returns {Uint8Array|null} Copy of the technical record hash, or `null` if absent.
     */
    get technicalRecordHash() {
        return this._technicalRecordHash ? new Uint8Array(this._technicalRecordHash) : null;
    }
    /**
     * @returns {bigint} Wire-format version of this certificate.
     */
    get version() {
        return UnicityCertificate.VERSION;
    }
    /**
     * Calculate the root hash of the shard tree certificate.
     *
     * @param {InputRecord} inputRecord            input record
     * @param {Uint8Array | null} technicalRecordHash    technical record hash
     * @param {Uint8Array} shardConfigurationHash shard configuration hash
     * @param {ShardTreeCertificate} shardTreeCertificate   shard tree certificate
     * @return root hash
     */
    static async calculateShardTreeCertificateRootHash(inputRecord, technicalRecordHash, shardConfigurationHash, shardTreeCertificate) {
        let rootHash = await new DataHasher(HashAlgorithm.SHA256)
            .update(inputRecord.toCBOR())
            .update(CborSerializer.encodeNullable(technicalRecordHash, CborSerializer.encodeByteString))
            .update(CborSerializer.encodeByteString(shardConfigurationHash))
            .digest();
        const shardId = shardTreeCertificate.shard;
        const siblingHashes = shardTreeCertificate.siblingHashList;
        for (let i = 0; i < siblingHashes.length; i++) {
            const isRight = shardId.getBit(shardId.length - 1 - i);
            if (isRight) {
                rootHash = await new DataHasher(HashAlgorithm.SHA256)
                    .update(CborSerializer.encodeByteString(siblingHashes[i]))
                    .update(CborSerializer.encodeByteString(rootHash.data))
                    .digest();
            }
            else {
                rootHash = await new DataHasher(HashAlgorithm.SHA256)
                    .update(CborSerializer.encodeByteString(rootHash.data))
                    .update(CborSerializer.encodeByteString(siblingHashes[i]))
                    .digest();
            }
        }
        return rootHash;
    }
    /**
     * Create unicity certificate from CBOR bytes.
     *
     * @param bytes CBOR bytes
     * @return unicity certificate
     */
    static fromCBOR(bytes) {
        const tag = CborDeserializer.decodeTag(bytes);
        if (tag.tag !== UnicityCertificate.CBOR_TAG) {
            throw new CborError(`Invalid CBOR tag for UnicityCertificate: ${tag.tag}`);
        }
        const data = CborDeserializer.decodeArray(tag.data, 7);
        const version = CborDeserializer.decodeUnsignedInteger(data[0]);
        if (version !== UnicityCertificate.VERSION) {
            throw new CborError(`Unsupported UnicityCertificate version: ${version}`);
        }
        return new UnicityCertificate(InputRecord.fromCBOR(data[1]), CborDeserializer.decodeNullable(data[2], CborDeserializer.decodeByteString), CborDeserializer.decodeByteString(data[3]), ShardTreeCertificate.fromCBOR(data[4]), UnicityTreeCertificate.fromCBOR(data[5]), UnicitySeal.fromCBOR(data[6]));
    }
    /**
     * Create UnicityCertificate from JSON.
     *
     * @param {unknown} data Hex string of CBOR-encoded certificate.
     * @returns {UnicityCertificate} Decoded certificate.
     * @throws {InvalidJsonStructureError} If the input is not a string.
     */
    static fromJSON(data) {
        if (!UnicityCertificate.isJSON(data)) {
            throw new InvalidJsonStructureError();
        }
        return UnicityCertificate.fromCBOR(HexConverter.decode(data));
    }
    static isJSON(input) {
        return typeof input === 'string';
    }
    /**
     * Convert unicity certificate to CBOR bytes.
     *
     * @return CBOR bytes
     */
    toCBOR() {
        return CborSerializer.encodeTag(UnicityCertificate.CBOR_TAG, CborSerializer.encodeArray(CborSerializer.encodeUnsignedInteger(this.version), this.inputRecord.toCBOR(), CborSerializer.encodeNullable(this.technicalRecordHash, CborSerializer.encodeByteString), CborSerializer.encodeByteString(this.shardConfigurationHash), this.shardTreeCertificate.toCBOR(), this.unicityTreeCertificate.toCBOR(), this.unicitySeal.toCBOR()));
    }
    /**
     * Returns a string representation of the UnicityCertificate.
     * @returns The string representation.
     */
    toString() {
        return dedent `
      Unicity Certificate
        Version: ${this.version.toString()}
        ${this.inputRecord.toString()}
        Technical Record Hash: ${this.technicalRecordHash ? HexConverter.encode(this.technicalRecordHash) : 'null'}
        Shard Configuration Hash: ${HexConverter.encode(this.shardConfigurationHash)}
        ${this.shardTreeCertificate.toString()}
        ${this.unicityTreeCertificate.toString()}
        ${this.unicitySeal.toString()}`;
    }
}
