// @unicitylabs/state-transition-sdk/lib/api/bft/UnicityTreeCertificate.js
import { CborDeserializer } from '../../serialization/cbor/CborDeserializer.js';
import { CborError } from '../../serialization/cbor/CborError.js';
import { CborSerializer } from '../../serialization/cbor/CborSerializer.js';
import { HexConverter } from '../../util/HexConverter.js';
import { dedent } from '../../util/StringUtils.js';
/**
 * Hash step in the certificate.
 */
class HashStep {
    key;
    _hash;
    constructor(key, _hash) {
        this.key = key;
        this._hash = _hash;
        this._hash = new Uint8Array(_hash);
    }
    /**
     * Get the hash.
     */
    get hash() {
        return new Uint8Array(this._hash);
    }
    /**
     * Create hash step from CBOR bytes.
     *
     * @param bytes CBOR bytes
     * @return hash step
     */
    static fromCBOR(bytes) {
        const data = CborDeserializer.decodeArray(bytes, 2);
        return new HashStep(CborDeserializer.decodeUnsignedInteger(data[0]), CborDeserializer.decodeByteString(data[1]));
    }
    /**
     * Convert hash step to CBOR bytes.
     *
     * @return CBOR bytes
     */
    toCBOR() {
        return CborSerializer.encodeArray(CborSerializer.encodeUnsignedInteger(this.key), CborSerializer.encodeByteString(this.hash));
    }
    /**
     * Returns a string representation of the UnicityTreeCertificate.
     * @returns The string representation.
     */
    toString() {
        return dedent `
      Hash Step
        Key: ${this.key}
        Hash: ${HexConverter.encode(this._hash)}`;
    }
}
/**
 * Unicity tree certificate.
 */
export class UnicityTreeCertificate {
    partitionIdentifier;
    _steps;
    static CBOR_TAG = 39004n;
    static VERSION = 1n;
    constructor(partitionIdentifier, _steps) {
        this.partitionIdentifier = partitionIdentifier;
        this._steps = _steps;
        this._steps = _steps.slice();
    }
    /**
     * Get the steps.
     */
    get steps() {
        return this._steps.slice();
    }
    /**
     * @returns {bigint} Wire-format version of this certificate.
     */
    get version() {
        return UnicityTreeCertificate.VERSION;
    }
    /**
     * Create certificate from CBOR bytes.
     *
     * @param bytes CBOR bytes
     * @return certificate
     */
    static fromCBOR(bytes) {
        const tag = CborDeserializer.decodeTag(bytes);
        if (tag.tag !== UnicityTreeCertificate.CBOR_TAG) {
            throw new CborError(`Invalid CBOR tag for UnicityTreeCertificate: ${tag.tag}`);
        }
        const data = CborDeserializer.decodeArray(tag.data, 3);
        const version = CborDeserializer.decodeUnsignedInteger(data[0]);
        if (version !== UnicityTreeCertificate.VERSION) {
            throw new CborError(`Unsupported UnicityTreeCertificate version: ${version}`);
        }
        return new UnicityTreeCertificate(CborDeserializer.decodeUnsignedInteger(data[1]), CborDeserializer.decodeArray(data[2]).map((step) => HashStep.fromCBOR(step)));
    }
    /**
     * Convert certificate to CBOR bytes.
     *
     * @return CBOR bytes
     */
    toCBOR() {
        return CborSerializer.encodeTag(UnicityTreeCertificate.CBOR_TAG, CborSerializer.encodeArray(CborSerializer.encodeUnsignedInteger(this.version), CborSerializer.encodeUnsignedInteger(this.partitionIdentifier), CborSerializer.encodeArray(...this.steps.map((step) => step.toCBOR()))));
    }
    /**
     * Returns a string representation of the UnicityTreeCertificate.
     * @returns The string representation.
     */
    toString() {
        return dedent `
      Unicity Tree Certificate
        Version: ${this.version}
        Partition Identifier: ${this.partitionIdentifier}
        Steps: [
          ${this.steps.map((step) => step.toString()).join('\n')}
        ]`;
    }
}
