// @unicitylabs/state-transition-sdk/lib/unicity-id/CertifiedUnicityIdMintTransaction.js
import { UnicityIdMintTransaction } from './UnicityIdMintTransaction.js';
import { InclusionProof } from '../api/InclusionProof.js';
import { CborDeserializer } from '../serialization/cbor/CborDeserializer.js';
import { CborSerializer } from '../serialization/cbor/CborSerializer.js';
import { InclusionProofVerificationRule, InclusionProofVerificationStatus, } from '../transaction/verification/rule/InclusionProofVerificationRule.js';
import { dedent } from '../util/StringUtils.js';
/**
 * Unicity-id mint transaction bundled with a verified inclusion proof.
 */
export class CertifiedUnicityIdMintTransaction {
    transaction;
    inclusionProof;
    constructor(transaction, inclusionProof) {
        this.transaction = transaction;
        this.inclusionProof = inclusionProof;
    }
    /**
     * @returns {Uint8Array} Data payload of the inner transaction.
     */
    get data() {
        return this.transaction.data;
    }
    /**
     * @returns {EncodedPredicate} Lock script of the inner transaction.
     */
    get lockScript() {
        return this.transaction.lockScript;
    }
    /**
     * @returns {EncodedPredicate} Recipient predicate of the inner transaction.
     */
    get recipient() {
        return this.transaction.recipient;
    }
    /**
     * @returns {DataHash} Source state hash of the inner transaction.
     */
    get sourceStateHash() {
        return this.transaction.sourceStateHash;
    }
    /**
     * @returns {Uint8Array} State mask of the inner transaction.
     */
    get stateMask() {
        return this.transaction.stateMask;
    }
    /**
     * @returns {SignaturePredicate} Target predicate of the inner transaction.
     */
    get targetPredicate() {
        return this.transaction.targetPredicate;
    }
    /**
     * @returns {TokenId} Token id of the inner transaction.
     */
    get tokenId() {
        return this.transaction.tokenId;
    }
    /**
     * @returns {TokenType} Token type of the inner transaction.
     */
    get tokenType() {
        return this.transaction.tokenType;
    }
    /**
     * @returns {UnicityId} Unicity id of the inner transaction.
     */
    get unicityId() {
        return this.transaction.unicityId;
    }
    /**
     * Create CertifiedUnicityIdMintTransaction from CBOR bytes.
     *
     * @param {Uint8Array} bytes CBOR bytes.
     * @returns {Promise<CertifiedUnicityIdMintTransaction>} Decoded certified transaction.
     */
    static async fromCBOR(bytes) {
        const data = CborDeserializer.decodeArray(bytes, 2);
        return new CertifiedUnicityIdMintTransaction(await UnicityIdMintTransaction.fromCBOR(data[0]), InclusionProof.fromCBOR(data[1]));
    }
    /**
     * Create CertifiedUnicityIdMintTransaction from unicity-id mint transaction and inclusion proof.
     *
     * @param {RootTrustBase} trustBase Root trust base.
     * @param {PredicateVerifierService} predicateVerifier Predicate verifier service.
     * @param {UnicityIdMintTransaction} transaction Transaction to certify.
     * @param {InclusionProof} inclusionProof Inclusion proof for the transaction.
     * @returns {Promise<CertifiedUnicityIdMintTransaction>} Verified certified transaction.
     * @throws {Error} If the inclusion proof does not verify.
     */
    static async fromTransaction(trustBase, predicateVerifier, transaction, inclusionProof) {
        const result = await InclusionProofVerificationRule.verify(trustBase, predicateVerifier, inclusionProof, transaction);
        if (result.status !== InclusionProofVerificationStatus.OK) {
            throw new Error(`Inclusion proof verification failed: ${result.status.toString()}`);
        }
        return new CertifiedUnicityIdMintTransaction(transaction, inclusionProof);
    }
    /**
     * @inheritDoc
     */
    calculateStateHash() {
        return this.transaction.calculateStateHash();
    }
    /**
     * @inheritDoc
     */
    calculateTransactionHash() {
        return this.transaction.calculateTransactionHash();
    }
    /**
     * @inheritDoc
     */
    toCBOR() {
        return CborSerializer.encodeArray(this.transaction.toCBOR(), this.inclusionProof.toCBOR());
    }
    /**
     * @returns {string} String representation of the certified transaction.
     */
    toString() {
        return dedent `
      CertifiedUnicityIdMintTransaction
        ${this.transaction.toString()}
        ${this.inclusionProof.toString()}`;
    }
}
