// @unicitylabs/state-transition-sdk/lib/unicity-id/UnicityIdMintTransaction.js
import { CertifiedUnicityIdMintTransaction } from './CertifiedUnicityIdMintTransaction.js';
import { UnicityId } from './UnicityId.js';
import { DataHasher } from '../crypto/hash/DataHasher.js';
import { HashAlgorithm } from '../crypto/hash/HashAlgorithm.js';
import { SignaturePredicate } from '../predicate/builtin/SignaturePredicate.js';
import { EncodedPredicate } from '../predicate/EncodedPredicate.js';
import { CborDeserializer } from '../serialization/cbor/CborDeserializer.js';
import { CborError } from '../serialization/cbor/CborError.js';
import { CborSerializer } from '../serialization/cbor/CborSerializer.js';
import { MintTransactionState } from '../transaction/MintTransactionState.js';
import { TokenType } from '../transaction/TokenType.js';
import { dedent } from '../util/StringUtils.js';
/**
 * Mint transaction for a unicity-id token.
 */
export class UnicityIdMintTransaction {
    sourceStateHash;
    lockScript;
    recipient;
    tokenId;
    tokenType;
    targetPredicate;
    unicityId;
    static CBOR_TAG = 39041n;
    static VERSION = 1n;
    constructor(sourceStateHash, lockScript, recipient, tokenId, tokenType, targetPredicate, unicityId) {
        this.sourceStateHash = sourceStateHash;
        this.lockScript = lockScript;
        this.recipient = recipient;
        this.tokenId = tokenId;
        this.tokenType = tokenType;
        this.targetPredicate = targetPredicate;
        this.unicityId = unicityId;
    }
    /**
     * @returns {Uint8Array} CBOR-encoded target predicate.
     */
    get data() {
        return EncodedPredicate.fromPredicate(this.targetPredicate).toCBOR();
    }
    /**
     * @returns {Uint8Array} State mask used when computing the resulting state hash.
     */
    get stateMask() {
        return new Uint8Array(this.tokenId.bytes);
    }
    /**
     * @returns {bigint} Wire-format version of this transaction.
     */
    get version() {
        return UnicityIdMintTransaction.VERSION;
    }
    /**
     * Create a UnicityIdMintTransaction.
     *
     * @param {SignaturePredicate} lockScript Issuer lock script.
     * @param {IPredicate} recipient Predicate that will lock the minted state.
     * @param {UnicityId} unicityId Unicity id being minted.
     * @param {TokenType} tokenType Token type.
     * @param {SignaturePredicate} targetPredicate Predicate the unicity id resolves to.
     * @returns {Promise<UnicityIdMintTransaction>} New mint transaction.
     */
    static async create(lockScript, recipient, unicityId, tokenType, targetPredicate) {
        const tokenId = await unicityId.toTokenId();
        return new UnicityIdMintTransaction(await MintTransactionState.create(tokenId), EncodedPredicate.fromPredicate(lockScript), EncodedPredicate.fromPredicate(recipient), tokenId, tokenType, targetPredicate, unicityId);
    }
    /**
     * Create UnicityIdMintTransaction from CBOR bytes.
     *
     * @param {Uint8Array} bytes CBOR bytes.
     * @returns {Promise<UnicityIdMintTransaction>} Decoded transaction.
     * @throws {CborError} On wrong tag or unsupported version.
     */
    static fromCBOR(bytes) {
        const tag = CborDeserializer.decodeTag(bytes);
        if (tag.tag !== UnicityIdMintTransaction.CBOR_TAG) {
            throw new CborError(`Invalid CBOR tag for UnicityIdMintTransaction: ${tag.tag}`);
        }
        const data = CborDeserializer.decodeArray(tag.data, 6);
        const version = CborDeserializer.decodeUnsignedInteger(data[0]);
        if (version !== UnicityIdMintTransaction.VERSION) {
            throw new CborError(`Unsupported UnicityIdMintTransaction version: ${version}`);
        }
        return UnicityIdMintTransaction.create(SignaturePredicate.fromPredicate(EncodedPredicate.fromCBOR(data[1])), EncodedPredicate.fromCBOR(data[2]), UnicityId.fromCBOR(data[3]), TokenType.fromCBOR(data[4]), SignaturePredicate.fromPredicate(EncodedPredicate.fromCBOR(data[5])));
    }
    /**
     * @inheritDoc
     */
    calculateStateHash() {
        return new DataHasher(HashAlgorithm.SHA256)
            .update(CborSerializer.encodeArray(CborSerializer.encodeByteString(this.sourceStateHash.imprint), CborSerializer.encodeByteString(this.stateMask)))
            .digest();
    }
    /**
     * @inheritDoc
     */
    calculateTransactionHash() {
        return new DataHasher(HashAlgorithm.SHA256).update(this.toCBOR()).digest();
    }
    /**
     * @inheritDoc
     */
    toCBOR() {
        return CborSerializer.encodeTag(UnicityIdMintTransaction.CBOR_TAG, CborSerializer.encodeArray(CborSerializer.encodeUnsignedInteger(this.version), this.lockScript.toCBOR(), this.recipient.toCBOR(), this.unicityId.toCBOR(), this.tokenType.toCBOR(), EncodedPredicate.fromPredicate(this.targetPredicate).toCBOR()));
    }
    /**
     * Bundle this transaction with its inclusion proof.
     *
     * @param {RootTrustBase} trustBase Root trust base used to verify the inclusion certificate.
     * @param {PredicateVerifierService} predicateVerifier Verifier for embedded predicates.
     * @param {InclusionProof} inclusionProof Inclusion proof for this transaction.
     * @returns {Promise<CertifiedUnicityIdMintTransaction>} Verified certified transaction.
     */
    toCertifiedTransaction(trustBase, predicateVerifier, inclusionProof) {
        return CertifiedUnicityIdMintTransaction.fromTransaction(trustBase, predicateVerifier, this, inclusionProof);
    }
    /**
     * @returns {string} String representation of the transaction.
     */
    toString() {
        return dedent `
      UnicityIdMintTransaction
        Version: ${this.version.toString()}
        Lock Script:
          ${this.lockScript.toString()}
        Recipient: ${this.recipient.toString()}
        Token ID: ${this.tokenId.toString()}
        Token Type: ${this.tokenType.toString()}
        Unicity ID:
          ${this.unicityId.toString()}
        Target Predicate:
          ${this.targetPredicate.toString()}`;
    }
}
