// @unicitylabs/nostr-js-sdk/dist/cjs/messaging/nip17.js
"use strict";
/**
 * NIP-17 Private Direct Messages Protocol.
 * Implements gift-wrapping for sender anonymity using NIP-44 encryption.
 *
 * Message flow:
 * 1. Create Rumor (kind 14, unsigned) with actual message content
 * 2. Create Seal (kind 13, signed by sender) encrypting the rumor
 * 3. Create Gift Wrap (kind 1059, signed by random ephemeral key) encrypting the seal
 *
 * Only the recipient can decrypt and verify the true sender.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.createGiftWrap = createGiftWrap;
exports.createReadReceipt = createReadReceipt;
exports.unwrap = unwrap;
const sha256_1 = require("@noble/hashes/sha256");
const utils_1 = require("@noble/hashes/utils");
const NostrKeyManager_js_1 = require("../NostrKeyManager.js");
const Event_js_1 = require("../protocol/Event.js");
const EventKinds = __importStar(require("../protocol/EventKinds.js"));
const NIP44 = __importStar(require("../crypto/nip44.js"));
const Schnorr = __importStar(require("../crypto/schnorr.js"));
// Randomization window for timestamps (+/- 2 days in seconds)
const TIMESTAMP_RANDOMIZATION = 2 * 24 * 60 * 60;
/**
 * Create a gift-wrapped private message.
 *
 * @param senderKeys Sender's key manager
 * @param recipientPubkeyHex Recipient's public key (hex)
 * @param content Message content
 * @param options Optional message options (reply-to, etc.)
 * @returns Gift-wrapped event (kind 1059)
 */
function createGiftWrap(senderKeys, recipientPubkeyHex, content, options) {
    // 1. Create Rumor (kind 14, unsigned)
    const rumor = createRumor(senderKeys.getPublicKeyHex(), recipientPubkeyHex, content, EventKinds.CHAT_MESSAGE, options?.replyToEventId);
    // 2. Create Seal (kind 13, signed by sender, encrypts rumor)
    const seal = createSeal(senderKeys, recipientPubkeyHex, rumor);
    // 3. Create Gift Wrap (kind 1059, signed by ephemeral key, encrypts seal)
    return wrapSeal(seal, recipientPubkeyHex);
}
/**
 * Create a gift-wrapped read receipt.
 *
 * @param senderKeys Sender's key manager
 * @param recipientPubkeyHex Recipient (original sender) public key
 * @param messageEventId Event ID of the message being acknowledged
 * @returns Gift-wrapped read receipt event
 */
function createReadReceipt(senderKeys, recipientPubkeyHex, messageEventId) {
    // Create rumor with kind 15 (read receipt)
    const tags = [
        ['p', recipientPubkeyHex],
        ['e', messageEventId],
    ];
    // Use actual timestamp for rumor (privacy via outer layers)
    const actualTimestamp = Math.floor(Date.now() / 1000);
    const rumor = {
        id: '', // Will be computed
        pubkey: senderKeys.getPublicKeyHex(),
        created_at: actualTimestamp,
        kind: EventKinds.READ_RECEIPT,
        tags,
        content: '', // Read receipts have empty content
    };
    // Compute the rumor ID
    rumor.id = computeRumorId(rumor);
    const seal = createSeal(senderKeys, recipientPubkeyHex, rumor);
    return wrapSeal(seal, recipientPubkeyHex);
}
/**
 * Unwrap a gift-wrapped message.
 *
 * @param giftWrap Gift wrap event (kind 1059)
 * @param recipientKeys Recipient's key manager
 * @returns Parsed private message
 */
function unwrap(giftWrap, recipientKeys) {
    if (giftWrap.kind !== EventKinds.GIFT_WRAP) {
        throw new Error(`Event is not a gift wrap (kind ${giftWrap.kind})`);
    }
    // Get ephemeral sender's pubkey from gift wrap
    const ephemeralPubkey = giftWrap.pubkey;
    const ephemeralPubkeyBytes = (0, utils_1.hexToBytes)(ephemeralPubkey);
    // Decrypt seal from gift wrap content
    const sealJson = NIP44.decrypt(giftWrap.content, recipientKeys.getPrivateKey(), ephemeralPubkeyBytes);
    const sealData = JSON.parse(sealJson);
    if (sealData.kind !== EventKinds.SEAL) {
        throw new Error(`Inner event is not a seal (kind ${sealData.kind})`);
    }
    // Verify seal signature
    const sealPubkey = sealData.pubkey;
    const sealIdBytes = (0, utils_1.hexToBytes)(sealData.id);
    const sigBytes = (0, utils_1.hexToBytes)(sealData.sig);
    const pubkeyBytes = (0, utils_1.hexToBytes)(sealPubkey);
    if (!Schnorr.verify(sigBytes, sealIdBytes, pubkeyBytes)) {
        throw new Error('Seal signature verification failed');
    }
    // Decrypt rumor from seal content
    const rumorJson = NIP44.decrypt(sealData.content, recipientKeys.getPrivateKey(), pubkeyBytes);
    const rumor = JSON.parse(rumorJson);
    // Extract reply-to event ID if present
    const replyToEventId = getTagValue(rumor.tags, 'e');
    return {
        eventId: giftWrap.id,
        senderPubkey: sealPubkey,
        recipientPubkey: recipientKeys.getPublicKeyHex(),
        content: rumor.content,
        timestamp: rumor.created_at,
        kind: rumor.kind,
        replyToEventId,
    };
}
// ========== Helper Functions ==========
/**
 * Create an unsigned rumor (kind 14 or 15).
 * Note: Rumor uses actual timestamp for correct message ordering.
 * Only seal and gift wrap use randomized timestamps for privacy.
 */
function createRumor(senderPubkey, recipientPubkey, content, kind, replyToEventId) {
    const tags = [['p', recipientPubkey]];
    if (replyToEventId) {
        tags.push(['e', replyToEventId, '', 'reply']);
    }
    // Use actual timestamp for rumor (inner message) - needed for correct ordering
    // Privacy is provided by randomized timestamps on seal and gift wrap layers
    const actualTimestamp = Math.floor(Date.now() / 1000);
    const rumor = {
        id: '', // Will be computed
        pubkey: senderPubkey,
        created_at: actualTimestamp,
        kind,
        tags,
        content,
    };
    // Compute the rumor ID
    rumor.id = computeRumorId(rumor);
    return rumor;
}
/**
 * Compute the rumor ID from serialized data.
 * ID = SHA-256([0, pubkey, created_at, kind, tags, content])
 */
function computeRumorId(rumor) {
    const serialized = JSON.stringify([
        0,
        rumor.pubkey,
        rumor.created_at,
        rumor.kind,
        rumor.tags,
        rumor.content,
    ]);
    const hash = (0, sha256_1.sha256)(new TextEncoder().encode(serialized));
    return (0, utils_1.bytesToHex)(hash);
}
/**
 * Create a seal (kind 13) that encrypts a rumor.
 */
function createSeal(senderKeys, recipientPubkeyHex, rumor) {
    const rumorJson = JSON.stringify(rumor);
    // Encrypt rumor with NIP-44
    const recipientPubkey = (0, utils_1.hexToBytes)(recipientPubkeyHex);
    const encryptedRumor = NIP44.encrypt(rumorJson, senderKeys.getPrivateKey(), recipientPubkey);
    // Create seal data
    const pubkey = senderKeys.getPublicKeyHex();
    const created_at = randomizeTimestamp();
    const kind = EventKinds.SEAL;
    const tags = []; // Seals have no tags
    const content = encryptedRumor;
    // Calculate ID
    const sealId = Event_js_1.Event.calculateId(pubkey, created_at, kind, tags, content);
    // Sign
    const sealIdBytes = (0, utils_1.hexToBytes)(sealId);
    const sig = senderKeys.signHex(sealIdBytes);
    return new Event_js_1.Event({
        id: sealId,
        pubkey,
        created_at,
        kind,
        tags,
        content,
        sig,
    });
}
/**
 * Wrap a seal in a gift wrap (kind 1059) using an ephemeral key.
 */
function wrapSeal(seal, recipientPubkeyHex) {
    // Generate ephemeral key for the gift wrap
    const ephemeralKeys = NostrKeyManager_js_1.NostrKeyManager.generate();
    const sealJson = JSON.stringify(seal.toJSON());
    // Encrypt seal with NIP-44 using ephemeral key
    const recipientPubkey = (0, utils_1.hexToBytes)(recipientPubkeyHex);
    const encryptedSeal = NIP44.encrypt(sealJson, ephemeralKeys.getPrivateKey(), recipientPubkey);
    // Create gift wrap data
    const pubkey = ephemeralKeys.getPublicKeyHex();
    const created_at = randomizeTimestamp();
    const kind = EventKinds.GIFT_WRAP;
    const tags = [['p', recipientPubkeyHex]];
    const content = encryptedSeal;
    // Calculate ID
    const giftWrapId = Event_js_1.Event.calculateId(pubkey, created_at, kind, tags, content);
    // Sign with ephemeral key
    const giftWrapIdBytes = (0, utils_1.hexToBytes)(giftWrapId);
    const sig = ephemeralKeys.signHex(giftWrapIdBytes);
    // Clear ephemeral key from memory
    ephemeralKeys.clear();
    return new Event_js_1.Event({
        id: giftWrapId,
        pubkey,
        created_at,
        kind,
        tags,
        content,
        sig,
    });
}
/**
 * Generate a randomized timestamp for privacy (+/- 2 days).
 */
function randomizeTimestamp() {
    const now = Math.floor(Date.now() / 1000);
    const randomOffset = Math.floor(Math.random() * 2 * TIMESTAMP_RANDOMIZATION) - TIMESTAMP_RANDOMIZATION;
    return now + randomOffset;
}
/**
 * Get the first value of a tag by name from a tags array.
 */
function getTagValue(tags, tagName) {
    const tag = tags.find((t) => t[0] === tagName);
    return tag?.[1];
}
