// @unicitylabs/nostr-js-sdk/dist/cjs/token/TokenTransferProtocol.js
"use strict";
/**
 * TokenTransferProtocol - Encapsulate Unicity token transfers over Nostr.
 * Uses NIP-04 encryption with optional GZIP compression.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.createTokenTransferEvent = createTokenTransferEvent;
exports.parseTokenTransfer = parseTokenTransfer;
exports.getAmount = getAmount;
exports.getSymbol = getSymbol;
exports.getReplyToEventId = getReplyToEventId;
exports.isTokenTransfer = isTokenTransfer;
exports.getRecipient = getRecipient;
exports.getSender = getSender;
const Event_js_1 = require("../protocol/Event.js");
const EventKinds = __importStar(require("../protocol/EventKinds.js"));
/** Prefix for token transfer messages */
const MESSAGE_PREFIX = 'token_transfer:';
/**
 * Create a token transfer event.
 *
 * Event structure:
 * - Kind: 31113 (TOKEN_TRANSFER - Unicity custom)
 * - Tags:
 *   - ["p", "<recipient_pubkey_hex>"] - Recipient
 *   - ["type", "token_transfer"] - Event type
 *   - ["amount", "<amount>"] - Optional amount
 *   - ["symbol", "<symbol>"] - Optional token symbol
 *   - ["e", "<event_id>", "", "reply"] - Optional reply-to event (for payment request correlation)
 * - Content: NIP-04 encrypted "token_transfer:{tokenJson}"
 *
 * @param keyManager Key manager with signing keys
 * @param recipientPubkeyHex Recipient's public key (hex)
 * @param tokenJson Token JSON string
 * @param amountOrOptions Optional amount for metadata, or options object
 * @param symbol Optional token symbol for metadata (ignored if options object used)
 * @returns Signed event
 */
async function createTokenTransferEvent(keyManager, recipientPubkeyHex, tokenJson, amountOrOptions, symbol) {
    // Parse options (support both old and new signatures)
    let amount;
    let tokenSymbol;
    let replyToEventId;
    if (amountOrOptions !== undefined && typeof amountOrOptions === 'object') {
        // New options object signature
        amount = amountOrOptions.amount;
        tokenSymbol = amountOrOptions.symbol;
        replyToEventId = amountOrOptions.replyToEventId;
    }
    else {
        // Old positional arguments signature
        amount = amountOrOptions;
        tokenSymbol = symbol;
    }
    // Encrypt the token data
    const message = MESSAGE_PREFIX + tokenJson;
    const encryptedContent = await keyManager.encryptHex(message, recipientPubkeyHex);
    // Build tags
    const tags = [
        ['p', recipientPubkeyHex],
        ['type', 'token_transfer'],
    ];
    if (amount !== undefined) {
        tags.push(['amount', String(amount)]);
    }
    if (tokenSymbol !== undefined) {
        tags.push(['symbol', tokenSymbol]);
    }
    // Add optional reply-to event reference (for payment request correlation)
    if (replyToEventId !== undefined && replyToEventId.length > 0) {
        tags.push(['e', replyToEventId, '', 'reply']);
    }
    const event = Event_js_1.Event.create(keyManager, {
        kind: EventKinds.TOKEN_TRANSFER,
        tags,
        content: encryptedContent,
    });
    return event;
}
/**
 * Parse a token transfer event.
 * Decrypts and decompresses the token data.
 *
 * @param event Token transfer event
 * @param keyManager Key manager for decryption
 * @returns Token JSON string
 * @throws Error if the event is not a valid token transfer
 */
async function parseTokenTransfer(event, keyManager) {
    // Verify event kind
    if (event.kind !== EventKinds.TOKEN_TRANSFER) {
        throw new Error('Event is not a token transfer');
    }
    // Verify event type tag
    const eventType = event.getTagValue('type');
    if (eventType !== 'token_transfer') {
        throw new Error('Event type is not token_transfer');
    }
    // Determine the sender's public key
    let senderPubkeyHex;
    if (keyManager.isMyPublicKey(event.pubkey)) {
        // We sent this event, decrypt with recipient's key
        const recipientPubkey = event.getTagValue('p');
        if (!recipientPubkey) {
            throw new Error('No recipient found in event');
        }
        senderPubkeyHex = recipientPubkey;
    }
    else {
        // We received this event, decrypt with sender's key
        senderPubkeyHex = event.pubkey;
    }
    // Decrypt the content
    let decrypted;
    try {
        decrypted = await keyManager.decryptHex(event.content, senderPubkeyHex);
    }
    catch (error) {
        // Fallback: try hex decoding for backward compatibility
        try {
            const { hexToBytes } = await Promise.resolve().then(() => __importStar(require('@noble/hashes/utils')));
            decrypted = new TextDecoder().decode(hexToBytes(event.content));
        }
        catch {
            throw error; // Re-throw original error
        }
    }
    // Strip the message prefix
    if (!decrypted.startsWith(MESSAGE_PREFIX)) {
        throw new Error('Invalid token transfer format');
    }
    return decrypted.slice(MESSAGE_PREFIX.length);
}
/**
 * Get the amount from a token transfer event.
 * @param event Token transfer event
 * @returns Amount, or undefined if not specified
 */
function getAmount(event) {
    const amount = event.getTagValue('amount');
    if (amount === undefined) {
        return undefined;
    }
    try {
        return BigInt(amount);
    }
    catch {
        return undefined;
    }
}
/**
 * Get the symbol from a token transfer event.
 * @param event Token transfer event
 * @returns Symbol, or undefined if not specified
 */
function getSymbol(event) {
    return event.getTagValue('symbol');
}
/**
 * Get the reply-to event ID from a token transfer event.
 * Used to correlate token transfers with payment requests.
 * @param event Token transfer event
 * @returns Referenced event ID, or undefined if not present
 */
function getReplyToEventId(event) {
    return event.getTagValue('e');
}
/**
 * Check if an event is a token transfer.
 * @param event Event to check
 * @returns true if the event is a token transfer
 */
function isTokenTransfer(event) {
    return (event.kind === EventKinds.TOKEN_TRANSFER &&
        event.getTagValue('type') === 'token_transfer');
}
/**
 * Get the recipient public key from a token transfer event.
 * @param event Token transfer event
 * @returns Recipient public key (hex), or undefined if not found
 */
function getRecipient(event) {
    return event.getTagValue('p');
}
/**
 * Get the sender public key from a token transfer event.
 * @param event Token transfer event
 * @returns Sender public key (hex)
 */
function getSender(event) {
    return event.pubkey;
}
